---
title: "Plotting"
date: '`r format(Sys.Date(), "%B %d %Y")`'
output:
  rmarkdown::html_vignette:
    toc_float: true
    toc_depth: 2
    vignette: >
      %\VignetteIndexEntry{Plotting}
      %\VignetteEncoding{UTF-8}
      %\VignetteEngine{knitr::rmarkdown}
---

```{r "setup", include = FALSE}
# Chunk opts
knitr::opts_chunk$set(
  collapse   = TRUE,
  comment    = "#>",
  warning    = FALSE,
  message    = FALSE,
  fig.width  = 7,
  fig.height = 4
)
```

<br>

This vignette provides detailed examples for plotting V(D)J data imported into a single-cell object using djvdj. For the examples shown below, we use data for splenocytes from BL6 and MD4 mice collected using the 10X Genomics scRNA-seq platform. MD4 B cells are monoclonal and specifically bind hen egg lysozyme.

```{r}
library(djvdj)
library(Seurat)
library(ggplot2)

# Add V(D)J data to object
vdj_dirs <- c(
  BL6 = system.file("extdata/splen/BL6_BCR", package = "djvdj"),
  MD4 = system.file("extdata/splen/MD4_BCR", package = "djvdj")
)

so <- splen_so |>
  import_vdj(
    vdj_dirs,
    define_clonotypes = "cdr3_gene",
    include_mutations = TRUE
  )
```

<br>

## Histograms

The `plot_histogram()` function can be used to plot numerical V(D)J data present in the object. By default, this will plot the values present in the `data_col` column for every cell. If `data_col` contains per-chain data, the values will be summarized for each cell. By default the mean will be calculated. The `trans` argument can be used to specify an x-axis transformation.

In the example below, the mean number of UMIs per cell is being plotted.

```{r}
so |>
  plot_histogram(
    data_col    = "umis",
    cluster_col = "orig.ident",
    trans       = "log10"
  )
```

The summary function can be changed using the `summary_fn` argument. The total number of UMIs per cell can be plotted by passing the `sum()` function.

```{r}
so |>
  plot_histogram(
    data_col    = "umis",
    cluster_col = "orig.ident",
    summary_fn  = sum,
    trans       = "log10"
  )
```

To only plot values for a specific chain, the `chain` argument can be used. In this example the mean number of UMIs is shown for each IGK chain.

```{r}
so |>
  plot_histogram(
    data_col    = "umis",
    cluster_col = "orig.ident",
    chain       = "IGK",
    trans       = "log10"
  )
```

The `method` argument can be used to create a density plot instead of a histogram.

```{r}
so |>
  plot_histogram(
    data_col    = "reads",
    cluster_col = "orig.ident",
    method      = "density",
    trans       = "log10"
  )
```

<br>

## Violin plots

```{r}
so |>
  plot_violin(
    data_col    = "reads",
    cluster_col = "sample",
    trans       = "log10"
  )
```

Instead of summarizing per-chain values for each cell, values can be plotted for each chain.

```{r}
so |>
  plot_violin(
    data_col    = "reads",
    cluster_col = "sample",
    trans       = "log10",
    per_chain   = TRUE
  )
```

<br>

## UMAP projections

To plot V(D)J information on UMAP projections, the `plot_scatter()` can be used. A summary function to use for summarizing per-chain values for each cell must be specified using the `summary_fn` argument. In this example we are plotting the mean number of mutations for each cell.

```{r, fig.width = 5, fig.height = 4}
so |>
  plot_scatter(data_col = "all_mis")
```

Instead of summarizing the per-chain values for each cell, we can also specify a specific chain to use for plotting. In this example we are plotting the CDR3 length for IGK chains. If a cell does not have an IGK chain or has multiple IGK chains, it will be plotted as an `NA`.

```{r, fig.width = 5.5, fig.height = 4}
so |>
  plot_scatter(
    data_col = "cdr3_length",
    chain    = "IGK"
  )
```

Like `plot_vdj()`, the scale can also be transformed using the `trans` argument.

```{r, fig.width = 5.25, fig.height = 4}
so |>
  plot_scatter(
    data_col    = "umis",
    trans       = "log10",
    plot_colors = c("#fd8d3c", "black")
  )
```

<br>

## Session info

```{r, echo = FALSE}
sessionInfo()
```
