<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Fetch per-chain V(D)J data from object. Within the object meta.data, each
row represents a single cell and can include information for multiple
chains. This function can return a data.frame where each row
represents a single chain. This is useful for plotting per-chain metrics
such as CDR3 length or the number of insertions/deletions."><title>Fetch V(D)J data from object — fetch_vdj • djvdj</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.7/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.7/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Fetch V(D)J data from object — fetch_vdj"><meta property="og:description" content="Fetch per-chain V(D)J data from object. Within the object meta.data, each
row represents a single cell and can include information for multiple
chains. This function can return a data.frame where each row
represents a single chain. This is useful for plotting per-chain metrics
such as CDR3 length or the number of insertions/deletions."><meta property="og:image" content="https://rnabioco.github.io/djvdj/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">djvdj</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Data import and manipulation</h6>
    <a class="dropdown-item" href="../articles/import.html">Data Import</a>
    <a class="dropdown-item" href="../articles/mutate.html">Data Manipulation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>VDJ analysis</h6>
    <a class="dropdown-item" href="../articles/frequency.html">Clonotype Frequencies</a>
    <a class="dropdown-item" href="../articles/diversity.html">Repertoire Diversity</a>
    <a class="dropdown-item" href="../articles/similarity.html">Repertoire Similarity</a>
    <a class="dropdown-item" href="../articles/gene-usage.html">Gene Usage</a>
    <a class="dropdown-item" href="../articles/clustering.html">CDR3 Clustering</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualization</h6>
    <a class="dropdown-item" href="../articles/plotting.html">Plotting</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/djvdj/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Fetch V(D)J data from object</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/djvdj/blob/HEAD/R/mutate-vdj.R" class="external-link"><code>R/mutate-vdj.R</code></a></small>
      <div class="d-none name"><code>fetch_vdj.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Fetch per-chain V(D)J data from object. Within the object meta.data, each
row represents a single cell and can include information for multiple
chains. This function can return a data.frame where each row
represents a single chain. This is useful for plotting per-chain metrics
such as CDR3 length or the number of insertions/deletions.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fetch_vdj</span><span class="op">(</span></span>
<span>  <span class="va">input</span>,</span>
<span>  data_cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clonotype_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filter_cells <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  per_chain <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  unnest <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  sep <span class="op">=</span> <span class="va">global</span><span class="op">$</span><span class="va">sep</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>input</dt>
<dd><p>Single cell object or data.frame containing V(D)J data. If a
data.frame is provided, the cell barcodes should be stored as row names.</p></dd>


<dt>data_cols</dt>
<dd><p>meta.data columns containing per-chain V(D)J data to unnest.
If NULL, V(D)J data are automatically selected by identifying columns that
have NAs in the same rows as clonotype_col.</p></dd>


<dt>clonotype_col</dt>
<dd><p>meta.data column containing clonotype IDs. This column
is used to determine which columns have V(D)J data. If both clonotype_col
and data_cols are NULL, all columns are included.</p></dd>


<dt>filter_cells</dt>
<dd><p>Remove cells that do not have V(D)J data, clonotype_col
must be provided to determine which cells to filter.</p></dd>


<dt>per_chain</dt>
<dd><p>If <code>TRUE</code> return per-chain data, i.e. each row represents a
chain.</p></dd>


<dt>unnest</dt>
<dd><p>If <code>FALSE</code>, a nested data.frame is returned where each row
represents a cell and V(D)J data is stored as list-cols. If <code>TRUE</code>, columns
are unnested so each row represents a single chain.</p></dd>


<dt>sep</dt>
<dd><p>Separator used for storing per cell V(D)J data. This is used to
identify columns containing per-chain data that can be unnested.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>data.frame containing V(D)J data</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Fetch per-chain V(D)J data</span></span></span>
<span class="r-in"><span><span class="fu">fetch_vdj</span><span class="op">(</span><span class="va">vdj_sce</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 216 × 50</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .cell_id             orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> 1_AAGCCGCAGCTTATCG-1 avid_1              0            0 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> 1_AATCCAGCATTACGAC-1 avid_1              6            4 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> 1_ACAGCTAGTCTGGTCG-1 avid_1             15            4 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> 1_ACCAGTAGTGCAGTAG-1 avid_1              4            2 1              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> 1_ACCTTTATCGACGGAA-1 avid_1              5            5 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> 1_ACGAGGAGTGACTCAT-1 avid_1              1            1 1              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> 1_ACGCAGCTCGTGACAT-1 avid_1              3            2 1              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> 1_ACGCCGACACGTCAGC-1 avid_1              5            2 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> 1_ACGGAGACATGCTGGC-1 avid_1              5            3 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> 1_ACTTTCATCGCTAGCG-1 avid_1              1            1 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 206 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 45 more variables: seurat_clusters &lt;fct&gt;, UMAP_1 &lt;dbl&gt;, UMAP_2 &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   clonotype_id &lt;chr&gt;, exact_subclonotype_id &lt;dbl&gt;, chains &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   n_chains &lt;int&gt;, cdr3 &lt;chr&gt;, cdr3_nt &lt;chr&gt;, cdr3_length &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   cdr3_nt_length &lt;dbl&gt;, v_gene &lt;chr&gt;, d_gene &lt;chr&gt;, j_gene &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   c_gene &lt;chr&gt;, isotype &lt;chr&gt;, reads &lt;dbl&gt;, umis &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   productive &lt;lgl&gt;, full_length &lt;lgl&gt;, paired &lt;lgl&gt;, v_ins &lt;dbl&gt;, …</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># To increase performance, specify which columns to return per-chain data,</span></span></span>
<span class="r-in"><span><span class="co"># per-cell data will be returned for all other columns</span></span></span>
<span class="r-in"><span><span class="fu">fetch_vdj</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">vdj_sce</span>,</span></span>
<span class="r-in"><span>  data_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chains"</span>, <span class="st">"reads"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 216 × 50</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .cell_id             orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> 1_AAGCCGCAGCTTATCG-1 avid_1              0            0 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> 1_AATCCAGCATTACGAC-1 avid_1              6            4 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> 1_ACAGCTAGTCTGGTCG-1 avid_1             15            4 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> 1_ACCAGTAGTGCAGTAG-1 avid_1              4            2 1              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> 1_ACCTTTATCGACGGAA-1 avid_1              5            5 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> 1_ACGAGGAGTGACTCAT-1 avid_1              1            1 1              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> 1_ACGCAGCTCGTGACAT-1 avid_1              3            2 1              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> 1_ACGCCGACACGTCAGC-1 avid_1              5            2 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> 1_ACGGAGACATGCTGGC-1 avid_1              5            3 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> 1_ACTTTCATCGCTAGCG-1 avid_1              1            1 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 206 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 45 more variables: seurat_clusters &lt;fct&gt;, UMAP_1 &lt;dbl&gt;, UMAP_2 &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   clonotype_id &lt;chr&gt;, exact_subclonotype_id &lt;dbl&gt;, chains &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   n_chains &lt;int&gt;, cdr3 &lt;chr&gt;, cdr3_nt &lt;chr&gt;, cdr3_length &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   cdr3_nt_length &lt;chr&gt;, v_gene &lt;chr&gt;, d_gene &lt;chr&gt;, j_gene &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   c_gene &lt;chr&gt;, isotype &lt;chr&gt;, reads &lt;dbl&gt;, umis &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   productive &lt;chr&gt;, full_length &lt;chr&gt;, paired &lt;lgl&gt;, v_ins &lt;chr&gt;, …</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Only include cells that have V(D)J data</span></span></span>
<span class="r-in"><span><span class="co"># clonotype_col must be specified to identify cells with V(D)J data</span></span></span>
<span class="r-in"><span><span class="fu">fetch_vdj</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">vdj_sce</span>,</span></span>
<span class="r-in"><span>  filter_cells <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  clonotype_col <span class="op">=</span> <span class="st">"clonotype_id"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 83 × 50</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .cell_id             orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> 1_ACGGAGACATGCTGGC-1 avid_1              5            3 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> 1_AGGCCACTCGGTCCGA-1 avid_1              3            2 1              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> 1_AGGCCACTCGGTCCGA-1 avid_1              3            2 1              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> 1_AGGCCACTCGGTCCGA-1 avid_1              3            2 1              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> 1_ATCATCTAGGCTAGAC-1 avid_1              3            3 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> 1_CACAGGCGTGGTCCGT-1 avid_1              9            4 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> 1_CAGCAGCGTAAAGTCA-1 avid_1              3            3 0              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> 1_CAGCTGGGTGCGATAG-1 avid_1              2            2 1              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> 1_CAGCTGGGTGCGATAG-1 avid_1              2            2 1              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> 1_CAGCTGGGTGCGATAG-1 avid_1              2            2 1              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 73 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 45 more variables: seurat_clusters &lt;fct&gt;, UMAP_1 &lt;dbl&gt;, UMAP_2 &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   clonotype_id &lt;chr&gt;, exact_subclonotype_id &lt;dbl&gt;, chains &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   n_chains &lt;int&gt;, cdr3 &lt;chr&gt;, cdr3_nt &lt;chr&gt;, cdr3_length &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   cdr3_nt_length &lt;dbl&gt;, v_gene &lt;chr&gt;, d_gene &lt;chr&gt;, j_gene &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   c_gene &lt;chr&gt;, isotype &lt;chr&gt;, reads &lt;dbl&gt;, umis &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   productive &lt;lgl&gt;, full_length &lt;lgl&gt;, paired &lt;lgl&gt;, v_ins &lt;dbl&gt;, …</span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

