<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Import V(D)J data"><title>Import V(D)J data — import_vdj • djvdj</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Import V(D)J data — import_vdj"><meta property="og:description" content="Import V(D)J data"><meta property="og:image" content="https://rnabioco.github.io/djvdj/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">djvdj</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/avid.html">V(D)J analysis</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/djvdj/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Import V(D)J data</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/djvdj/blob/HEAD/R/import-vdj.R" class="external-link"><code>R/import-vdj.R</code></a></small>
      <div class="d-none name"><code>import_vdj.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Import V(D)J data</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">import_vdj</span><span class="op">(</span>
  input <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">vdj_dir</span>,
  prefix <span class="op">=</span> <span class="st">""</span>,
  cell_prefix <span class="op">=</span> <span class="cn">NULL</span>,
  filter_chains <span class="op">=</span> <span class="cn">TRUE</span>,
  filter_paired <span class="op">=</span> <span class="cn">FALSE</span>,
  define_clonotypes <span class="op">=</span> <span class="cn">NULL</span>,
  include_indels <span class="op">=</span> <span class="cn">TRUE</span>,
  sep <span class="op">=</span> <span class="st">";"</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>input</dt>
<dd><p>Object containing single cell data, if set to NULL a data.frame
containing V(D)J results will be returned</p></dd>
<dt>vdj_dir</dt>
<dd><p>Directory containing the output from cellranger vdj. A vector
or named vector can be given to load data from multiple runs. If a named
vector is given, the cell barcodes will be prefixed with the provided names.
This mimics the behavior of Seurat::Read10X(). Cell barcode prefixes can
also be provided using the cell_prefix argument.</p></dd>
<dt>prefix</dt>
<dd><p>Prefix to add to new columns</p></dd>
<dt>cell_prefix</dt>
<dd><p>Prefix to add to cell barcodes, this is helpful when
loading data from multiple runs into a single object. If NULL, cell barcode
prefixes are automatically generated in a similar manner as
Seurat::Read10X().</p>
<p>For the V(D)J data to be successfully added to the object, the cell prefixes
must match the prefixes that are already present in the object. If the cell
prefixes are incorrect, import_vdj will be unable to assign the V(D)J data
to the correct cells.</p></dd>
<dt>filter_chains</dt>
<dd><p>Only include chains with at least one productive and
full length contig.</p></dd>
<dt>filter_paired</dt>
<dd><p>Only include clonotypes with paired chains. For TCR
data each clonotype must have at least one TRA and TRB chain, for BCR data
each clonotype must have at least one IGH chain and at least one IGK or IGL
chain.</p></dd>
<dt>define_clonotypes</dt>
<dd><p>Define clonotype IDs based on V(D)J data. This is
useful if the V(D)J datasets being loaded do not have consistent clonotype
IDs, i.e., clonotype1 is not the same across samples. Possible values are:</p><ul><li><p>cdr3aa, define clonotypes based on the CDR3 amino acid sequence</p></li>
<li><p>cdr3nt, define clonotypes based on the CDR3 nucleotide sequence</p></li>
<li><p>cdr3_gene, define clonotypes based on the combination of the CDR3
nucleotide sequence and the V(D)J genes.</p></li>
</ul><p>When defining clonotypes, only productive full length chains will be used.
Set to NULL (default) to use the clonotype IDs already present in the input
data.</p></dd>
<dt>include_indels</dt>
<dd><p>Include the number of insertions/deletions for each
chain. This requires the concat_ref.bam file from cellranger vdj to be
present the directory provided to vdj_dir. If include_indels is TRUE,
filter_chains is also automatically set TRUE since indel data is only
available for productive chains.</p></dd>
<dt>sep</dt>
<dd><p>Separator to use for storing per cell V(D)J data</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Single cell object or data.frame with added V(D)J data</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co"># Loading multiple datasets</span></span>
<span class="r-in"><span class="va">vdj_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span class="r-in">  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/bcr_1"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span>,</span>
<span class="r-in">  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/bcr_2"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span><span class="va">tiny_so</span>, <span class="va">vdj_dir</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1     avid_1          1            1               1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters   UMAP_1     UMAP_2 clonotype_id   v_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1               1 2.561658 -0.6746224   clonotype1 IGKV5-43</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      d_gene j_gene c_gene chains isotype        cdr3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1   None  IGKJ2   IGKC    IGK    None CQQSNSWPYTF</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                cdr3_nt reads umis productive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1 TGTCAACAGAGTAACAGCTGGCCGTACACGTTC  1054   70       TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      full_length paired n_insertion n_deletion n_mismatch</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1        TRUE  FALSE           0          4          2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      cdr3_length cdr3_nt_length n_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1          11             33        1</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Specifying cell prefixes</span></span>
<span class="r-in"><span class="co"># if cell prefixes are not specified when loading multiple datasets,</span></span>
<span class="r-in"><span class="co"># prefixes will be automatically generated in a similar manner as</span></span>
<span class="r-in"><span class="co"># Seurat::Read10X</span></span>
<span class="r-in"><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span></span>
<span class="r-in">  <span class="va">tiny_so</span>,</span>
<span class="r-in">  vdj_dir <span class="op">=</span> <span class="va">vdj_dir</span>,</span>
<span class="r-in">  cell_prefix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1_"</span>, <span class="st">"2_"</span><span class="op">)</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1     avid_1          1            1               1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters   UMAP_1     UMAP_2 clonotype_id   v_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1               1 2.561658 -0.6746224   clonotype1 IGKV5-43</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      d_gene j_gene c_gene chains isotype        cdr3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1   None  IGKJ2   IGKC    IGK    None CQQSNSWPYTF</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                cdr3_nt reads umis productive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1 TGTCAACAGAGTAACAGCTGGCCGTACACGTTC  1054   70       TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      full_length paired n_insertion n_deletion n_mismatch</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1        TRUE  FALSE           0          4          2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      cdr3_length cdr3_nt_length n_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1          11             33        1</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Specifying cell prefixes using vector names</span></span>
<span class="r-in"><span class="co"># if a named vector is passed, the names will be used as the cell prefixes</span></span>
<span class="r-in"><span class="va">vdj_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span class="r-in">  <span class="st">"1_"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/bcr_1"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span>,</span>
<span class="r-in">  <span class="st">"2_"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/bcr_2"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span><span class="va">tiny_so</span>, <span class="va">vdj_dir</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1     avid_1          1            1               1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters   UMAP_1     UMAP_2 clonotype_id   v_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1               1 2.561658 -0.6746224   clonotype1 IGKV5-43</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      d_gene j_gene c_gene chains isotype        cdr3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1   None  IGKJ2   IGKC    IGK    None CQQSNSWPYTF</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                cdr3_nt reads umis productive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1 TGTCAACAGAGTAACAGCTGGCCGTACACGTTC  1054   70       TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      full_length paired n_insertion n_deletion n_mismatch</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1        TRUE  FALSE           0          4          2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      cdr3_length cdr3_nt_length n_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1          11             33        1</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Only include V(D)J data for productive full length chains</span></span>
<span class="r-in"><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span></span>
<span class="r-in">  <span class="va">tiny_so</span>,</span>
<span class="r-in">  vdj_dir <span class="op">=</span> <span class="va">vdj_dir</span>,</span>
<span class="r-in">  filter_chains <span class="op">=</span> <span class="cn">TRUE</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1     avid_1          1            1               1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters   UMAP_1     UMAP_2 clonotype_id   v_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1               1 2.561658 -0.6746224   clonotype1 IGKV5-43</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      d_gene j_gene c_gene chains isotype        cdr3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1   None  IGKJ2   IGKC    IGK    None CQQSNSWPYTF</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                cdr3_nt reads umis productive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1 TGTCAACAGAGTAACAGCTGGCCGTACACGTTC  1054   70       TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      full_length paired n_insertion n_deletion n_mismatch</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1        TRUE  FALSE           0          4          2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      cdr3_length cdr3_nt_length n_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1          11             33        1</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Only include V(D)J data for cells with paired chains</span></span>
<span class="r-in"><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span></span>
<span class="r-in">  <span class="va">tiny_so</span>,</span>
<span class="r-in">  vdj_dir <span class="op">=</span> <span class="va">vdj_dir</span>,</span>
<span class="r-in">  filter_paired <span class="op">=</span> <span class="cn">TRUE</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1     avid_1          1            1               1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters   UMAP_1     UMAP_2 clonotype_id v_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1               1 2.561658 -0.6746224         &lt;NA&gt;   &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      d_gene j_gene c_gene chains isotype cdr3 cdr3_nt reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1   &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;    &lt;NA&gt; &lt;NA&gt;    &lt;NA&gt;  &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      umis productive full_length paired n_insertion n_deletion</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1 &lt;NA&gt;       &lt;NA&gt;        &lt;NA&gt;     NA        &lt;NA&gt;       &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      n_mismatch cdr3_length cdr3_nt_length n_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1       &lt;NA&gt;        &lt;NA&gt;           &lt;NA&gt;       NA</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Defining clonotypes</span></span>
<span class="r-in"><span class="co"># this is useful if the original clonotype IDs are not consistent across</span></span>
<span class="r-in"><span class="co"># datasets, i.e. clonotype1 is not the same for all samples</span></span>
<span class="r-in"><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span></span>
<span class="r-in">  <span class="va">tiny_so</span>,</span>
<span class="r-in">  vdj_dir <span class="op">=</span> <span class="va">vdj_dir</span>,</span>
<span class="r-in">  define_clonotypes <span class="op">=</span> <span class="st">"cdr3_gene"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1     avid_1          1            1               1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters   UMAP_1     UMAP_2 clonotype_id   v_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1               1 2.561658 -0.6746224  clonotype16 IGKV5-43</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      d_gene j_gene c_gene chains isotype        cdr3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1   None  IGKJ2   IGKC    IGK    None CQQSNSWPYTF</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                cdr3_nt reads umis productive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1 TGTCAACAGAGTAACAGCTGGCCGTACACGTTC  1054   70       TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      full_length paired n_insertion n_deletion n_mismatch</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1        TRUE  FALSE           0          4          2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      cdr3_length cdr3_nt_length n_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1          11             33        1</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Omit indel information for each chain</span></span>
<span class="r-in"><span class="co"># this information will be included if the file concat_ref.bam is present</span></span>
<span class="r-in"><span class="co"># to speed up data import, omit indel information</span></span>
<span class="r-in"><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span></span>
<span class="r-in">  <span class="va">tiny_so</span>,</span>
<span class="r-in">  vdj_dir <span class="op">=</span> <span class="va">vdj_dir</span>,</span>
<span class="r-in">  include_indels <span class="op">=</span> <span class="cn">FALSE</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1     avid_1          1            1               1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters   UMAP_1     UMAP_2 clonotype_id   v_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1               1 2.561658 -0.6746224   clonotype1 IGKV5-43</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      d_gene j_gene c_gene chains isotype        cdr3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1   None  IGKJ2   IGKC    IGK    None CQQSNSWPYTF</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                cdr3_nt reads umis productive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1 TGTCAACAGAGTAACAGCTGGCCGTACACGTTC  1054   70       TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      full_length paired cdr3_length cdr3_nt_length n_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1        TRUE  FALSE          11             33        1</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Loading both BCR and TCR data</span></span>
<span class="r-in"><span class="co"># load each dataset separately and assign unique column names using the</span></span>
<span class="r-in"><span class="co"># prefix argument</span></span>
<span class="r-in"><span class="va">bcr_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/bcr_1"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">tcr_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/tcr_1"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span></span>
<span class="r-in">  <span class="va">tiny_so</span>,</span>
<span class="r-in">  vdj_dir     <span class="op">=</span> <span class="va">bcr_dir</span>,</span>
<span class="r-in">  prefix      <span class="op">=</span> <span class="st">"bcr_"</span>,</span>
<span class="r-in">  cell_prefix <span class="op">=</span> <span class="st">"1_"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span></span>
<span class="r-in">  <span class="va">vdj_so</span>,</span>
<span class="r-in">  vdj_dir     <span class="op">=</span> <span class="va">tcr_dir</span>,</span>
<span class="r-in">  prefix      <span class="op">=</span> <span class="st">"tcr_"</span>,</span>
<span class="r-in">  cell_prefix <span class="op">=</span> <span class="st">"1_"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1     avid_1          1            1               1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters   UMAP_1     UMAP_2 bcr_clonotype_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1               1 2.561658 -0.6746224       clonotype1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      bcr_v_gene bcr_d_gene bcr_j_gene bcr_c_gene bcr_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1   IGKV5-43       None      IGKJ2       IGKC        IGK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      bcr_isotype    bcr_cdr3                       bcr_cdr3_nt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1        None CQQSNSWPYTF TGTCAACAGAGTAACAGCTGGCCGTACACGTTC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      bcr_reads bcr_umis bcr_productive bcr_full_length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1      1054       70           TRUE            TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      bcr_paired bcr_n_insertion bcr_n_deletion bcr_n_mismatch</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1      FALSE               0              4              2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      bcr_cdr3_length bcr_cdr3_nt_length bcr_n_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1              11                 33            1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      tcr_clonotype_id       tcr_v_gene tcr_d_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1     clonotype116 TRAV12D-2;TRBV20  None;None</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                          tcr_j_gene tcr_c_gene tcr_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1 TRAJ40;TRBJ2-5  TRAC;None    TRA;TRB</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                          tcr_cdr3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1 CALSDRWDSGNYKYVF;CGAGAGDTQYF</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                                                             tcr_cdr3_nt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1 TGTGCTTTGAGTGATCGATGGGACTCAGGAAACTACAAATACGTCTTT;TGTGGTGCTGGGGCGGGGGACACCCAGTACTTT</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      tcr_reads tcr_umis tcr_productive tcr_full_length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1   244;856      3;6      TRUE;TRUE       TRUE;TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      tcr_paired tcr_n_insertion tcr_n_deletion tcr_n_mismatch</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1       TRUE             4;0            0;2            5;6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      tcr_cdr3_length tcr_cdr3_nt_length tcr_n_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1           16;11              48;33            2</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Using import_vdj outside of Seurat</span></span>
<span class="r-in"><span class="co"># SingleCellExperiment objects are also compatible, or if an input object is</span></span>
<span class="r-in"><span class="co"># omitted, a data.frame containing the V(D)J data will be returned</span></span>
<span class="r-in"><span class="va">vdj_sce</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span><span class="va">tiny_sce</span>, <span class="va">vdj_dir</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_sce</span><span class="op">@</span><span class="va">colData</span>, <span class="fl">1</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 27 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      &lt;character&gt;  &lt;numeric&gt;    &lt;integer&gt;        &lt;factor&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1      avid_1          1            1               1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters    UMAP_1    UMAP_2 clonotype_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                             &lt;factor&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1               1   2.56166 -0.674622   clonotype1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           v_gene      d_gene      j_gene      c_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1    IGKV5-43        None       IGKJ2        IGKC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           chains     isotype        cdr3                cdr3_nt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      &lt;character&gt; &lt;character&gt; &lt;character&gt;            &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1         IGK        None CQQSNSWPYTF TGTCAACAGAGTAACAGCTG..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                            reads        umis  productive full_length    paired</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;logical&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1        1054          70        TRUE        TRUE     FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      n_insertion  n_deletion  n_mismatch cdr3_length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1           0           4           2          11</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      cdr3_nt_length  n_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                         &lt;character&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAACCTGAGTAGGTGC-1             33         1</span>
<span class="r-in"></span>
<span class="r-in"><span class="va">vdj_df</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span>vdj_dir <span class="op">=</span> <span class="va">vdj_dir</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_df</span>, <span class="fl">1</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       clonotype_id                     v_gene         d_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_CTTCTCTTCACCTTAT-1 clonotype2350 IGHV5-17;IGKV5-43;IGKV5-43 None;None;None</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 j_gene         c_gene      chains isotype</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_CTTCTCTTCACCTTAT-1 IGHJ4;IGKJ2;IGKJ2 IGHA;IGKC;IGKC IGH;IGK;IGK    IGHA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                             cdr3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_CTTCTCTTCACCTTAT-1 CARPYYYTTIYVDYAMAYW;CQQSNSWPYTF;CQQSNCWPFTF</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                                                                                                                            cdr3_nt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_CTTCTCTTCACCTTAT-1 TGTGCAAGGCCTTATTACTACACTACTATCTACGTTGACTACGCTATGGCCTACTGG;TGTCAACAGAGTAACAGCTGGCCGTACACGTTC;TGTCAACAGAGTAACTGCTGGCCGTTCACGTTC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                 reads          umis     productive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_CTTCTCTTCACCTTAT-1 14930;18030;4334 1112;1408;351 TRUE;TRUE;TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                         full_length paired n_insertion n_deletion n_mismatch</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_CTTCTCTTCACCTTAT-1 TRUE;TRUE;TRUE   TRUE      25;0;0      0;4;4     14;2;6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      cdr3_length cdr3_nt_length n_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_CTTCTCTTCACCTTAT-1    19;11;11       57;33;33        3</span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Ryan Sheridan, <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer></div>

  

  

  </body></html>

