<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Import V(D)J data"><title>Import V(D)J data — import_vdj • djvdj</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.4/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.4/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Import V(D)J data — import_vdj"><meta property="og:description" content="Import V(D)J data"><meta property="og:image" content="https://rnabioco.github.io/djvdj/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">djvdj</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Data import and manipulation</h6>
    <a class="dropdown-item" href="../articles/import_vdj.html">Data Import</a>
    <a class="dropdown-item" href="../articles/mutate.html">Data Manipulation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>VDJ analysis</h6>
    <a class="dropdown-item" href="../articles/frequency.html">Clonotype Frequencies</a>
    <a class="dropdown-item" href="../articles/diversity.html">Repertoire Diversity</a>
    <a class="dropdown-item" href="../articles/similarity.html">Repertoire Similarity</a>
    <a class="dropdown-item" href="../articles/gene-usage.html">Gene Usage</a>
    <a class="dropdown-item" href="../articles/clustering.html">CDR3 Clustering</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualization</h6>
    <a class="dropdown-item" href="../articles/plotting.html">Plotting</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/djvdj/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Import V(D)J data</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/djvdj/blob/HEAD/R/import-vdj.R" class="external-link"><code>R/import-vdj.R</code></a></small>
      <div class="d-none name"><code>import_vdj.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Import V(D)J data</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">import_vdj</span><span class="op">(</span></span>
<span>  input <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  vdj_dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  prefix <span class="op">=</span> <span class="st">""</span>,</span>
<span>  data_cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filter_chains <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filter_paired <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  define_clonotypes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_mutations <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  aggr_dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">";"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>input</dt>
<dd><p>Object containing single cell data, if set to NULL a data.frame
containing V(D)J results will be returned</p></dd>


<dt>vdj_dir</dt>
<dd><p>Directory containing the output from cellranger vdj. A vector
or named vector can be given to load data from multiple runs. If a named
vector is given, the cell barcodes will be prefixed with the provided names.
This mimics the behavior of Seurat::Read10X().</p></dd>


<dt>prefix</dt>
<dd><p>Prefix to add to new columns</p></dd>


<dt>data_cols</dt>
<dd><p>Additional columns from filtered_contig_annotations.csv to
include in object.</p></dd>


<dt>filter_chains</dt>
<dd><p>Only include chains with at least one productive and
full length contig.</p></dd>


<dt>filter_paired</dt>
<dd><p>Only include clonotypes with paired chains. For TCR
data each clonotype must have at least one TRA and TRB chain, for BCR data
each clonotype must have at least one IGH chain and at least one IGK or IGL
chain.</p></dd>


<dt>define_clonotypes</dt>
<dd><p>Define clonotype IDs based on V(D)J data. This is
useful if the V(D)J datasets being loaded do not have consistent clonotype
IDs, i.e., clonotype1 is not the same across samples. Possible values are:</p><ul><li><p>'cdr3aa', define clonotypes based on the CDR3 amino acid sequence</p></li>
<li><p>'cdr3nt', define clonotypes based on the CDR3 nucleotide sequence</p></li>
<li><p>'cdr3_gene', define clonotypes based on the combination of the CDR3
nucleotide sequence and the V(D)J genes.</p></li>
</ul><p>When defining clonotypes, only productive full length chains will be used.
Set to NULL (default) to use the clonotype IDs already present in the input
data.</p></dd>


<dt>include_mutations</dt>
<dd><p>Include information about the number of
insertions/deletions/mismatches for each chain. This requires the
concat_ref.bam file from cellranger vdj to be present the directory provided
to vdj_dir. If include_mutations is TRUE, filter_chains is also
automatically set TRUE since indel data is only available for productive
chains.</p></dd>


<dt>aggr_dir</dt>
<dd><p>Path to cellranger aggr output. To include mutation
information for each chain, also provide paths to the original cellranger
vdj output directories using the vdj_dir argument.</p>
<p>To correctly match cell barcodes to those in the object, gene expression
data for each sample must be loaded in the same order as the samples were
specified in the cellranger aggr config file. In addition, if loading
mutation data, sample paths provided to the vdj_dir argument must also be in
the same order as the samples were specified in the cellranger aggr config
file.</p></dd>


<dt>sep</dt>
<dd><p>Separator to use for storing per cell V(D)J data</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Single cell object or data.frame with added V(D)J data</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Loading multiple datasets</span></span></span>
<span class="r-in"><span><span class="va">vdj_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/bcr_1/outs"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/bcr_2/outs"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span><span class="va">tiny_so</span>, <span class="va">vdj_dir</span>, include_mutations <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1     avid_1          0            0               0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters    UMAP_1    UMAP_2 clonotype_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1               0 -5.977054 -2.418108         &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      exact_subclonotype_id chains n_chains cdr3 cdr3_nt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1                    NA   &lt;NA&gt;       NA &lt;NA&gt;    &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      cdr3_length cdr3_nt_length v_gene d_gene j_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1        &lt;NA&gt;           &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      c_gene isotype reads umis productive full_length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1   &lt;NA&gt;    &lt;NA&gt;  &lt;NA&gt; &lt;NA&gt;       &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      paired</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1     NA</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Specifying cell prefixes</span></span></span>
<span class="r-in"><span><span class="co"># if cell prefixes are not specified when loading multiple datasets,</span></span></span>
<span class="r-in"><span><span class="co"># prefixes will be automatically generated in a similar manner as</span></span></span>
<span class="r-in"><span><span class="co"># Seurat::Read10X</span></span></span>
<span class="r-in"><span><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">tiny_so</span>,</span></span>
<span class="r-in"><span>  vdj_dir <span class="op">=</span> <span class="va">vdj_dir</span>,</span></span>
<span class="r-in"><span>  include_mutations <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1     avid_1          0            0               0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters    UMAP_1    UMAP_2 clonotype_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1               0 -5.977054 -2.418108         &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      exact_subclonotype_id chains n_chains cdr3 cdr3_nt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1                    NA   &lt;NA&gt;       NA &lt;NA&gt;    &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      cdr3_length cdr3_nt_length v_gene d_gene j_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1        &lt;NA&gt;           &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      c_gene isotype reads umis productive full_length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1   &lt;NA&gt;    &lt;NA&gt;  &lt;NA&gt; &lt;NA&gt;       &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      paired</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1     NA</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Specifying cell prefixes using vector names</span></span></span>
<span class="r-in"><span><span class="co"># if a named vector is passed, the names will be used as the cell prefixes</span></span></span>
<span class="r-in"><span><span class="va">vdj_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="st">"1"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/bcr_1/outs"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  <span class="st">"2"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/bcr_2/outs"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span><span class="va">tiny_so</span>, <span class="va">vdj_dir</span>, include_mutations <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1     avid_1          0            0               0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters    UMAP_1    UMAP_2 clonotype_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1               0 -5.977054 -2.418108         &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      exact_subclonotype_id chains n_chains cdr3 cdr3_nt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1                    NA   &lt;NA&gt;       NA &lt;NA&gt;    &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      cdr3_length cdr3_nt_length v_gene d_gene j_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1        &lt;NA&gt;           &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      c_gene isotype reads umis productive full_length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1   &lt;NA&gt;    &lt;NA&gt;  &lt;NA&gt; &lt;NA&gt;       &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      paired</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1     NA</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Only include V(D)J data for productive full length chains</span></span></span>
<span class="r-in"><span><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">tiny_so</span>,</span></span>
<span class="r-in"><span>  vdj_dir <span class="op">=</span> <span class="va">vdj_dir</span>,</span></span>
<span class="r-in"><span>  filter_chains <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  include_mutations <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1     avid_1          0            0               0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters    UMAP_1    UMAP_2 clonotype_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1               0 -5.977054 -2.418108         &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      exact_subclonotype_id chains n_chains cdr3 cdr3_nt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1                    NA   &lt;NA&gt;       NA &lt;NA&gt;    &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      cdr3_length cdr3_nt_length v_gene d_gene j_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1        &lt;NA&gt;           &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      c_gene isotype reads umis productive full_length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1   &lt;NA&gt;    &lt;NA&gt;  &lt;NA&gt; &lt;NA&gt;       &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      paired</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1     NA</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Only include V(D)J data for cells with paired chains</span></span></span>
<span class="r-in"><span><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">tiny_so</span>,</span></span>
<span class="r-in"><span>  vdj_dir <span class="op">=</span> <span class="va">vdj_dir</span>,</span></span>
<span class="r-in"><span>  filter_paired <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  include_mutations <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1     avid_1          0            0               0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters    UMAP_1    UMAP_2 clonotype_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1               0 -5.977054 -2.418108         &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      exact_subclonotype_id chains n_chains cdr3 cdr3_nt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1                    NA   &lt;NA&gt;       NA &lt;NA&gt;    &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      cdr3_length cdr3_nt_length v_gene d_gene j_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1        &lt;NA&gt;           &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      c_gene isotype reads umis productive full_length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1   &lt;NA&gt;    &lt;NA&gt;  &lt;NA&gt; &lt;NA&gt;       &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      paired</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1     NA</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Defining clonotypes</span></span></span>
<span class="r-in"><span><span class="co"># this is useful if the original clonotype IDs are not consistent across</span></span></span>
<span class="r-in"><span><span class="co"># datasets, i.e. clonotype1 is not the same for all samples</span></span></span>
<span class="r-in"><span><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">tiny_so</span>,</span></span>
<span class="r-in"><span>  vdj_dir <span class="op">=</span> <span class="va">vdj_dir</span>,</span></span>
<span class="r-in"><span>  define_clonotypes <span class="op">=</span> <span class="st">"cdr3_gene"</span>,</span></span>
<span class="r-in"><span>  include_mutations <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1     avid_1          0            0               0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters    UMAP_1    UMAP_2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1               0 -5.977054 -2.418108</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      exact_subclonotype_id chains n_chains cdr3 cdr3_nt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1                    NA   &lt;NA&gt;       NA &lt;NA&gt;    &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      cdr3_length cdr3_nt_length v_gene d_gene j_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1        &lt;NA&gt;           &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      c_gene isotype reads umis productive full_length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1   &lt;NA&gt;    &lt;NA&gt;  &lt;NA&gt; &lt;NA&gt;       &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      paired clonotype_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1     NA         &lt;NA&gt;</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Include mutation information for each chain</span></span></span>
<span class="r-in"><span><span class="co"># this information will be included if the file concat_ref.bam is present</span></span></span>
<span class="r-in"><span><span class="co"># including mutation information will cause data import to be slower</span></span></span>
<span class="r-in"><span><span class="va">vdj_so</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">tiny_so</span>,</span></span>
<span class="r-in"><span>  vdj_dir <span class="op">=</span> <span class="va">vdj_dir</span>,</span></span>
<span class="r-in"><span>  include_mutations <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>airr_rearrangement.tsv not found in /home/runner/work/_temp/Library/djvdj/extdata/bcr_1/outs.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>airr_rearrangement.tsv not found in /home/runner/work/_temp/Library/djvdj/extdata/bcr_2/outs.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>To add mutation data to object, concat_ref.bam and airr_rearrangement.tsv must be present for all samples, check that these files are in the provided directory paths, mutation data not added to object.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_so</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1     avid_1          0            0               0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters    UMAP_1    UMAP_2 clonotype_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1               0 -5.977054 -2.418108         &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      exact_subclonotype_id chains n_chains cdr3 cdr3_nt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1                    NA   &lt;NA&gt;       NA &lt;NA&gt;    &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      cdr3_length cdr3_nt_length v_gene d_gene j_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1        &lt;NA&gt;           &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      c_gene isotype reads umis productive full_length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1   &lt;NA&gt;    &lt;NA&gt;  &lt;NA&gt; &lt;NA&gt;       &lt;NA&gt;        &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      paired</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1     NA</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Using import_vdj outside of Seurat</span></span></span>
<span class="r-in"><span><span class="co"># SingleCellExperiment objects are also compatible, or if an input object is</span></span></span>
<span class="r-in"><span><span class="co"># omitted, a data.frame containing the V(D)J data will be returned</span></span></span>
<span class="r-in"><span><span class="va">vdj_sce</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span><span class="va">tiny_sce</span>, <span class="va">vdj_dir</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_sce</span><span class="op">@</span><span class="va">colData</span>, <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 1 row and 25 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      &lt;character&gt;  &lt;numeric&gt;    &lt;integer&gt;        &lt;factor&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1      avid_1          0            0               0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      seurat_clusters    UMAP_1    UMAP_2 clonotype_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                             &lt;factor&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1               0  -5.97705  -2.41811           NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      exact_subclonotype_id      chains  n_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                  &lt;numeric&gt; &lt;character&gt; &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1                    NA          NA        NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                             cdr3     cdr3_nt cdr3_length cdr3_nt_length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      &lt;character&gt; &lt;character&gt; &lt;character&gt;    &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1          NA          NA          NA             NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           v_gene      d_gene      j_gene      c_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1          NA          NA          NA          NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                          isotype       reads        umis  productive</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1          NA          NA          NA          NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      full_length    paired</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      &lt;character&gt; &lt;logical&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1_AAGCCGCAGCTTATCG-1          NA        NA</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">vdj_df</span> <span class="op">&lt;-</span> <span class="fu">import_vdj</span><span class="op">(</span>vdj_dir <span class="op">=</span> <span class="va">vdj_dir</span>, include_mutations <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">vdj_df</span>, <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     clonotype_id exact_subclonotype_id chains n_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1ACGGAGACATGCTGGC-1   clonotype1                     1    IGK        1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                            cdr3                           cdr3_nt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1ACGGAGACATGCTGGC-1 CQQSNSWPYTF TGTCAACAGAGTAACAGCTGGCCGTACACGTTC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     cdr3_length cdr3_nt_length   v_gene d_gene j_gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1ACGGAGACATGCTGGC-1          11             33 IGKV5-43   None  IGKJ2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     c_gene isotype reads umis productive full_length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1ACGGAGACATGCTGGC-1   IGKC    None   394   24       TRUE        TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     paired</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1ACGGAGACATGCTGGC-1  FALSE</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

