<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Modify V(D)J data in object — mutate_vdj • djvdj</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Modify V(D)J data in object — mutate_vdj"><meta property="og:description" content="Modify per-chain V(D)J data for each cell. This function offers greater
flexibility than summarize_vdj(), but is less user-friendly."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">djvdj</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/avid.html">V(D)J analysis</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/rnabioco/djvdj/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Modify V(D)J data in object</h1>
    <small class="dont-index">Source: <a href="https://github.com/rnabioco/djvdj/blob/HEAD/R/mutate-vdj.R" class="external-link"><code>R/mutate-vdj.R</code></a></small>
    <div class="hidden name"><code>mutate_vdj.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Modify per-chain V(D)J data for each cell. This function offers greater
flexibility than <code><a href="summarize_vdj.html">summarize_vdj()</a></code>, but is less user-friendly.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">mutate_vdj</span><span class="op">(</span>
  <span class="va">input</span>,
  <span class="va">...</span>,
  clonotype_col <span class="op">=</span> <span class="st">"clonotype_id"</span>,
  vdj_cols <span class="op">=</span> <span class="cn">NULL</span>,
  return_df <span class="op">=</span> <span class="cn">FALSE</span>,
  sep <span class="op">=</span> <span class="st">";"</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>input</dt>
<dd><p>Single cell object or data.frame containing V(D)J data. If a
data.frame is provided, cell barcodes should be stored as row names.</p></dd>
<dt>...</dt>
<dd><p>Name-value pairs to use for creating or modifying per-chain V(D)J
meta.data, e.g. mean_umis = mean(umis).</p>
<p>To allow for modification of per-chain V(D)J data, the data for each cell is
converted into a vector, e.g. 'IGH;IGK' is equivalent to c('IGH', 'IGK').
This allows R vector operations to be performed on the per-chain values.
Any operations that produce a result greater than length 1 need to be
returned as a list(), e.g. new_col = umis + 1 will return a new value for
each chain, to prevent an error this must be written as
new_col = list(umis + 1).</p></dd>
<dt>clonotype_col</dt>
<dd><p>meta.data column containing clonotype IDs. This is used
to identify columns containing V(D)J data.</p></dd>
<dt>vdj_cols</dt>
<dd><p>meta.data columns containing V(D)J data to modify. If NULL,
data are automatically selected by identifying columns that have NAs in the
same rows as clonotype_col.</p></dd>
<dt>return_df</dt>
<dd><p>Return results as a data.frame. If FALSE, results will be
added to the input object.</p></dd>
<dt>sep</dt>
<dd><p>Separator used for storing per cell V(D)J data</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Object with modified meta.data</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co"># Calculate mean reads and UMIs per cell</span></span>
<span class="r-in"><span class="fu">mutate_vdj</span><span class="op">(</span></span>
<span class="r-in">  <span class="va">vdj_sce</span>,</span>
<span class="r-in">  mean_umis  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">umis</span><span class="op">)</span>,</span>
<span class="r-in">  mean_reads <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">reads</span><span class="op">)</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SingleCellExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 200 204 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(200): Gins2 Klf11 ... Gm29265 Pkd1l1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(204): 1_AAACCTGAGTAGGTGC-1 1_AAACCTGAGTCAAGCG-1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2_TTTACTGGTTGGTTTG-1 2_TTTGTCATCACCCGAG-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(30): orig.ident nCount_RNA ... mean_umis mean_reads</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> reducedDimNames(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> mainExpName: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> altExpNames(0):</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Calculate the total number of insertions + deletions for each chain</span></span>
<span class="r-in"><span class="co"># we have to wrap our expression in list() since a value is returned for</span></span>
<span class="r-in"><span class="co"># each chain</span></span>
<span class="r-in"><span class="fu">mutate_vdj</span><span class="op">(</span></span>
<span class="r-in">  <span class="va">vdj_sce</span>,</span>
<span class="r-in">  indels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">n_insertion</span> <span class="op">+</span> <span class="va">n_deletion</span><span class="op">)</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SingleCellExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 200 204 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(200): Gins2 Klf11 ... Gm29265 Pkd1l1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(204): 1_AAACCTGAGTAGGTGC-1 1_AAACCTGAGTCAAGCG-1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2_TTTACTGGTTGGTTTG-1 2_TTTGTCATCACCCGAG-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(29): orig.ident nCount_RNA ... n_chains indels</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> reducedDimNames(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> mainExpName: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> altExpNames(0):</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Create a new column showing the unique chains for each cell</span></span>
<span class="r-in"><span class="fu">mutate_vdj</span><span class="op">(</span></span>
<span class="r-in">  <span class="va">vdj_sce</span>,</span>
<span class="r-in">  unique_chains <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_c.html" class="external-link">str_c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">chains</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SingleCellExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 200 204 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(200): Gins2 Klf11 ... Gm29265 Pkd1l1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(204): 1_AAACCTGAGTAGGTGC-1 1_AAACCTGAGTCAAGCG-1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2_TTTACTGGTTGGTTTG-1 2_TTTGTCATCACCCGAG-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(29): orig.ident nCount_RNA ... n_chains unique_chains</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> reducedDimNames(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> mainExpName: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> altExpNames(0):</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Determine which cells have both an IGK and IGL chain</span></span>
<span class="r-in"><span class="fu">mutate_vdj</span><span class="op">(</span></span>
<span class="r-in">  <span class="va">vdj_sce</span>,</span>
<span class="r-in">  both_light <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IGK"</span>, <span class="st">"IGL"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">chains</span><span class="op">)</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SingleCellExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 200 204 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): counts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(200): Gins2 Klf11 ... Gm29265 Pkd1l1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(204): 1_AAACCTGAGTAGGTGC-1 1_AAACCTGAGTCAAGCG-1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2_TTTACTGGTTGGTTTG-1 2_TTTGTCATCACCCGAG-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(29): orig.ident nCount_RNA ... n_chains both_light</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> reducedDimNames(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> mainExpName: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> altExpNames(0):</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Determine which cells have multiple light chains</span></span>
<span class="r-in"><span class="fu">mutate_vdj</span><span class="op">(</span></span>
<span class="r-in">  <span class="va">vdj_so</span>,</span>
<span class="r-in">  multi_light <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">chains</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IGK"</span>, <span class="st">"IGL"</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class Seurat </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 200 features across 204 samples within 1 assay </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Active assay: RNA (200 features, 67 variable features)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  2 dimensional reductions calculated: pca, umap</span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Sheridan, <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer></div>

  


  

  </body></html>

