<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Visualize numerical single-cell data by creating a histogram, density plot,
violin plots, or boxplots. When plotting V(D)J data, values can be plotted
separately for each chain or summarized and plotted for each cell."><title>Plot numerical single-cell data — plot_numerical • djvdj</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.5/font.css" rel="stylesheet"><link href="../deps/Source_Code_Pro-0.4.5/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Plot numerical single-cell data — plot_numerical"><meta property="og:description" content="Visualize numerical single-cell data by creating a histogram, density plot,
violin plots, or boxplots. When plotting V(D)J data, values can be plotted
separately for each chain or summarized and plotted for each cell."><meta property="og:image" content="https://rnabioco.github.io/djvdj/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">djvdj</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Data import and manipulation</h6>
    <a class="dropdown-item" href="../articles/import.html">Data Import</a>
    <a class="dropdown-item" href="../articles/mutate.html">Data Manipulation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>VDJ analysis</h6>
    <a class="dropdown-item" href="../articles/frequency.html">Clonotype Frequencies</a>
    <a class="dropdown-item" href="../articles/diversity.html">Repertoire Diversity</a>
    <a class="dropdown-item" href="../articles/similarity.html">Repertoire Similarity</a>
    <a class="dropdown-item" href="../articles/gene-usage.html">Gene Usage</a>
    <a class="dropdown-item" href="../articles/clustering.html">CDR3 Clustering</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualization</h6>
    <a class="dropdown-item" href="../articles/plotting.html">Plotting</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/djvdj/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Plot numerical single-cell data</h1>
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/djvdj/blob/HEAD/R/plot-vdj.R" class="external-link"><code>R/plot-vdj.R</code></a></small>
      <div class="d-none name"><code>plot_numerical.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Visualize numerical single-cell data by creating a histogram, density plot,
violin plots, or boxplots. When plotting V(D)J data, values can be plotted
separately for each chain or summarized and plotted for each cell.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_histogram</span><span class="op">(</span></span>
<span>  <span class="va">input</span>,</span>
<span>  <span class="va">data_col</span>,</span>
<span>  cluster_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"histogram"</span>,</span>
<span>  top <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  other_label <span class="op">=</span> <span class="st">"other"</span>,</span>
<span>  units <span class="op">=</span> <span class="st">"frequency"</span>,</span>
<span>  plot_colors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_lvls <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  trans <span class="op">=</span> <span class="st">"identity"</span>,</span>
<span>  panel_nrow <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  panel_scales <span class="op">=</span> <span class="st">"fixed"</span>,</span>
<span>  na_color <span class="op">=</span> <span class="st">"grey80"</span>,</span>
<span>  n_label <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  label_params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  per_chain <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  chain <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chain_col <span class="op">=</span> <span class="va">global</span><span class="op">$</span><span class="va">chain_col</span>,</span>
<span>  summary_fn <span class="op">=</span> <span class="va">mean</span>,</span>
<span>  sep <span class="op">=</span> <span class="va">global</span><span class="op">$</span><span class="va">sep</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot_violin</span><span class="op">(</span></span>
<span>  <span class="va">input</span>,</span>
<span>  <span class="va">data_col</span>,</span>
<span>  cluster_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"violin"</span>,</span>
<span>  top <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  other_label <span class="op">=</span> <span class="st">"other"</span>,</span>
<span>  plot_colors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_lvls <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  trans <span class="op">=</span> <span class="st">"identity"</span>,</span>
<span>  panel_nrow <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  panel_scales <span class="op">=</span> <span class="st">"free_x"</span>,</span>
<span>  na_color <span class="op">=</span> <span class="st">"grey80"</span>,</span>
<span>  n_label <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  label_params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  per_chain <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  chain <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chain_col <span class="op">=</span> <span class="va">global</span><span class="op">$</span><span class="va">chain_col</span>,</span>
<span>  summary_fn <span class="op">=</span> <span class="va">mean</span>,</span>
<span>  sep <span class="op">=</span> <span class="va">global</span><span class="op">$</span><span class="va">sep</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>input</dt>
<dd><p>Single cell object or data.frame, if a
data.frame is provided, cell barcodes should be stored as row names.</p></dd>


<dt>data_col</dt>
<dd><p>meta.data column containing data to plot</p></dd>


<dt>cluster_col</dt>
<dd><p>meta.data column containing cluster IDs to use for
grouping cells for plotting</p></dd>


<dt>group_col</dt>
<dd><p>meta.data column to use for grouping clusters into separate
panels</p></dd>


<dt>method</dt>
<dd><p>Method to use for plotting, possible values include,
'histogram', 'density', 'boxplot', or 'violin'</p></dd>


<dt>top</dt>
<dd><p>To only show the top cell groups, provide
one of the following, all other cells will be labeled using the value
provided to the <code>other_label</code> argument. If <code>NULL</code> this will be automatically
set.</p><ul><li><p>Integer specifying the number of top groups to show</p></li>
<li><p>Vector specifying the names of cell groups to show</p></li>
</ul></dd>


<dt>other_label</dt>
<dd><p>Label to use for 'other' cells when <code>top</code> is specified, if
<code>NULL</code> all cell groups will be shown.</p></dd>


<dt>units</dt>
<dd><p>Units to use for y-axis when method is 'histogram'. Use
'frequency' to show the number of values or 'percent' to show the percentage
of total values.</p></dd>


<dt>plot_colors</dt>
<dd><p>Character vector specifying colors to use for cell
clusters specified by cluster_col. When cluster_col is <code>NULL</code>, plot colors
can be directly modified with the ggplot2 parameters <code>color</code> and <code>fill</code>,
e.g. <code>fill = "red"</code>, <code>color = "black"</code></p></dd>


<dt>plot_lvls</dt>
<dd><p>Character vector containing order to use for plotting cell
clusters specified by cluster_col</p></dd>


<dt>trans</dt>
<dd><p>Transformation to use when plotting data, e.g. 'log10'. By
default values are not transformed, refer to <code><a href="https://ggplot2.tidyverse.org/reference/continuous_scale.html" class="external-link">ggplot2::continuous_scale()</a></code>
for more options.</p></dd>


<dt>panel_nrow</dt>
<dd><p>The number of rows to use for arranging plot panels</p></dd>


<dt>panel_scales</dt>
<dd><p>Should scales for plot panels be fixed or free. This
passes a scales specification to <code><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">ggplot2::facet_wrap()</a></code>, can be 'fixed',
'free', 'free_x', or 'free_y'. 'fixed' will cause panels to share the same
scales.</p></dd>


<dt>na_color</dt>
<dd><p>Color to use for missing values. If plotting V(D)J data,
cells lacking data will be plotted as <code>NA</code>s.</p></dd>


<dt>n_label</dt>
<dd><p>Location on plot where n label should be added, this can be
any combination of the following:</p><ul><li><p>'corner', display the total number of cells plotted in the top right
corner, the position of the label can be modified by passing <code>x</code> and <code>y</code>
specifications with the <code>label_params</code> argument</p></li>
<li><p>'axis', display the number of cells plotted for each group shown on the
x-axis</p></li>
<li><p>'legend', display the number of cells plotted for each group shown in the
plot legend</p></li>
<li><p>'none', do not display the number of cells plotted</p></li>
</ul></dd>


<dt>label_params</dt>
<dd><p>Named list providing additional parameters to modify
n label aesthetics, e.g. list(size = 4, color = "red")</p></dd>


<dt>...</dt>
<dd><p>Additional arguments to pass to ggplot2, e.g. color, fill, size,
linetype, etc.</p></dd>


<dt>per_chain</dt>
<dd><p>If <code>TRUE</code> values will be plotted for each chain, i.e. each
data point represents a chain. If <code>FALSE</code> values will be summarized for each
cell using <code>summary_fn</code> before plotting, i.e. each data point represents a
cell.</p></dd>


<dt>chain</dt>
<dd><p>Chain(s) to use for filtering data before plotting. If NULL
data will not be filtered based on chain.</p></dd>


<dt>chain_col</dt>
<dd><p>meta.data column containing chains for each cell</p></dd>


<dt>summary_fn</dt>
<dd><p>Function to use for summarizing values when <code>per_chain</code> is
<code>FALSE</code>, can be either a function, e.g. <code>mean</code>, or a purrr-style
lambda, e.g. <code>~ mean(.x, na.rm = TRUE)</code> where <code>.x</code> refers to the column. If
<code>NULL</code>, the mean will be calculated.</p></dd>


<dt>sep</dt>
<dd><p>Separator used for storing per-chain V(D)J data for each cell</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>ggplot object</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="summarize_vdj.html">summarize_vdj()</a></code> for more examples on how per-chain data can be
summarized for each cell</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Create histogram</span></span></span>
<span class="r-in"><span><span class="fu">plot_histogram</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">vdj_so</span>,</span></span>
<span class="r-in"><span>  data_col    <span class="op">=</span> <span class="st">"reads"</span>,</span></span>
<span class="r-in"><span>  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,</span></span>
<span class="r-in"><span>  trans       <span class="op">=</span> <span class="st">"log10"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 133 rows containing missing values</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span>
<span class="r-plt img"><img src="plot_numerical-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create violin plots</span></span></span>
<span class="r-in"><span><span class="fu">plot_violin</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">vdj_sce</span>,</span></span>
<span class="r-in"><span>  data_col    <span class="op">=</span> <span class="st">"nFeature_RNA"</span>,</span></span>
<span class="r-in"><span>  cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plot_numerical-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

