<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="djvdj">
<title>Clonal Abundance • djvdj</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.2/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.2/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Clonal Abundance">
<meta property="og:description" content="djvdj">
<meta property="og:image" content="https://rnabioco.github.io/djvdj/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">djvdj</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Data import and manipulation</h6>
    <a class="dropdown-item" href="../articles/import_vdj.html">Data Import</a>
    <a class="dropdown-item" href="../articles/mutate.html">Data Manipulation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>VDJ analysis</h6>
    <a class="dropdown-item" href="../articles/gene-usage.html">Gene Usage</a>
    <a class="dropdown-item" href="../articles/abundance.html">Clonal Abundance</a>
    <a class="dropdown-item" href="../articles/diversity.html">Repertoire Diversity</a>
    <a class="dropdown-item" href="../articles/similarity.html">Repertoire Similarity</a>
    <a class="dropdown-item" href="../articles/clustering.html">CDR3 Clustering</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualization</h6>
    <a class="dropdown-item" href="../articles/plotting.html">Plotting</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/djvdj/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Clonal Abundance</h1>
            
            <h4 data-toc-skip class="date">October 31 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/djvdj/blob/HEAD/vignettes/abundance.Rmd" class="external-link"><code>vignettes/abundance.Rmd</code></a></small>
      <div class="d-none name"><code>abundance.Rmd</code></div>
    </div>

    
    
<p><br></p>
<p>This vignette provides detailed examples for quantifying differences
in clonal abundance. For the examples shown below, we use data for
splenocytes from BL6 and MD4 mice collected using the 10X Genomics
scRNA-seq platform. MD4 B cells are monoclonal and specifically bind hen
egg lysozyme. A Seurat object along with the raw matrices can be
downloaded <a href="https://djvdj-data.s3.us-west-2.amazonaws.com/splenocytes.zip" class="external-link">here</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/djvdj/">djvdj</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"https://djvdj-data.s3.us-west-2.amazonaws.com/splenocytes.zip"</span>,</span>
<span>  <span class="st">"splenocytes.zip"</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span><span class="st">"splenocytes.zip"</span>, overwrite <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load Seurat object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"splenocytes/splen_so.rda"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add V(D)J data to object</span></span>
<span><span class="va">vdj_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  BL6 <span class="op">=</span> <span class="st">"splenocytes/BL6_BCR"</span>,</span>
<span>  MD4 <span class="op">=</span> <span class="st">"splenocytes/MD4_BCR"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">so</span> <span class="op">&lt;-</span> <span class="va">splen_so</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span><span class="va">vdj_dirs</span>, define_clonotypes <span class="op">=</span> <span class="st">"cdr3_gene"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<div class="section level2">
<h2 id="calculating-clonal-abundance">Calculating clonal abundance<a class="anchor" aria-label="anchor" href="#calculating-clonal-abundance"></a>
</h2>
<p>To quantify clonotype abundance and store the results in the object
meta.data, the <code><a href="../reference/calc_frequency.html">calc_frequency()</a></code> function can be used. This
will add columns showing the number of occurrences of each clonotype
(‘freq’), the percentage of cells sharing the clonotype (‘pct’), and a
label that can be used for plotting (‘grp’). By default these
calculations will be performed for all cells in the object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/calc_frequency.html">calc_frequency</a></span><span class="op">(</span>data_col <span class="op">=</span> <span class="st">"clonotype_id"</span><span class="op">)</span></span></code></pre></div>
<p>To calculate clonotype abundance separately for samples or clusters,
the <code>cluster_col</code> argument can be used. To do this just
specify the name of the column containing the sample or cluster IDs for
each cell.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/calc_frequency.html">calc_frequency</a></span><span class="op">(</span></span>
<span>    data_col <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>When <code>cluster_col</code> is specified, an additional meta.data
column (‘shared’) will be added indicating whether the clonotype is
shared between multiple clusters.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span><span class="op">@</span><span class="va">meta.data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt;                        orig.ident nCount_RNA nFeature_RNA RNA_snn_res.1</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGACCTTTG-1        BL6        273          208             0</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGGAATCGC-1        BL6        243          199            12</span></span>
<span><span class="co">#&gt;                        seurat_clusters cell_type sample</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGACCTTTG-1               0   B cells  BL6-1</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGGAATCGC-1              12   B cells  BL6-1</span></span>
<span><span class="co">#&gt;                        exact_subclonotype_id  chains n_chains</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGACCTTTG-1                     1 IGH;IGK        2</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGGAATCGC-1                     1     IGL        1</span></span>
<span><span class="co">#&gt;                                          cdr3</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGACCTTTG-1 CARHPAMDYW;CWQGTHFPRTF</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGGAATCGC-1            CALWYSTHYVF</span></span>
<span><span class="co">#&gt;                                                                                 cdr3_nt</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGACCTTTG-1 TGTGCCAGACATCCCGCTATGGACTACTGG;TGCTGGCAAGGTACACATTTTCCTCGGACGTTC</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGGAATCGC-1                                TGTGCTCTATGGTACAGCACCCATTATGTTTTC</span></span>
<span><span class="co">#&gt;                        cdr3_length cdr3_nt_length            v_gene</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGACCTTTG-1       10;11          30;33 IGHV2-6;IGKV1-135</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGGAATCGC-1          11             33             IGLV2</span></span>
<span><span class="co">#&gt;                           d_gene      j_gene    c_gene isotype  reads</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGACCTTTG-1 None;None IGHJ4;IGKJ1 IGHM;IGKC    IGHM 60;376</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGGAATCGC-1      None       IGLJ2     IGLC2    None    360</span></span>
<span><span class="co">#&gt;                        umis productive full_length paired  clonotype_id</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGACCTTTG-1 6;22  TRUE;TRUE   TRUE;TRUE   TRUE clonotype5467</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGGAATCGC-1   13       TRUE        TRUE  FALSE clonotype5730</span></span>
<span><span class="co">#&gt;                        clonotype_id_freq clonotype_id_pct</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGACCTTTG-1                 1        0.1620746</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGGAATCGC-1                 5        0.8103728</span></span>
<span><span class="co">#&gt;                        clonotype_id_shared clonotype_id_grp</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGACCTTTG-1               FALSE                1</span></span>
<span><span class="co">#&gt; BL6_AAACCTGAGGAATCGC-1                TRUE              2-6</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="plotting-clonal-abundance">Plotting clonal abundance<a class="anchor" aria-label="anchor" href="#plotting-clonal-abundance"></a>
</h2>
<p>djvdj includes the <code><a href="../reference/plot_clonal_abundance.html">plot_clonal_abundance()</a></code> function to
visualize differences in clonotype frequency between samples or
clusters. By default this will produce bargraphs. Plot colors can be
adjusted using the <code>plot_colors</code> argument.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/plot_clonal_abundance.html">plot_clonal_abundance</a></span><span class="op">(</span></span>
<span>    clonotype_col <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="st">"#3182bd"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="abundance_files/figure-html/unnamed-chunk-6-1.png" width="768"></p>
<p>Abundance values can be calculated and plotted separately for each
sample or cluster using the <code>cluster_col</code> argument. The
<code>panel_nrow</code> and <code>panel_scales</code> arguments can be
used to add separate scales for each sample or to adjust the number of
rows used to arrange plots.</p>
<p>As expected we see that most MD4 B cells share the same clonotype,
while BL6 cells have a diverse repertoire.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/plot_clonal_abundance.html">plot_clonal_abundance</a></span><span class="op">(</span></span>
<span>    clonotype_col <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    panel_scales  <span class="op">=</span> <span class="st">"free"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="abundance_files/figure-html/unnamed-chunk-7-1.png" width="768"></p>
<p>Rank-abundance plots can also be generated by setting the
<code>method</code> argument to ‘line’. Most djvdj plotting functions
return ggplot objects that can be further modified with ggplot2
functions. Here we log10-transform the y-axis using the
<code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">ggplot2::scale_y_log10()</a></code> function.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/plot_clonal_abundance.html">plot_clonal_abundance</a></span><span class="op">(</span></span>
<span>    clonotype_col <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    method        <span class="op">=</span> <span class="st">"line"</span>,</span>
<span>    plot_colors   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>MD4 <span class="op">=</span> <span class="st">"#fec44f"</span>, BL6 <span class="op">=</span> <span class="st">"#3182bd"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="abundance_files/figure-html/unnamed-chunk-8-1.png" width="768"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="umap-projections">UMAP projections<a class="anchor" aria-label="anchor" href="#umap-projections"></a>
</h2>
<p>By default <code><a href="../reference/calc_frequency.html">calc_frequency()</a></code> will divide clonotypes into
groups based on abundance and add a column to the meta.data containing
these group labels. Clonotype abundance can be summarized on a UMAP
projection by plotting the added ‘grp’ column using the generic plotting
function <code><a href="../reference/plot_features.html">plot_features()</a></code></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create UMAP summarizing samples</span></span>
<span><span class="va">mouse_gg</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/plot_features.html">plot_features</a></span><span class="op">(</span></span>
<span>    feature <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.25</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Create UMAP summarizing clonotype abundance</span></span>
<span><span class="va">abun_gg</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/calc_frequency.html">calc_frequency</a></span><span class="op">(</span></span>
<span>    data_col <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/plot_features.html">plot_features</a></span><span class="op">(</span></span>
<span>    feature <span class="op">=</span> <span class="st">"clonotype_id_grp"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.25</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">mouse_gg</span> <span class="op">+</span> <span class="va">abun_gg</span></span></code></pre></div>
<p><img src="abundance_files/figure-html/unnamed-chunk-9-1.png" width="1056"></p>
<p>Highly abundant clonotypes can also be specifically labeled on a UMAP
projection. To do this, add a new meta.data column with the desired
label using the <code><a href="../reference/mutate_vdj.html">mutate_vdj()</a></code> function. This function works
in a similar manner as <code><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">dplyr::mutate()</a></code>, but will
specifically modify the object meta.data and allows to the user to parse
per-chain information for each cell.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_gg</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/mutate_vdj.html">mutate_vdj</a></span><span class="op">(</span></span>
<span>    top_clonotype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">clonotype_id</span> <span class="op">==</span> <span class="st">"clonotype907"</span>, <span class="va">clonotype_id</span>, <span class="st">"other"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/plot_features.html">plot_features</a></span><span class="op">(</span></span>
<span>    feature     <span class="op">=</span> <span class="st">"top_clonotype"</span>,</span>
<span>    size        <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>other <span class="op">=</span> <span class="st">"#fec44f"</span>, clonotype907 <span class="op">=</span> <span class="st">"#3182bd"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">mouse_gg</span> <span class="op">+</span> <span class="va">top_gg</span></span></code></pre></div>
<p><img src="abundance_files/figure-html/unnamed-chunk-10-1.png" width="1056"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="other-frequency-calculations">Other frequency calculations<a class="anchor" aria-label="anchor" href="#other-frequency-calculations"></a>
</h2>
<p>In addition to clonotype abundance, <code><a href="../reference/calc_frequency.html">calc_frequency()</a></code> can
be used to summarize the frequency of any cell label present in the
object. In this example we count the number of cells present for each
cell type in each sample.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/calc_frequency.html">calc_frequency</a></span><span class="op">(</span></span>
<span>    data_col <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>To plot the fraction of cells present for each cell type, we can use
the generic plotting function, <code><a href="../reference/plot_frequency.html">plot_frequency()</a></code>. This will
create stacked bargraphs summarizing each cell label present in the
<code>data_col</code> column. The color of each group can be specified
with the <code>plot_colors</code> argument.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/plot_frequency.html">plot_frequency</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#3182bd"</span>, <span class="st">"#fec44f"</span>, <span class="st">"#31a354"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="abundance_files/figure-html/unnamed-chunk-12-1.png" width="768"></p>
<p>To summarize the number cells present for each cell type, set the
<code>units</code> argument to ‘frequency’.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/plot_frequency.html">plot_frequency</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    units       <span class="op">=</span> <span class="st">"frequency"</span>,</span>
<span>    stack       <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="abundance_files/figure-html/unnamed-chunk-13-1.png" width="768"></p>
<p>Frequency plots can also be separated based on an additional grouping
variable such as treatment group (e.g. placebo vs drug) or disease
status (e.g. healthy vs disease). This will generate boxplots with each
point representing a label present in the <code>cluster_col</code>
column. In this example we have 5 BL6 and 5 MD4 samples, so there are 5
points shown for each boxplot.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/plot_frequency.html">plot_frequency</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    group_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>MD4 <span class="op">=</span> <span class="st">"#fec44f"</span>, BL6 <span class="op">=</span> <span class="st">"#3182bd"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="abundance_files/figure-html/unnamed-chunk-14-1.png" width="768"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
