<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="djvdj">
<title>V(D)J analysis • djvdj</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="../deps/_Source%20Sans%20Pro-0.4.1/font.css" rel="stylesheet">
<link href="../deps/_Source%20Code%20Pro-0.4.1/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="V(D)J analysis">
<meta property="og:description" content="djvdj">
<meta property="og:image" content="https://rnabioco.github.io/djvdj/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">djvdj</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/avid.html">V(D)J analysis</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/djvdj/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>V(D)J analysis</h1>
            
            <h4 data-toc-skip class="date">May 27 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/djvdj/blob/HEAD/vignettes/avid.Rmd" class="external-link"><code>vignettes/avid.Rmd</code></a></small>
      <div class="d-none name"><code>avid.Rmd</code></div>
    </div>

    
    
<p><br></p>
<p>For this vignette we are using AVID-seq data. This is a method developed by the Hesselberth lab which involves staining cells with DNA-tagged antigen. The DNA tag is similar to the tags present on CITE-seq antibodies and can be detected using the 10x Genomics 5’ immune profiling kit. For this experiment we mixed splenocytes from BL6 and MD4 mice and stained with a HEL-DNA conjugate. MD4 B cells are monoclonal and specifically bind HEL.</p>
<p><img src="avid_files/figure-html/rna_umap-1.png" width="1200"></p>
<p><br></p>
<div class="section level2">
<h2 id="import-vdj-data">Import V(D)J data<a class="anchor" aria-label="anchor" href="#import-vdj-data"></a>
</h2>
<p><code><a href="../reference/import_vdj.html">import_vdj()</a></code> takes the output files from <a href="https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/using/vdj#header" class="external-link">Cell Ranger</a> and adds clonotype information to the meta.data for an existing <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a> or SingleCellExperiment object. For cells with multiple chains, the information for each chain is stored as a single row, separated by a semi-colon. For cells that do not have V(D)J sequencing data, <code>NA</code>s are added to the meta.data.</p>
<p>If the Seurat object contains data for multiple runs, a vector containing paths to the <code>cellranger vdj</code> output files for each sample can be given. When multiple runs are included in the same object, the cell barcodes will commonly contain a unique prefix for each sample. The cell prefixes can be specified using the <code>cell_prefix</code> argument.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/djvdj">djvdj</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org" class="external-link">purrr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span>

<span class="co"># Create vector of paths pointing to cellranger output</span>
<span class="va">paths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data/avid/bcr/outs"</span><span class="op">)</span>

<span class="va">so_avid</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span>
  input         <span class="op">=</span> <span class="va">so_avid</span>,                  <span class="co"># Seurat object</span>
  vdj_dir       <span class="op">=</span> <span class="va">paths</span>,                    <span class="co"># cellranger directories</span>
  filter_chains <span class="op">=</span> <span class="cn">TRUE</span>,                     <span class="co"># Only include productive chains</span>
  filter_paired <span class="op">=</span> <span class="cn">FALSE</span>                     <span class="co"># Only include clonotypes with paired chains</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
<p>Take a look at the meta.data to see the V(D)J data added to the object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">so_avid</span><span class="op">@</span><span class="va">meta.data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fl">19</span><span class="op">:</span><span class="fl">35</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 7,137 × 17</span></span>
<span class="co">#&gt;    clonotype_id chains  n_chains cdr3   cdr3_nt cdr3_length cdr3_nt_length</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> clonotype91  IGH;IGK        2 CVKGY… TGTGTG… 14;10       42;30         </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span>            <span style="color: #BB0000;">NA</span> <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>            </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> clonotype92  IGH;IGK        2 CARGR… TGTGCA… 13;11       39;33         </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span>            <span style="color: #BB0000;">NA</span> <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>            </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span>            <span style="color: #BB0000;">NA</span> <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>            </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span>            <span style="color: #BB0000;">NA</span> <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>            </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> <span style="color: #BB0000;">NA</span>           <span style="color: #BB0000;">NA</span>            <span style="color: #BB0000;">NA</span> <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>      <span style="color: #BB0000;">NA</span>          <span style="color: #BB0000;">NA</span>            </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> clonotype94  IGH;IGK        2 CTVSY… TGTACC… 14;11       42;33         </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> clonotype95  IGH;IGK        2 CARSY… TGTGCA… 17;11       51;33         </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> clonotype96  IGH;IGK        2 CARSR… TGTGCA… 9;11        27;33         </span>
<span class="co">#&gt; <span style="color: #949494;"># … with 7,127 more rows, and 10 more variables: v_gene &lt;chr&gt;,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   d_gene &lt;chr&gt;, j_gene &lt;chr&gt;, c_gene &lt;chr&gt;, isotype &lt;chr&gt;, reads &lt;chr&gt;,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   umis &lt;chr&gt;, productive &lt;chr&gt;, full_length &lt;chr&gt;, paired &lt;lgl&gt;</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="clonotype-abundance">Clonotype Abundance<a class="anchor" aria-label="anchor" href="#clonotype-abundance"></a>
</h2>
<p>To identify the top clonotypes in each sample or cluster, clonotype abundance can be calculated using the <code>calc_abundance()</code> function. These calculations can be performed on a per-cluster or per-sample basis by also providing a meta.data column containing cell labels.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">so_avid</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_frequency.html">calc_frequency</a></span><span class="op">(</span>
  input       <span class="op">=</span> <span class="va">so_avid</span>,
  cluster_col <span class="op">=</span> <span class="st">"mouse"</span>,                  <span class="co"># Column containing cell clusters to compare</span>
  data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>            <span class="co"># Column containing clonotype IDs to use</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
<p>For each ‘calc’ function provided by djvdj, there is a matching ‘plot’ function that will generate a summary plot. The <code>plot_abundance()</code> function will plot clonotypes ranked by abundance. As expected we see that most MD4 B cells share the same clonotype.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clrs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>
  BL6 <span class="op">=</span> <span class="st">"#E69F00"</span>,
  MD4 <span class="op">=</span> <span class="st">"#56B4E9"</span>
<span class="op">)</span>

<span class="fu"><a href="../reference/plot_clonal_abundance.html">plot_clonal_abundance</a></span><span class="op">(</span>
  input       <span class="op">=</span> <span class="va">so_avid</span>,
  cluster_col <span class="op">=</span> <span class="st">"mouse"</span>,                   <span class="co"># Column containing cell clusters to compare</span>
  n_clones    <span class="op">=</span> <span class="fl">10</span>,                        <span class="co"># Number of top clonotypes to plot</span>
  type        <span class="op">=</span> <span class="st">"bar"</span>,                     <span class="co"># Type of plot, 'bar' or 'line'</span>
  plot_colors <span class="op">=</span> <span class="va">clrs</span>
<span class="op">)</span></code></pre></div>
<p><img src="avid_files/figure-html/abund_plots-1.png" width="768"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="repertoire-diversity">Repertoire Diversity<a class="anchor" aria-label="anchor" href="#repertoire-diversity"></a>
</h2>
<p>The function <code><a href="../reference/calc_diversity.html">calc_diversity()</a></code> will calculate repertoire diversity based on the number of cells that share each clonotype. The <code>cluster_col</code> argument can be used to group cells based on a meta.data column prior to calculating diversity. <code><a href="../reference/calc_diversity.html">calc_diversity()</a></code> uses the R package <a href="https://github.com/kylebittinger/abdiv" class="external-link">abdiv</a> for performing diversity calculations and any <a href="https://github.com/kylebittinger/abdiv" class="external-link">abdiv</a> diversity function can be specified using the <code>method</code> argument. It is important to read the <a href="https://github.com/kylebittinger/abdiv" class="external-link">abdiv</a> documentation to ensure the selected function is appropriate for your analysis.</p>
<p>Possible methods for calculating diversity include:</p>
<pre><code> [1] "berger_parker_d"  "brillouin_d"      "dominance"       
 [4] "heip_e"           "invsimpson"       "kempton_taylor_q"
 [7] "margalef"         "mcintosh_d"       "mcintosh_e"      
[10] "menhinick"        "pielou_e"         "richness"        
[13] "shannon"          "simpson"          "simpson_e"       
[16] "strong"          </code></pre>
<p><br></p>
<p>In this example we are calculating the Shannon diversity for each sample in the orig.ident meta.data column.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">so_avid</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_diversity.html">calc_diversity</a></span><span class="op">(</span>
  input       <span class="op">=</span> <span class="va">so_avid</span>,
  cluster_col <span class="op">=</span> <span class="st">"mouse"</span>,                    <span class="co"># Column containing cell clusters to compare</span>
  method      <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/shannon.html" class="external-link">shannon</a></span>              <span class="co"># abdiv method to use</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
<p>The <code><a href="../reference/plot_diversity.html">plot_diversity()</a></code> function will create plots summarizing repertoire diversity for each sample. A named list of functions can also be passed to plot multiple metrics. Two different metrics are shown in the example below. As expected, BL6 B cells have a very diverse repertoire, while MD4 cells have a restricted repertoire.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Metrics to plot</span>
<span class="va">fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
  <span class="st">"simpson"</span>  <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/simpson.html" class="external-link">simpson</a></span>,
  <span class="st">"mcintosh"</span> <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/mcintosh_d.html" class="external-link">mcintosh_d</a></span>
<span class="op">)</span>

<span class="fu"><a href="../reference/plot_diversity.html">plot_diversity</a></span><span class="op">(</span>
  input       <span class="op">=</span> <span class="va">so_avid</span>,
  cluster_col <span class="op">=</span> <span class="st">"mouse"</span>,                    <span class="co"># Column containing cell clusters to compare</span>
  method      <span class="op">=</span> <span class="va">fns</span>,                        <span class="co"># abdiv method to use</span>
  plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#E69F00"</span>, <span class="st">"#56B4E9"</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p><img src="avid_files/figure-html/div_plots-1.png" width="336"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="repertoire-overlap">Repertoire Overlap<a class="anchor" aria-label="anchor" href="#repertoire-overlap"></a>
</h2>
<p>To compare repertoires for different samples or clusters, <code><a href="../reference/calc_similarity.html">calc_similarity()</a></code> can calculate a variety of different similarity metrics. The <code>cluster_col</code> argument should be used to specify the meta.data column containing cell groups to use for comparison. Like <code><a href="../reference/calc_diversity.html">calc_diversity()</a></code>, an <a href="https://github.com/kylebittinger/abdiv" class="external-link">abdiv</a> function can be specified with the <code>method</code> argument. It is important to read the <a href="https://github.com/kylebittinger/abdiv" class="external-link">abdiv</a> documentation to ensure the selected function is appropriate for your analysis.</p>
<p>Possible methods for calculating repertoire similarity include:</p>
<pre><code> [1] "binomial_deviance"                 
 [2] "bray_curtis"                       
 [3] "bray_curtis_balanced"              
 [4] "bray_curtis_gradient"              
 [5] "canberra"                          
 [6] "chebyshev"                         
 [7] "chord"                             
 [8] "clark_coefficient_of_divergence"   
 [9] "correlation_distance"              
[10] "cosine_distance"                   
[11] "cy_dissimilarity"                  
[12] "euclidean"                         
[13] "geodesic_metric"                   
[14] "hamming"                           
[15] "hellinger"                         
[16] "horn_morisita"                     
[17] "jaccard"                           
[18] "jaccard_nestedness"                
[19] "jaccard_turnover"                  
[20] "kulczynski_first"                  
[21] "kulczynski_second"                 
[22] "kullback_leibler_divergence"       
[23] "manhattan"                         
[24] "mean_character_difference"         
[25] "minkowski"                         
[26] "modified_mean_character_difference"
[27] "morisita"                          
[28] "rms_distance"                      
[29] "rogers_tanimoto"                   
[30] "russel_rao"                        
[31] "ruzicka"                           
[32] "ruzicka_balanced"                  
[33] "ruzicka_gradient"                  
[34] "sokal_michener"                    
[35] "sokal_sneath"                      
[36] "sorenson"                          
[37] "sorenson_nestedness"               
[38] "sorenson_turnover"                 
[39] "weighted_kulczynski_second"        
[40] "yule_dissimilarity"                </code></pre>
<p><br></p>
<p>By default <code><a href="../reference/calc_similarity.html">calc_similarity()</a></code> will add a new meta.data column for each comparison. In this example we are calculating the jaccard dissimilarity index for all combinations of clusters present in the <code>seurat_clusters</code> column.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">so_avid</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_similarity.html">calc_similarity</a></span><span class="op">(</span>
  input       <span class="op">=</span> <span class="va">so_avid</span>,
  cluster_col <span class="op">=</span> <span class="st">"seurat_clusters"</span>,          <span class="co"># Column containing cell clusters to compare</span>
  method      <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/jaccard.html" class="external-link">jaccard</a></span>              <span class="co"># abdiv method to use</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
<p>A heatmap summarizing the results can be generated using the <code><a href="../reference/plot_similarity.html">plot_similarity()</a></code> function. Values closer to 1 indicate minimal overlap between the clusters.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_similarity.html">plot_similarity</a></span><span class="op">(</span>
  input       <span class="op">=</span> <span class="va">so_avid</span>,
  cluster_col <span class="op">=</span> <span class="st">"seurat_clusters"</span>,          <span class="co"># Column containing cell clusters to compare</span>
  method      <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/jaccard.html" class="external-link">jaccard</a></span>,             <span class="co"># abdiv method to use</span>
  plot_colors <span class="op">=</span> <span class="st">"#009E73"</span>,
  color       <span class="op">=</span> <span class="st">"white"</span>                     <span class="co"># Additional ggplot options</span>
<span class="op">)</span></code></pre></div>
<p><img src="avid_files/figure-html/sim_plots-1.png" width="480"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="gene-usage">Gene Usage<a class="anchor" aria-label="anchor" href="#gene-usage"></a>
</h2>
<p>The V(D)J data imported from Cell Ranger also includes the V(D)J identified for each chain. The function <code><a href="../reference/calc_gene_usage.html">calc_gene_usage()</a></code> will calculate the fraction of cells expressing each V(D)J gene and produce a table summarizing the results. The <code>chain</code> argument can be used to specify the chain(s) to use for calculating gene usage, by default results for all chains will be included.</p>
<p>In this example we are summarizing the usage of different V segments for the IGH chain</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/calc_gene_usage.html">calc_gene_usage</a></span><span class="op">(</span>
  input       <span class="op">=</span> <span class="va">so_avid</span>,
  gene_cols   <span class="op">=</span> <span class="st">"v_gene"</span>,                   <span class="co"># Column containing genes</span>
  cluster_col <span class="op">=</span> <span class="st">"mouse"</span>,                    <span class="co"># Column containing cell clusters to compare</span>
  chain       <span class="op">=</span> <span class="st">"IGH"</span>,                      <span class="co"># Chain to calculate gene usage for</span>
  chain_col   <span class="op">=</span> <span class="st">"chains"</span>                    <span class="co"># Column containing chains</span>
<span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 192 × 5</span></span>
<span class="co">#&gt;    v_gene   mouse n_cells  freq    pct</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> IGHV1-11 BL6      <span style="text-decoration: underline;">3</span>694     2 0.054<span style="text-decoration: underline;">1</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> IGHV1-11 MD4       126     0 0     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> IGHV1-12 BL6      <span style="text-decoration: underline;">3</span>694    27 0.731 </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> IGHV1-12 MD4       126     0 0     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> IGHV1-15 BL6      <span style="text-decoration: underline;">3</span>694    75 2.03  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> IGHV1-15 MD4       126     0 0     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> IGHV1-18 BL6      <span style="text-decoration: underline;">3</span>694    82 2.22  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> IGHV1-18 MD4       126     0 0     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> IGHV1-19 BL6      <span style="text-decoration: underline;">3</span>694    59 1.60  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> IGHV1-19 MD4       126     0 0     </span>
<span class="co">#&gt; <span style="color: #949494;"># … with 182 more rows</span></span></code></pre></div>
<p><br></p>
<p>The function <code><a href="../reference/plot_gene_usage.html">plot_gene_usage()</a></code> can be used visualize gene usage across clusters. Using the <code>yaxis</code> argument, the percentage of cells (percent) or total number of cells (frequency) expressing each gene can be shown. The number of top genes (most frequent) to plot can also be specified with <code>n_genes</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_gene_usage.html">plot_gene_usage</a></span><span class="op">(</span>
  input       <span class="op">=</span> <span class="va">so_avid</span>,
  gene_cols   <span class="op">=</span> <span class="st">"v_gene"</span>,                   <span class="co"># Column(s) containing genes</span>
  type        <span class="op">=</span> <span class="st">"bar"</span>,                      <span class="co"># Type of plot, 'heatmap' or 'bar'</span>
  chain       <span class="op">=</span> <span class="st">"IGH"</span>,                      <span class="co"># Chain to plot</span>
  n_genes     <span class="op">=</span> <span class="fl">50</span>,                         <span class="co"># The number of top genes to plot</span>
  plot_colors <span class="op">=</span> <span class="st">"#0072B2"</span>
<span class="op">)</span></code></pre></div>
<p><img src="avid_files/figure-html/gene_usage_1-1.png" width="1056"></p>
<p><br></p>
<p>By passing multiple columns to <code>gene_cols</code>, the frequency of V-J gene pairings can also be summarized. In this example we are only looking at IGK chains.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/calc_gene_usage.html">calc_gene_usage</a></span><span class="op">(</span>
  input       <span class="op">=</span> <span class="va">so_avid</span>,
  gene_cols   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"v_gene"</span>, <span class="st">"j_gene"</span><span class="op">)</span>,      <span class="co"># Column(s) containing genes</span>
  cluster_col <span class="op">=</span> <span class="st">"mouse"</span>,                    <span class="co"># Column containing cell clusters to compare</span>
  chain       <span class="op">=</span> <span class="st">"IGK"</span>                       <span class="co"># Chain to plot</span>
<span class="op">)</span></code></pre></div>
<p><br></p>
<p>When multiple gene columns are passed to <code><a href="../reference/plot_gene_usage.html">plot_gene_usage()</a></code>, a list of plots will be returned, one for each cluster in the <code>cluster_col</code> column.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ggs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_gene_usage.html">plot_gene_usage</a></span><span class="op">(</span>
  input       <span class="op">=</span> <span class="va">so_avid</span>,
  gene_cols   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"v_gene"</span>, <span class="st">"j_gene"</span><span class="op">)</span>,      <span class="co"># Column(s) containing genes</span>
  cluster_col <span class="op">=</span> <span class="st">"mouse"</span>,                    <span class="co"># Column containing cell clusters to compare</span>
  chain       <span class="op">=</span> <span class="st">"IGK"</span>,                      <span class="co"># Chain to plot</span>
  plot_colors <span class="op">=</span> <span class="st">"#6A51A3"</span>,
  n_genes     <span class="op">=</span> <span class="fl">20</span>
<span class="op">)</span>

<span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">ggs</span><span class="op">)</span></code></pre></div>
<p><img src="avid_files/figure-html/gene_usage_2-1.png" width="1152"></p>
<p><br></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
