<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="djvdj">
<title>CDR3 Clustering • djvdj</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.4/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.4/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="CDR3 Clustering">
<meta property="og:description" content="djvdj">
<meta property="og:image" content="https://rnabioco.github.io/djvdj/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">djvdj</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Data import and manipulation</h6>
    <a class="dropdown-item" href="../articles/import.html">Data Import</a>
    <a class="dropdown-item" href="../articles/mutate.html">Data Manipulation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>VDJ analysis</h6>
    <a class="dropdown-item" href="../articles/frequency.html">Clonotype Frequencies</a>
    <a class="dropdown-item" href="../articles/diversity.html">Repertoire Diversity</a>
    <a class="dropdown-item" href="../articles/similarity.html">Repertoire Similarity</a>
    <a class="dropdown-item" href="../articles/gene-usage.html">Gene Usage</a>
    <a class="dropdown-item" href="../articles/clustering.html">CDR3 Clustering</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualization</h6>
    <a class="dropdown-item" href="../articles/plotting.html">Plotting</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/djvdj/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>CDR3 Clustering</h1>
            
            <h4 data-toc-skip class="date">December 15 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/djvdj/blob/HEAD/vignettes/clustering.Rmd" class="external-link"><code>vignettes/clustering.Rmd</code></a></small>
      <div class="d-none name"><code>clustering.Rmd</code></div>
    </div>

    
    
<p><br></p>
<p>This vignette provides detailed examples for clustering cells based
on CDR3 sequences. For the examples shown below, we use data for
splenocytes from BL6 and MD4 mice collected using the 10X Genomics
scRNA-seq platform. MD4 B cells are monoclonal and specifically bind hen
egg lysozyme.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/djvdj/">djvdj</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add V(D)J data to object</span></span>
<span><span class="va">vdj_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  BL6 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/splen/BL6_BCR"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span>,</span>
<span>  MD4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/splen/MD4_BCR"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">so</span> <span class="op">&lt;-</span> <span class="va">splen_so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span><span class="va">vdj_dirs</span>, define_clonotypes <span class="op">=</span> <span class="st">"cdr3_gene"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<div class="section level2">
<h2 id="clustering-sequences">Clustering sequences<a class="anchor" aria-label="anchor" href="#clustering-sequences"></a>
</h2>
<p>The <code><a href="../reference/cluster_sequences.html">cluster_sequences()</a></code> function can be used to cluster
cells based on CDR3 sequences, or any other sequences present in the
object. Provide the meta.data column containing sequences to the
<code>data_col</code> argument. By default if a cell has multiple
chains, the sequences will be concatenated.</p>
<p>By default, distances are calculated for amino acid sequences using
the <a href="https://pubmed.ncbi.nlm.nih.gov/1438297/" class="external-link">BLOSUM62</a>
substitution matrix, which is based on observed amino acid frequencies
and substitution probabilities. The calculated distances are then used
to cluster cells using either the Louvain or Leiden clustering
algorithms. The coarseness of the clusters an be adjusted using the
<code>resolution</code> argument with smaller values returning fewer
clusters.</p>
<p>In this example we are clustering cells based on the CDR3 amino acid
sequence.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/cluster_sequences.html">cluster_sequences</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"cdr3"</span>,</span>
<span>    method      <span class="op">=</span> <span class="st">"louvain"</span>,   <span class="co"># clustering method</span></span>
<span>    dist_method <span class="op">=</span> <span class="st">"BLOSUM62"</span>,  <span class="co"># method for calculating sequence distances</span></span>
<span>    resolution  <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Use the <code>chain</code> argument to cluster using only sequences
from a specific chain.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/cluster_sequences.html">cluster_sequences</a></span><span class="op">(</span></span>
<span>    data_col <span class="op">=</span> <span class="st">"cdr3"</span>,</span>
<span>    chain    <span class="op">=</span> <span class="st">"IGK"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>By default the Uniform Manifold Approximation and Projection (UMAP)
dimensional reduction method will be performed and UMAP coordinates will
be added to the object. To skip this step, set the <code>run_umap</code>
argument to <code>FALSE</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/cluster_sequences.html">cluster_sequences</a></span><span class="op">(</span></span>
<span>    data_col <span class="op">=</span> <span class="st">"cdr3"</span>,</span>
<span>    chain    <span class="op">=</span> <span class="st">"IGK"</span>,</span>
<span>    run_umap <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>To return clustering results for multiple resolutions, a vector can
be provided to the <code>resolution</code> argument.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span></span>
<span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/cluster_sequences.html">cluster_sequences</a></span><span class="op">(</span></span>
<span>    data_col   <span class="op">=</span> <span class="st">"cdr3"</span>,</span>
<span>    chain      <span class="op">=</span> <span class="st">"IGK"</span>,</span>
<span>    resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="fl">0.8</span>, <span class="fl">1.6</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="plotting-clusters">Plotting clusters<a class="anchor" aria-label="anchor" href="#plotting-clusters"></a>
</h2>
<p>Clustering results can be visualized on a UMAP projection using the
generic plotting function <code><a href="../reference/plot_features.html">plot_features()</a></code>. Colors can be
adjusted using the <code>plot_colors</code> argument.</p>
<p>This function will often generate a warning saying that rows with
missing values were removed, this is expected since some cells do not
have V(D)J data and will not have UMAP coordinates.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clrs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">11</span>, <span class="st">"Paired"</span><span class="op">)</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">11</span><span class="op">)</span></span>
<span></span>
<span><span class="va">so_vdj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_features.html">plot_features</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"cdr3_UMAP_1"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"cdr3_UMAP_2"</span>,</span>
<span>    feature <span class="op">=</span> <span class="st">"cdr3_cluster_0.4"</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="va">clrs</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="clustering_files/figure-html/unnamed-chunk-6-1.png" width="768"></p>
<p>To visualize the proportion of BL6 and MD4 cells in each cluster we
can create a stacked bargraph using the <code><a href="../reference/plot_frequency.html">plot_frequency()</a></code>
function. MD4 cells are monoclonal and as expected these cells are found
almost exclusively in a single cluster.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_frequency.html">plot_frequency</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"cdr3_cluster_0.4"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="va">clrs</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="clustering_files/figure-html/unnamed-chunk-7-1.png" width="768"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="cdr3-motifs">CDR3 motifs<a class="anchor" aria-label="anchor" href="#cdr3-motifs"></a>
</h2>
<p>Using the <code><a href="../reference/plot_motifs.html">plot_motifs()</a></code> function we can generate
sequence motifs for each cluster. We just need to provide the same
<code>data_col</code> and <code>chain</code> used for clustering. To
create a separate motif for each cluster, we also need to provide the
column containing cluster IDs to the <code>cluster_col</code>
argument.</p>
<p>As expected we see that most cells within the MD4 cluster have the
exact same IGK CDR3 sequence.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_motifs.html">plot_motifs</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"cdr3"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"cdr3_cluster_0.4"</span>,</span>
<span>    chain       <span class="op">=</span> <span class="st">"IGK"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="clustering_files/figure-html/unnamed-chunk-10-1.png" width="960"></p>
<p>By default, sequences are aligned to the 5’ end and trimmed based on
the <code>width</code> parameter. Sequences can be aligned to the 3’ end
using the <code>align_end</code> parameter. Sequences longer than the
width cutoff are trimmed, sequences shorter than the width cutoff are
removed. By default the width cutoff is automatically selected for each
cluster to include at least 75% of sequences.</p>
<p>In this example we generate motifs for the last 11 amino acids of the
CDR3.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_motifs.html">plot_motifs</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"cdr3"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"cdr3_cluster_0.4"</span>,</span>
<span>    chain       <span class="op">=</span> <span class="st">"IGK"</span>,</span>
<span>    align_end   <span class="op">=</span> <span class="st">"3"</span>,</span>
<span>    width       <span class="op">=</span> <span class="fl">11</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="clustering_files/figure-html/unnamed-chunk-11-1.png" width="960"></p>
<p>Plot colors can be modified using the <code>plot_colors</code>
argument and the number of rows used to arrange panels can be adjusted
with the <code>panel_nrow</code> argument. Like most djvdj plotting
functions, <code><a href="../reference/plot_motifs.html">plot_motifs()</a></code> will return a ggplot2 object that
can be modified with other ggplot2 functions.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_motifs.html">plot_motifs</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"cdr3"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"cdr3_cluster_0.4"</span>,</span>
<span>    chain       <span class="op">=</span> <span class="st">"IGK"</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">5</span>, <span class="st">"Set1"</span><span class="op">)</span>,</span>
<span>    panel_nrow  <span class="op">=</span> <span class="fl">4</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    axis.text.x  <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="clustering_files/figure-html/unnamed-chunk-12-1.png" width="768"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.1 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] RColorBrewer_1.1-3 ggplot2_3.4.0      SeuratObject_4.1.3</span></span>
<span><span class="co">#&gt; [4] Seurat_4.3.0       djvdj_0.0.0.9000  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] utf8_1.2.2                  spatstat.explore_3.0-5     </span></span>
<span><span class="co">#&gt;   [3] reticulate_1.26             tidyselect_1.2.0           </span></span>
<span><span class="co">#&gt;   [5] htmlwidgets_1.5.4           grid_4.2.2                 </span></span>
<span><span class="co">#&gt;   [7] BiocParallel_1.30.4         Rtsne_0.16                 </span></span>
<span><span class="co">#&gt;   [9] munsell_0.5.0               codetools_0.2-18           </span></span>
<span><span class="co">#&gt;  [11] ragg_1.2.4                  ica_1.0-3                  </span></span>
<span><span class="co">#&gt;  [13] future_1.29.0               miniUI_0.1.1.1             </span></span>
<span><span class="co">#&gt;  [15] withr_2.5.0                 spatstat.random_3.0-1      </span></span>
<span><span class="co">#&gt;  [17] colorspace_2.0-3            progressr_0.12.0           </span></span>
<span><span class="co">#&gt;  [19] Biobase_2.56.0              highr_0.9                  </span></span>
<span><span class="co">#&gt;  [21] knitr_1.41                  stats4_4.2.2               </span></span>
<span><span class="co">#&gt;  [23] SingleCellExperiment_1.18.1 ROCR_1.0-11                </span></span>
<span><span class="co">#&gt;  [25] tensor_1.5                  listenv_0.8.0              </span></span>
<span><span class="co">#&gt;  [27] labeling_0.4.2              MatrixGenerics_1.8.1       </span></span>
<span><span class="co">#&gt;  [29] GenomeInfoDbData_1.2.8      polyclip_1.10-4            </span></span>
<span><span class="co">#&gt;  [31] farver_2.1.1                bit64_4.0.5                </span></span>
<span><span class="co">#&gt;  [33] rprojroot_2.0.3             parallelly_1.32.1          </span></span>
<span><span class="co">#&gt;  [35] vctrs_0.5.1                 generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [37] xfun_0.35                   ggseqlogo_0.1              </span></span>
<span><span class="co">#&gt;  [39] R6_2.5.1                    doParallel_1.0.17          </span></span>
<span><span class="co">#&gt;  [41] GenomeInfoDb_1.32.4         clue_0.3-63                </span></span>
<span><span class="co">#&gt;  [43] bitops_1.0-7                spatstat.utils_3.0-1       </span></span>
<span><span class="co">#&gt;  [45] cachem_1.0.6                DelayedArray_0.22.0        </span></span>
<span><span class="co">#&gt;  [47] promises_1.2.0.1            scales_1.2.1               </span></span>
<span><span class="co">#&gt;  [49] vroom_1.6.0                 gtable_0.3.1               </span></span>
<span><span class="co">#&gt;  [51] abdiv_0.2.0                 globals_0.16.2             </span></span>
<span><span class="co">#&gt;  [53] goftest_1.2-3               rlang_1.0.6                </span></span>
<span><span class="co">#&gt;  [55] systemfonts_1.0.4           GlobalOptions_0.1.2        </span></span>
<span><span class="co">#&gt;  [57] splines_4.2.2               lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt;  [59] spatstat.geom_3.0-3         broom_1.0.1                </span></span>
<span><span class="co">#&gt;  [61] yaml_2.3.6                  reshape2_1.4.4             </span></span>
<span><span class="co">#&gt;  [63] abind_1.4-5                 backports_1.4.1            </span></span>
<span><span class="co">#&gt;  [65] httpuv_1.6.6                tools_4.2.2                </span></span>
<span><span class="co">#&gt;  [67] ellipsis_0.3.2              jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt;  [69] BiocGenerics_0.42.0         ggridges_0.5.4             </span></span>
<span><span class="co">#&gt;  [71] Rcpp_1.0.9                  plyr_1.8.8                 </span></span>
<span><span class="co">#&gt;  [73] zlibbioc_1.42.0             purrr_0.3.5                </span></span>
<span><span class="co">#&gt;  [75] RCurl_1.98-1.9              dbscan_1.1-11              </span></span>
<span><span class="co">#&gt;  [77] deldir_1.0-6                pbapply_1.6-0              </span></span>
<span><span class="co">#&gt;  [79] GetoptLong_1.0.5            cowplot_1.1.1              </span></span>
<span><span class="co">#&gt;  [81] S4Vectors_0.34.0            zoo_1.8-11                 </span></span>
<span><span class="co">#&gt;  [83] iNEXT_3.0.0                 SummarizedExperiment_1.26.1</span></span>
<span><span class="co">#&gt;  [85] ggrepel_0.9.2               cluster_2.1.4              </span></span>
<span><span class="co">#&gt;  [87] fs_1.5.2                    magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;  [89] data.table_1.14.6           scattermore_0.8            </span></span>
<span><span class="co">#&gt;  [91] circlize_0.4.15             lmtest_0.9-40              </span></span>
<span><span class="co">#&gt;  [93] RANN_2.6.1                  fitdistrplus_1.1-8         </span></span>
<span><span class="co">#&gt;  [95] matrixStats_0.63.0          hms_1.1.2                  </span></span>
<span><span class="co">#&gt;  [97] patchwork_1.1.2             mime_0.12                  </span></span>
<span><span class="co">#&gt;  [99] evaluate_0.19               xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [101] IRanges_2.30.1              gridExtra_2.3              </span></span>
<span><span class="co">#&gt; [103] shape_1.4.6                 compiler_4.2.2             </span></span>
<span><span class="co">#&gt; [105] tibble_3.1.8                KernSmooth_2.23-20         </span></span>
<span><span class="co">#&gt; [107] crayon_1.5.2                htmltools_0.5.4            </span></span>
<span><span class="co">#&gt; [109] later_1.3.0                 tzdb_0.3.0                 </span></span>
<span><span class="co">#&gt; [111] tidyr_1.2.1                 ComplexHeatmap_2.12.1      </span></span>
<span><span class="co">#&gt; [113] MASS_7.3-58.1               boot_1.3-28                </span></span>
<span><span class="co">#&gt; [115] Matrix_1.5-1                readr_2.1.3                </span></span>
<span><span class="co">#&gt; [117] cli_3.4.1                   parallel_4.2.2             </span></span>
<span><span class="co">#&gt; [119] igraph_1.3.5                GenomicRanges_1.48.0       </span></span>
<span><span class="co">#&gt; [121] pkgconfig_2.0.3             pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt; [123] sp_1.5-1                    plotly_4.10.1              </span></span>
<span><span class="co">#&gt; [125] spatstat.sparse_3.0-0       foreach_1.5.2              </span></span>
<span><span class="co">#&gt; [127] bslib_0.4.1                 XVector_0.36.0             </span></span>
<span><span class="co">#&gt; [129] stringr_1.5.0               digest_0.6.31              </span></span>
<span><span class="co">#&gt; [131] sctransform_0.3.5           RcppAnnoy_0.0.20           </span></span>
<span><span class="co">#&gt; [133] spatstat.data_3.0-0         Biostrings_2.64.1          </span></span>
<span><span class="co">#&gt; [135] rmarkdown_2.18              leiden_0.4.3               </span></span>
<span><span class="co">#&gt; [137] uwot_0.1.14                 shiny_1.7.3                </span></span>
<span><span class="co">#&gt; [139] Rsamtools_2.12.0            rjson_0.2.21               </span></span>
<span><span class="co">#&gt; [141] lifecycle_1.0.3             nlme_3.1-160               </span></span>
<span><span class="co">#&gt; [143] jsonlite_1.8.4              desc_1.4.2                 </span></span>
<span><span class="co">#&gt; [145] viridisLite_0.4.1           fansi_1.0.3                </span></span>
<span><span class="co">#&gt; [147] pillar_1.8.1                lattice_0.20-45            </span></span>
<span><span class="co">#&gt; [149] fastmap_1.1.0               httr_1.4.4                 </span></span>
<span><span class="co">#&gt; [151] survival_3.4-0              glue_1.6.2                 </span></span>
<span><span class="co">#&gt; [153] png_0.1-8                   iterators_1.0.14           </span></span>
<span><span class="co">#&gt; [155] bit_4.0.5                   stringi_1.7.8              </span></span>
<span><span class="co">#&gt; [157] sass_0.4.4                  textshaping_0.3.6          </span></span>
<span><span class="co">#&gt; [159] memoise_2.0.1               dplyr_1.0.10               </span></span>
<span><span class="co">#&gt; [161] irlba_2.3.5.1               future.apply_1.10.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
