<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="djvdj">
<title>Repertoire Diversity â€¢ djvdj</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.5/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.5/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Repertoire Diversity">
<meta property="og:description" content="djvdj">
<meta property="og:image" content="https://rnabioco.github.io/djvdj/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">djvdj</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Data import and manipulation</h6>
    <a class="dropdown-item" href="../articles/import.html">Data Import</a>
    <a class="dropdown-item" href="../articles/mutate.html">Data Manipulation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>VDJ analysis</h6>
    <a class="dropdown-item" href="../articles/frequency.html">Clonotype Frequencies</a>
    <a class="dropdown-item" href="../articles/diversity.html">Repertoire Diversity</a>
    <a class="dropdown-item" href="../articles/similarity.html">Repertoire Similarity</a>
    <a class="dropdown-item" href="../articles/gene-usage.html">Gene Usage</a>
    <a class="dropdown-item" href="../articles/clustering.html">CDR3 Clustering</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualization</h6>
    <a class="dropdown-item" href="../articles/plotting.html">Plotting</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/djvdj/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Repertoire Diversity</h1>
            
            <h4 data-toc-skip class="date">March 26 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/djvdj/blob/HEAD/vignettes/diversity.Rmd" class="external-link"><code>vignettes/diversity.Rmd</code></a></small>
      <div class="d-none name"><code>diversity.Rmd</code></div>
    </div>

    
    
<p><br></p>
<p>This vignette provides detailed examples for quantifying repertoire
diversity. For the examples shown below, we use data for splenocytes
from BL6 and MD4 mice collected using the 10X Genomics scRNA-seq
platform. MD4 B cells are monoclonal and specifically bind hen egg
lysozyme.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/djvdj/">djvdj</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add V(D)J data to object</span></span>
<span><span class="va">vdj_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  BL6 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/splen/BL6_BCR"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span>,</span>
<span>  MD4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/splen/MD4_BCR"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">so</span> <span class="op">&lt;-</span> <span class="va">splen_so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span><span class="va">vdj_dirs</span>, define_clonotypes <span class="op">=</span> <span class="st">"cdr3_gene"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<div class="section level2">
<h2 id="calculating-diversity">Calculating diversity<a class="anchor" aria-label="anchor" href="#calculating-diversity"></a>
</h2>
<p>To calculate repertoire diversity and store the results in the object
meta.data, the <code><a href="../reference/calc_diversity.html">calc_diversity()</a></code> function can be used. This
function is designed to specifically work with the R package <a href="https://github.com/kylebittinger/abdiv" class="external-link">abdiv</a>. The diversity
metric can be selected by passing the name of the function to the
<code>method</code> argument. Any alpha diversity function from the
abdiv package that takes species counts as input can be used. Be sure to
read the documentation for the function you are using to ensure it is
appropriate for your analysis.</p>
<p>In this example we are calculating the Shannon entropy for BL6 and
MD4 samples.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_diversity.html">calc_diversity</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    method      <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/shannon.html" class="external-link">shannon</a></span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Estimations of species diversity are influenced by sample size. One
approach to address this is to equalize the number of cells present in
each cluster. The <code>downsample</code> argument will randomly sample
cells so each sample being tested has the same number of cells as the
smallest cluster. The bootstrapped standard error can also be calculated
by setting the number of bootstrap samples with the <code>n_boots</code>
argument.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_diversity.html">calc_diversity</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    downsample  <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    n_boots     <span class="op">=</span> <span class="fl">50</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Diversity metrics can also be calculated for a specific chain. To do
this, the column passed to the <code>data_col</code> argument must
contain per-chain data, such as CDR3 amino acid or nucleotide sequences.
In this example diversity is calculated based only on heavy chain CDR3
sequences.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_diversity.html">calc_diversity</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"cdr3_nt"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    chain       <span class="op">=</span> <span class="st">"IGH"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="plotting-diversity">Plotting diversity<a class="anchor" aria-label="anchor" href="#plotting-diversity"></a>
</h2>
<p>The <code><a href="../reference/plot_diversity.html">plot_diversity()</a></code> function will create plots
summarizing repertoire diversity for each sample. A named list of
functions can also be passed to plot multiple metrics. Two metrics for
measuring diversity are the Simpson index and Shannon entropy. Both of
these metrics are influenced by species richness (number of unique
sequences) and evenness (relative abundance of sequences). Pielouâ€™s
index will specifically measure species evenness. For these metrics,
maximally diverse samples will return a value of 1.</p>
<p>As expected, BL6 B cells have a very diverse repertoire, while MD4
cells have a restricted repertoire.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">div_fns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"simpson"</span> <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/simpson.html" class="external-link">simpson</a></span>,</span>
<span>  <span class="st">"shannon"</span> <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/shannon.html" class="external-link">shannon</a></span>,</span>
<span>  <span class="st">"pielou evenness"</span> <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/shannon.html" class="external-link">pielou_e</a></span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_diversity.html">plot_diversity</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    method      <span class="op">=</span> <span class="va">div_fns</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="diversity_files/figure-html/unnamed-chunk-5-1.png" width="768"></p>
<p>Diversity plots can also be separated based on an additional grouping
variable such as treatment group (e.g.Â pacebo vs drug) or disease status
(e.g.Â healthy vs disease). This will generate boxplots with each point
representing a label present in the <code>cluster_col</code> column. In
this example we have 3 BL6 and 3 MD4 samples, so there should be 5
points shown for each boxplot.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_diversity.html">plot_diversity</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    group_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    method      <span class="op">=</span> <span class="va">div_fns</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="diversity_files/figure-html/unnamed-chunk-6-1.png" width="768"></p>
<p>Additional arguments are provided to adjust plot aesthetics. The
<code>plot_colors</code> parameter can be used to modify colors, the
<code>panel_nrow</code> and <code>panel_scales</code> arguments will
adjust the plot scales and number of rows used to arrange plots.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_diversity.html">plot_diversity</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    group_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    method      <span class="op">=</span> <span class="va">div_fns</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>BL6 <span class="op">=</span> <span class="st">"#3182bd"</span>, MD4 <span class="op">=</span> <span class="st">"#fec44f"</span><span class="op">)</span>,</span>
<span>    panel_nrow  <span class="op">=</span> <span class="fl">2</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="diversity_files/figure-html/unnamed-chunk-7-1.png" width="576"></p>
<p><code><a href="../reference/plot_diversity.html">plot_diversity()</a></code> returns a ggplot object that can be
modified with ggplot2 functions such as <code><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">ggplot2::theme()</a></code>.
Plots can be further adjusted by passing aesthetic parameters directly
to ggplot2, e.g.Â <code>alpha</code>, <code>linetype</code>,
<code>color</code>, etc.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_diversity.html">plot_diversity</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    method      <span class="op">=</span> <span class="va">div_fns</span>,</span>
<span>    </span>
<span>    alpha    <span class="op">=</span> <span class="fl">0.5</span>,         <span class="co"># parameters to pass to ggplot2</span></span>
<span>    linetype <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    color    <span class="op">=</span> <span class="st">"black"</span>    </span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>strip.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="diversity_files/figure-html/unnamed-chunk-8-1.png" width="768"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="rarefaction-curves">Rarefaction curves<a class="anchor" aria-label="anchor" href="#rarefaction-curves"></a>
</h2>
<p>Another approach to ensure differences in sample size are not having
an undue influence on diversity results is to plot rarefaction curves.
This method involves calculating species diversity for different sized
samples generated by randomly downsampling each cluster. By default the
bootstrapped 95% confidence interval will also be plotted.</p>
<p>Calculations used to generate rarefaction curves are performed using
the iNEXT package. There are three diversity calculations that can be
specified with the <code>method</code> argument:</p>
<ul>
<li>â€˜richnessâ€™, species richness, this is equivalent to the calculation
performed by <code><a href="https://rdrr.io/pkg/abdiv/man/richness.html" class="external-link">abdiv::richness()</a></code>
</li>
<li>â€˜shannonâ€™, the exponential of Shannon entropy</li>
<li>â€˜invsimpsonâ€™, the inverse Simpson index, this is equivalent to the
calculation performed by <code><a href="https://rdrr.io/pkg/abdiv/man/simpson.html" class="external-link">abdiv::invsimpson()</a></code>
</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_rarefaction.html">plot_rarefaction</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    method      <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"richness"</span>, <span class="st">"shannon"</span>, <span class="st">"invsimpson"</span><span class="op">)</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#3182bd"</span>, <span class="st">"#fec44f"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="diversity_files/figure-html/unnamed-chunk-9-1.png" width="768"></p>
<p>If the 95% confidence interval is not desired, set
<code>n_boots</code> to 0. In this example we also plot a separate line
for each BL6 and MD4 sample.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_rarefaction.html">plot_rarefaction</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    method      <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"richness"</span>, <span class="st">"shannon"</span>, <span class="st">"invsimpson"</span><span class="op">)</span>,</span>
<span>    n_boots     <span class="op">=</span> <span class="fl">0</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="diversity_files/figure-html/unnamed-chunk-10-1.png" width="768"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.2.3 (2023-03-15)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggplot2_3.4.1      SeuratObject_4.1.3 Seurat_4.3.0      </span></span>
<span><span class="co">#&gt; [4] djvdj_0.1.0       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] backports_1.4.1             circlize_0.4.15            </span></span>
<span><span class="co">#&gt;   [3] systemfonts_1.0.4           plyr_1.8.8                 </span></span>
<span><span class="co">#&gt;   [5] igraph_1.4.1                lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt;   [7] sp_1.6-0                    splines_4.2.3              </span></span>
<span><span class="co">#&gt;   [9] listenv_0.9.0               scattermore_0.8            </span></span>
<span><span class="co">#&gt;  [11] GenomeInfoDb_1.34.9         digest_0.6.31              </span></span>
<span><span class="co">#&gt;  [13] foreach_1.5.2               htmltools_0.5.5            </span></span>
<span><span class="co">#&gt;  [15] fansi_1.0.4                 magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;  [17] memoise_2.0.1               tensor_1.5                 </span></span>
<span><span class="co">#&gt;  [19] cluster_2.1.4               doParallel_1.0.17          </span></span>
<span><span class="co">#&gt;  [21] ROCR_1.0-11                 tzdb_0.3.0                 </span></span>
<span><span class="co">#&gt;  [23] ComplexHeatmap_2.14.0       globals_0.16.2             </span></span>
<span><span class="co">#&gt;  [25] readr_2.1.4                 matrixStats_0.63.0         </span></span>
<span><span class="co">#&gt;  [27] vroom_1.6.1                 spatstat.sparse_3.0-1      </span></span>
<span><span class="co">#&gt;  [29] pkgdown_2.0.7               colorspace_2.1-0           </span></span>
<span><span class="co">#&gt;  [31] ggrepel_0.9.3               textshaping_0.3.6          </span></span>
<span><span class="co">#&gt;  [33] xfun_0.37                   dplyr_1.1.1                </span></span>
<span><span class="co">#&gt;  [35] crayon_1.5.2                RCurl_1.98-1.10            </span></span>
<span><span class="co">#&gt;  [37] jsonlite_1.8.4              spatstat.data_3.0-1        </span></span>
<span><span class="co">#&gt;  [39] progressr_0.13.0            survival_3.5-3             </span></span>
<span><span class="co">#&gt;  [41] zoo_1.8-11                  iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [43] glue_1.6.2                  polyclip_1.10-4            </span></span>
<span><span class="co">#&gt;  [45] gtable_0.3.3                zlibbioc_1.44.0            </span></span>
<span><span class="co">#&gt;  [47] XVector_0.38.0              leiden_0.4.3               </span></span>
<span><span class="co">#&gt;  [49] GetoptLong_1.0.5            DelayedArray_0.24.0        </span></span>
<span><span class="co">#&gt;  [51] future.apply_1.10.0         shape_1.4.6                </span></span>
<span><span class="co">#&gt;  [53] SingleCellExperiment_1.20.1 BiocGenerics_0.44.0        </span></span>
<span><span class="co">#&gt;  [55] abind_1.4-5                 scales_1.2.1               </span></span>
<span><span class="co">#&gt;  [57] spatstat.random_3.1-4       miniUI_0.1.1.1             </span></span>
<span><span class="co">#&gt;  [59] Rcpp_1.0.10                 viridisLite_0.4.1          </span></span>
<span><span class="co">#&gt;  [61] xtable_1.8-4                clue_0.3-64                </span></span>
<span><span class="co">#&gt;  [63] reticulate_1.28             bit_4.0.5                  </span></span>
<span><span class="co">#&gt;  [65] stats4_4.2.3                htmlwidgets_1.6.2          </span></span>
<span><span class="co">#&gt;  [67] httr_1.4.5                  RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;  [69] ellipsis_0.3.2              ica_1.0-3                  </span></span>
<span><span class="co">#&gt;  [71] farver_2.1.1                pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [73] uwot_0.1.14                 deldir_1.0-6               </span></span>
<span><span class="co">#&gt;  [75] sass_0.4.5                  utf8_1.2.3                 </span></span>
<span><span class="co">#&gt;  [77] labeling_0.4.2              tidyselect_1.2.0           </span></span>
<span><span class="co">#&gt;  [79] rlang_1.1.0                 reshape2_1.4.4             </span></span>
<span><span class="co">#&gt;  [81] later_1.3.0                 munsell_0.5.0              </span></span>
<span><span class="co">#&gt;  [83] tools_4.2.3                 cachem_1.0.7               </span></span>
<span><span class="co">#&gt;  [85] cli_3.6.1                   generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [87] broom_1.0.4                 ggridges_0.5.4             </span></span>
<span><span class="co">#&gt;  [89] evaluate_0.20               stringr_1.5.0              </span></span>
<span><span class="co">#&gt;  [91] fastmap_1.1.1               goftest_1.2-3              </span></span>
<span><span class="co">#&gt;  [93] yaml_2.3.7                  ragg_1.2.5                 </span></span>
<span><span class="co">#&gt;  [95] bit64_4.0.5                 knitr_1.42                 </span></span>
<span><span class="co">#&gt;  [97] fs_1.6.1                    fitdistrplus_1.1-8         </span></span>
<span><span class="co">#&gt;  [99] purrr_1.0.1                 RANN_2.6.1                 </span></span>
<span><span class="co">#&gt; [101] nlme_3.1-162                pbapply_1.7-0              </span></span>
<span><span class="co">#&gt; [103] future_1.32.0               mime_0.12                  </span></span>
<span><span class="co">#&gt; [105] compiler_4.2.3              plotly_4.10.1              </span></span>
<span><span class="co">#&gt; [107] png_0.1-8                   spatstat.utils_3.0-2       </span></span>
<span><span class="co">#&gt; [109] abdiv_0.2.0                 tibble_3.2.1               </span></span>
<span><span class="co">#&gt; [111] bslib_0.4.2                 stringi_1.7.12             </span></span>
<span><span class="co">#&gt; [113] highr_0.10                  desc_1.4.2                 </span></span>
<span><span class="co">#&gt; [115] lattice_0.20-45             Matrix_1.5-3               </span></span>
<span><span class="co">#&gt; [117] vctrs_0.6.1                 pillar_1.9.0               </span></span>
<span><span class="co">#&gt; [119] lifecycle_1.0.3             spatstat.geom_3.1-0        </span></span>
<span><span class="co">#&gt; [121] lmtest_0.9-40               jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt; [123] GlobalOptions_0.1.2         RcppAnnoy_0.0.20           </span></span>
<span><span class="co">#&gt; [125] data.table_1.14.8           cowplot_1.1.1              </span></span>
<span><span class="co">#&gt; [127] bitops_1.0-7                irlba_2.3.5.1              </span></span>
<span><span class="co">#&gt; [129] ggtrace_0.2.0               httpuv_1.6.9               </span></span>
<span><span class="co">#&gt; [131] patchwork_1.1.2             GenomicRanges_1.50.2       </span></span>
<span><span class="co">#&gt; [133] R6_2.5.1                    promises_1.2.0.1           </span></span>
<span><span class="co">#&gt; [135] gridExtra_2.3               KernSmooth_2.23-20         </span></span>
<span><span class="co">#&gt; [137] IRanges_2.32.0              parallelly_1.35.0          </span></span>
<span><span class="co">#&gt; [139] codetools_0.2-19            boot_1.3-28.1              </span></span>
<span><span class="co">#&gt; [141] MASS_7.3-58.2               SummarizedExperiment_1.28.0</span></span>
<span><span class="co">#&gt; [143] rprojroot_2.0.3             rjson_0.2.21               </span></span>
<span><span class="co">#&gt; [145] withr_2.5.0                 iNEXT_3.0.0                </span></span>
<span><span class="co">#&gt; [147] sctransform_0.3.5           S4Vectors_0.36.2           </span></span>
<span><span class="co">#&gt; [149] GenomeInfoDbData_1.2.9      parallel_4.2.3             </span></span>
<span><span class="co">#&gt; [151] hms_1.1.3                   grid_4.2.3                 </span></span>
<span><span class="co">#&gt; [153] tidyr_1.3.0                 rmarkdown_2.20             </span></span>
<span><span class="co">#&gt; [155] MatrixGenerics_1.10.0       Rtsne_0.16                 </span></span>
<span><span class="co">#&gt; [157] spatstat.explore_3.1-0      Biobase_2.58.0             </span></span>
<span><span class="co">#&gt; [159] shiny_1.7.4</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
