<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="djvdj">
<title>Clonotype Frequencies â€¢ djvdj</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.4/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.4/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Clonotype Frequencies">
<meta property="og:description" content="djvdj">
<meta property="og:image" content="https://rnabioco.github.io/djvdj/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">djvdj</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Data import and manipulation</h6>
    <a class="dropdown-item" href="../articles/import.html">Data Import</a>
    <a class="dropdown-item" href="../articles/mutate.html">Data Manipulation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>VDJ analysis</h6>
    <a class="dropdown-item" href="../articles/frequency.html">Clonotype Frequencies</a>
    <a class="dropdown-item" href="../articles/diversity.html">Repertoire Diversity</a>
    <a class="dropdown-item" href="../articles/similarity.html">Repertoire Similarity</a>
    <a class="dropdown-item" href="../articles/gene-usage.html">Gene Usage</a>
    <a class="dropdown-item" href="../articles/clustering.html">CDR3 Clustering</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualization</h6>
    <a class="dropdown-item" href="../articles/plotting.html">Plotting</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/djvdj/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Clonotype Frequencies</h1>
            
            <h4 data-toc-skip class="date">December 15 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/djvdj/blob/HEAD/vignettes/frequency.Rmd" class="external-link"><code>vignettes/frequency.Rmd</code></a></small>
      <div class="d-none name"><code>frequency.Rmd</code></div>
    </div>

    
    
<p><br></p>
<p>This vignette provides detailed examples for quantifying differences
in clonotype frequencies. For the examples shown below, we use data for
splenocytes from BL6 and MD4 mice collected using the 10X Genomics
scRNA-seq platform. MD4 B cells are monoclonal and specifically bind hen
egg lysozyme.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/djvdj/">djvdj</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add V(D)J data to object</span></span>
<span><span class="va">vdj_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  BL6 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/splen/BL6_BCR"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span>,</span>
<span>  MD4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/splen/MD4_BCR"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">so</span> <span class="op">&lt;-</span> <span class="va">splen_so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span><span class="va">vdj_dirs</span>, define_clonotypes <span class="op">=</span> <span class="st">"cdr3_gene"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<div class="section level2">
<h2 id="calculating-clonotype-frequencies">Calculating clonotype frequencies<a class="anchor" aria-label="anchor" href="#calculating-clonotype-frequencies"></a>
</h2>
<p>To quantify clonotype frequencies and store the results in the object
meta.data, the <code><a href="../reference/calc_frequency.html">calc_frequency()</a></code> function can be used. This
will add columns showing the number of occurrences of each clonotype
(â€˜freqâ€™), the percentage of cells sharing the clonotype (â€˜pctâ€™), and a
label that can be used for plotting (â€˜grpâ€™). By default these
calculations will be performed for all cells in the object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_frequency.html">calc_frequency</a></span><span class="op">(</span>data_col <span class="op">=</span> <span class="st">"clonotype_id"</span><span class="op">)</span></span></code></pre></div>
<p>To calculate clonotype frequencies separately for samples or
clusters, the <code>cluster_col</code> argument can be used. To do this
just specify the name of the column containing the sample or cluster IDs
for each cell.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_frequency.html">calc_frequency</a></span><span class="op">(</span></span>
<span>    data_col <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>When <code>cluster_col</code> is specified, an additional meta.data
column (â€˜sharedâ€™) will be added indicating whether the clonotype is
shared between multiple clusters.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="st">"meta.data"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt;                        orig.ident nCount_RNA nFeature_RNA RNA_snn_res.1</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1        BL6        202           25             0</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1        BL6         42           20             3</span></span>
<span><span class="co">#&gt;                        seurat_clusters     UMAP_1     UMAP_2</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1               0 -1.7410439  0.8840749</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1               3  0.9088528 -1.2614110</span></span>
<span><span class="co">#&gt;                                  type         r cell_type sample</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1 B cells (B.T3) 0.6256432   B cells  BL6-1</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1 B cells (B.T2) 0.6350011   B cells  BL6-1</span></span>
<span><span class="co">#&gt;                        exact_subclonotype_id chains n_chains        cdr3</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1                    NA   &lt;NA&gt;       NA        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1                     1    IGK        1 CFQGSHVPWTF</span></span>
<span><span class="co">#&gt;                                                  cdr3_nt cdr3_length</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1                              &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1 TGCTTTCAAGGTTCACATGTTCCGTGGACGTTC          11</span></span>
<span><span class="co">#&gt;                        cdr3_nt_length    v_gene d_gene j_gene c_gene</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1           &lt;NA&gt;      &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1             33 IGKV1-117   None  IGKJ1   IGKC</span></span>
<span><span class="co">#&gt;                        isotype reads umis productive full_length paired</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1    &lt;NA&gt;  &lt;NA&gt; &lt;NA&gt;       &lt;NA&gt;        &lt;NA&gt;     NA</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1    None   352   21       TRUE        TRUE  FALSE</span></span>
<span><span class="co">#&gt;                        clonotype_id clonotype_id_freq clonotype_id_pct</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1         &lt;NA&gt;                NA               NA</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1  clonotype34                 1         1.818182</span></span>
<span><span class="co">#&gt;                        clonotype_id_shared clonotype_id_grp</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1                  NA             &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1                TRUE                1</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="plotting-clonotype-frequencies">Plotting clonotype frequencies<a class="anchor" aria-label="anchor" href="#plotting-clonotype-frequencies"></a>
</h2>
<p>djvdj includes the <code><a href="../reference/plot_clone_frequency.html">plot_clone_frequency()</a></code> function to
visualize differences in clonotype frequency between samples or
clusters. By default this will produce bargraphs. Plot colors can be
adjusted using the <code>plot_colors</code> argument.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_clone_frequency.html">plot_clone_frequency</a></span><span class="op">(</span></span>
<span>    data_col <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="st">"#3182bd"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="frequency_files/figure-html/unnamed-chunk-5-1.png" width="768"></p>
<p>Frequencies can be calculated and plotted separately for each sample
or cluster using the <code>cluster_col</code> argument. The
<code>panel_nrow</code> and <code>panel_scales</code> arguments can be
used to add separate scales for each sample or to adjust the number of
rows used to arrange plots.</p>
<p>As expected we see that most MD4 B cells share the same clonotype,
while BL6 cells have a diverse repertoire.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_clone_frequency.html">plot_clone_frequency</a></span><span class="op">(</span></span>
<span>    data_col     <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col  <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    panel_scales <span class="op">=</span> <span class="st">"free"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="frequency_files/figure-html/unnamed-chunk-6-1.png" width="768"></p>
<p>Rank-abundance plots can also be generated by setting the
<code>method</code> argument to â€˜lineâ€™. Most djvdj plotting functions
return ggplot objects that can be further modified with ggplot2
functions. Here we further modify plot aesthetics using the
<code>ggplot::theme()</code> function. Most djvdj plotting function also
include the ability to transform the axis using the <code>trans</code>
argument.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_clone_frequency.html">plot_clone_frequency</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    method      <span class="op">=</span> <span class="st">"line"</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>MD4 <span class="op">=</span> <span class="st">"#fec44f"</span>, BL6 <span class="op">=</span> <span class="st">"#3182bd"</span><span class="op">)</span>,</span>
<span>    trans       <span class="op">=</span> <span class="st">"log10"</span>         <span class="co"># log-transform axis</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<p><img src="frequency_files/figure-html/unnamed-chunk-7-1.png" width="768"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="umap-projections">UMAP projections<a class="anchor" aria-label="anchor" href="#umap-projections"></a>
</h2>
<p>By default <code><a href="../reference/calc_frequency.html">calc_frequency()</a></code> will divide clonotypes into
groups based on frequency and add a column to the meta.data containing
these group labels. Clonotype frequencies can be summarized on a UMAP
projection by plotting the added â€˜grpâ€™ column using the generic plotting
function <code><a href="../reference/plot_features.html">plot_features()</a></code>.</p>
<p>Cells that lack BCR data will be plotted as <code>NA</code>s, the
color of these points can be adjusted using the <code>na_color</code>
argument.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create UMAP summarizing samples</span></span>
<span><span class="va">mouse_gg</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_features.html">plot_features</a></span><span class="op">(</span>feature <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create UMAP summarizing clonotype frequencies</span></span>
<span><span class="va">abun_gg</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_frequency.html">calc_frequency</a></span><span class="op">(</span></span>
<span>    data_col <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_features.html">plot_features</a></span><span class="op">(</span>feature <span class="op">=</span> <span class="st">"clonotype_id_grp"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mouse_gg</span> <span class="op">+</span> <span class="va">abun_gg</span></span></code></pre></div>
<p><img src="frequency_files/figure-html/unnamed-chunk-8-1.png" width="1056"></p>
<p>Highly abundant clonotypes can also be specifically labeled on a UMAP
projection. To do this, add a new meta.data column with the desired
label using the <code><a href="../reference/mutate_vdj.html">mutate_vdj()</a></code> function. This function works
in a similar manner as <code><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">dplyr::mutate()</a></code>, but will
specifically modify the object meta.data and allows to the user to parse
per-chain information for each cell.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_gg</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/mutate_vdj.html">mutate_vdj</a></span><span class="op">(</span></span>
<span>    top_clonotype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span></span>
<span>      <span class="va">clonotype_id</span> <span class="op">==</span> <span class="st">"clonotype56"</span>,</span>
<span>      <span class="va">clonotype_id</span>,</span>
<span>      <span class="st">"other"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_features.html">plot_features</a></span><span class="op">(</span></span>
<span>    feature     <span class="op">=</span> <span class="st">"top_clonotype"</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>other <span class="op">=</span> <span class="st">"#fec44f"</span>, clonotype56 <span class="op">=</span> <span class="st">"#3182bd"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">mouse_gg</span> <span class="op">+</span> <span class="va">top_gg</span></span></code></pre></div>
<p><img src="frequency_files/figure-html/unnamed-chunk-9-1.png" width="1056"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="other-frequency-calculations">Other frequency calculations<a class="anchor" aria-label="anchor" href="#other-frequency-calculations"></a>
</h2>
<p>In addition to clonotype abundance, <code><a href="../reference/calc_frequency.html">calc_frequency()</a></code> can
be used to summarize the frequency of any cell label present in the
object. In this example we count the number of cells present for each
cell type in each sample.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_frequency.html">calc_frequency</a></span><span class="op">(</span></span>
<span>    data_col <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>To plot the fraction of cells present for each cell type, we can use
the generic plotting function, <code><a href="../reference/plot_frequency.html">plot_frequency()</a></code>. This will
create stacked bargraphs summarizing each cell label present in the
<code>data_col</code> column. The color of each group can be specified
with the <code>plot_colors</code> argument.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_frequency.html">plot_frequency</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#3182bd"</span>, <span class="st">"#fec44f"</span>, <span class="st">"#31a354"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="frequency_files/figure-html/unnamed-chunk-11-1.png" width="768"></p>
<p>To summarize the number cells present for each cell type, set the
<code>units</code> argument to â€˜frequencyâ€™. To create grouped bargraphs,
set the <code>stack</code> argument to <code>FALSE</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_frequency.html">plot_frequency</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    units       <span class="op">=</span> <span class="st">"frequency"</span>,</span>
<span>    stack       <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="frequency_files/figure-html/unnamed-chunk-12-1.png" width="768"></p>
<p>Frequency plots can also be separated based on an additional grouping
variable such as treatment group (e.g.Â placebo vs drug) or disease
status (e.g.Â healthy vs disease). This will generate boxplots with each
point representing a label present in the <code>cluster_col</code>
column. In this example we have 3 BL6 and 3 MD4 samples, so there are 3
points shown for each boxplot.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_frequency.html">plot_frequency</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    group_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>MD4 <span class="op">=</span> <span class="st">"#fec44f"</span>, BL6 <span class="op">=</span> <span class="st">"#3182bd"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="frequency_files/figure-html/unnamed-chunk-13-1.png" width="480"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.1 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggplot2_3.4.0      SeuratObject_4.1.3 Seurat_4.3.0      </span></span>
<span><span class="co">#&gt; [4] djvdj_0.0.0.9000  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] utf8_1.2.2                  spatstat.explore_3.0-5     </span></span>
<span><span class="co">#&gt;   [3] reticulate_1.26             tidyselect_1.2.0           </span></span>
<span><span class="co">#&gt;   [5] htmlwidgets_1.5.4           grid_4.2.2                 </span></span>
<span><span class="co">#&gt;   [7] BiocParallel_1.30.4         Rtsne_0.16                 </span></span>
<span><span class="co">#&gt;   [9] munsell_0.5.0               codetools_0.2-18           </span></span>
<span><span class="co">#&gt;  [11] ragg_1.2.4                  ica_1.0-3                  </span></span>
<span><span class="co">#&gt;  [13] future_1.29.0               miniUI_0.1.1.1             </span></span>
<span><span class="co">#&gt;  [15] withr_2.5.0                 spatstat.random_3.0-1      </span></span>
<span><span class="co">#&gt;  [17] colorspace_2.0-3            progressr_0.12.0           </span></span>
<span><span class="co">#&gt;  [19] Biobase_2.56.0              highr_0.9                  </span></span>
<span><span class="co">#&gt;  [21] knitr_1.41                  stats4_4.2.2               </span></span>
<span><span class="co">#&gt;  [23] SingleCellExperiment_1.18.1 ROCR_1.0-11                </span></span>
<span><span class="co">#&gt;  [25] tensor_1.5                  listenv_0.8.0              </span></span>
<span><span class="co">#&gt;  [27] labeling_0.4.2              MatrixGenerics_1.8.1       </span></span>
<span><span class="co">#&gt;  [29] GenomeInfoDbData_1.2.8      polyclip_1.10-4            </span></span>
<span><span class="co">#&gt;  [31] farver_2.1.1                bit64_4.0.5                </span></span>
<span><span class="co">#&gt;  [33] rprojroot_2.0.3             parallelly_1.32.1          </span></span>
<span><span class="co">#&gt;  [35] vctrs_0.5.1                 generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [37] xfun_0.35                   R6_2.5.1                   </span></span>
<span><span class="co">#&gt;  [39] doParallel_1.0.17           GenomeInfoDb_1.32.4        </span></span>
<span><span class="co">#&gt;  [41] clue_0.3-63                 bitops_1.0-7               </span></span>
<span><span class="co">#&gt;  [43] spatstat.utils_3.0-1        cachem_1.0.6               </span></span>
<span><span class="co">#&gt;  [45] DelayedArray_0.22.0         promises_1.2.0.1           </span></span>
<span><span class="co">#&gt;  [47] scales_1.2.1                vroom_1.6.0                </span></span>
<span><span class="co">#&gt;  [49] gtable_0.3.1                abdiv_0.2.0                </span></span>
<span><span class="co">#&gt;  [51] globals_0.16.2              goftest_1.2-3              </span></span>
<span><span class="co">#&gt;  [53] rlang_1.0.6                 systemfonts_1.0.4          </span></span>
<span><span class="co">#&gt;  [55] GlobalOptions_0.1.2         splines_4.2.2              </span></span>
<span><span class="co">#&gt;  [57] lazyeval_0.2.2              spatstat.geom_3.0-3        </span></span>
<span><span class="co">#&gt;  [59] broom_1.0.1                 yaml_2.3.6                 </span></span>
<span><span class="co">#&gt;  [61] reshape2_1.4.4              abind_1.4-5                </span></span>
<span><span class="co">#&gt;  [63] backports_1.4.1             httpuv_1.6.6               </span></span>
<span><span class="co">#&gt;  [65] tools_4.2.2                 ellipsis_0.3.2             </span></span>
<span><span class="co">#&gt;  [67] jquerylib_0.1.4             RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;  [69] BiocGenerics_0.42.0         ggridges_0.5.4             </span></span>
<span><span class="co">#&gt;  [71] Rcpp_1.0.9                  plyr_1.8.8                 </span></span>
<span><span class="co">#&gt;  [73] zlibbioc_1.42.0             purrr_0.3.5                </span></span>
<span><span class="co">#&gt;  [75] RCurl_1.98-1.9              dbscan_1.1-11              </span></span>
<span><span class="co">#&gt;  [77] deldir_1.0-6                pbapply_1.6-0              </span></span>
<span><span class="co">#&gt;  [79] GetoptLong_1.0.5            cowplot_1.1.1              </span></span>
<span><span class="co">#&gt;  [81] S4Vectors_0.34.0            zoo_1.8-11                 </span></span>
<span><span class="co">#&gt;  [83] iNEXT_3.0.0                 SummarizedExperiment_1.26.1</span></span>
<span><span class="co">#&gt;  [85] ggrepel_0.9.2               cluster_2.1.4              </span></span>
<span><span class="co">#&gt;  [87] fs_1.5.2                    magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;  [89] data.table_1.14.6           scattermore_0.8            </span></span>
<span><span class="co">#&gt;  [91] circlize_0.4.15             lmtest_0.9-40              </span></span>
<span><span class="co">#&gt;  [93] RANN_2.6.1                  fitdistrplus_1.1-8         </span></span>
<span><span class="co">#&gt;  [95] matrixStats_0.63.0          hms_1.1.2                  </span></span>
<span><span class="co">#&gt;  [97] patchwork_1.1.2             mime_0.12                  </span></span>
<span><span class="co">#&gt;  [99] evaluate_0.19               xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [101] IRanges_2.30.1              gridExtra_2.3              </span></span>
<span><span class="co">#&gt; [103] shape_1.4.6                 compiler_4.2.2             </span></span>
<span><span class="co">#&gt; [105] tibble_3.1.8                KernSmooth_2.23-20         </span></span>
<span><span class="co">#&gt; [107] crayon_1.5.2                htmltools_0.5.4            </span></span>
<span><span class="co">#&gt; [109] later_1.3.0                 tzdb_0.3.0                 </span></span>
<span><span class="co">#&gt; [111] tidyr_1.2.1                 ComplexHeatmap_2.12.1      </span></span>
<span><span class="co">#&gt; [113] MASS_7.3-58.1               boot_1.3-28                </span></span>
<span><span class="co">#&gt; [115] Matrix_1.5-1                readr_2.1.3                </span></span>
<span><span class="co">#&gt; [117] cli_3.4.1                   parallel_4.2.2             </span></span>
<span><span class="co">#&gt; [119] igraph_1.3.5                GenomicRanges_1.48.0       </span></span>
<span><span class="co">#&gt; [121] pkgconfig_2.0.3             pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt; [123] sp_1.5-1                    plotly_4.10.1              </span></span>
<span><span class="co">#&gt; [125] spatstat.sparse_3.0-0       foreach_1.5.2              </span></span>
<span><span class="co">#&gt; [127] bslib_0.4.1                 XVector_0.36.0             </span></span>
<span><span class="co">#&gt; [129] stringr_1.5.0               digest_0.6.31              </span></span>
<span><span class="co">#&gt; [131] sctransform_0.3.5           RcppAnnoy_0.0.20           </span></span>
<span><span class="co">#&gt; [133] spatstat.data_3.0-0         Biostrings_2.64.1          </span></span>
<span><span class="co">#&gt; [135] rmarkdown_2.18              leiden_0.4.3               </span></span>
<span><span class="co">#&gt; [137] uwot_0.1.14                 shiny_1.7.3                </span></span>
<span><span class="co">#&gt; [139] Rsamtools_2.12.0            rjson_0.2.21               </span></span>
<span><span class="co">#&gt; [141] lifecycle_1.0.3             nlme_3.1-160               </span></span>
<span><span class="co">#&gt; [143] jsonlite_1.8.4              desc_1.4.2                 </span></span>
<span><span class="co">#&gt; [145] viridisLite_0.4.1           fansi_1.0.3                </span></span>
<span><span class="co">#&gt; [147] pillar_1.8.1                lattice_0.20-45            </span></span>
<span><span class="co">#&gt; [149] fastmap_1.1.0               httr_1.4.4                 </span></span>
<span><span class="co">#&gt; [151] survival_3.4-0              glue_1.6.2                 </span></span>
<span><span class="co">#&gt; [153] png_0.1-8                   iterators_1.0.14           </span></span>
<span><span class="co">#&gt; [155] bit_4.0.5                   stringi_1.7.8              </span></span>
<span><span class="co">#&gt; [157] sass_0.4.4                  textshaping_0.3.6          </span></span>
<span><span class="co">#&gt; [159] memoise_2.0.1               dplyr_1.0.10               </span></span>
<span><span class="co">#&gt; [161] irlba_2.3.5.1               future.apply_1.10.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
