<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="djvdj">
<title>Gene Usage • djvdj</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.7/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.7/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Gene Usage">
<meta property="og:description" content="djvdj">
<meta property="og:image" content="https://rnabioco.github.io/djvdj/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">djvdj</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Data import and manipulation</h6>
    <a class="dropdown-item" href="../articles/import.html">Data Import</a>
    <a class="dropdown-item" href="../articles/mutate.html">Data Manipulation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>VDJ analysis</h6>
    <a class="dropdown-item" href="../articles/frequency.html">Clonotype Frequencies</a>
    <a class="dropdown-item" href="../articles/diversity.html">Repertoire Diversity</a>
    <a class="dropdown-item" href="../articles/similarity.html">Repertoire Similarity</a>
    <a class="dropdown-item" href="../articles/gene-usage.html">Gene Usage</a>
    <a class="dropdown-item" href="../articles/clustering.html">CDR3 Clustering</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualization</h6>
    <a class="dropdown-item" href="../articles/plotting.html">Plotting</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/djvdj/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Gene Usage</h1>
            
            <h4 data-toc-skip class="date">August 29 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/djvdj/blob/HEAD/vignettes/gene-usage.Rmd" class="external-link"><code>vignettes/gene-usage.Rmd</code></a></small>
      <div class="d-none name"><code>gene-usage.Rmd</code></div>
    </div>

    
    
<p><br></p>
<p>This vignette provides detailed examples for calculating and
visualizing V(D)J gene usage. For the examples shown below, we use data
for splenocytes from BL6 and MD4 mice collected using the 10X Genomics
scRNA-seq platform. MD4 B cells are monoclonal and specifically bind hen
egg lysozyme.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/djvdj/">djvdj</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add V(D)J data to object</span></span>
<span><span class="va">vdj_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  BL6 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/splen/BL6_BCR"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span>,</span>
<span>  MD4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/splen/MD4_BCR"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">so</span> <span class="op">&lt;-</span> <span class="va">splen_so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span><span class="va">vdj_dirs</span>, define_clonotypes <span class="op">=</span> <span class="st">"cdr3_gene"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<div class="section level2">
<h2 id="calculating-gene-usage">Calculating gene usage<a class="anchor" aria-label="anchor" href="#calculating-gene-usage"></a>
</h2>
<p>The <code><a href="../reference/calc_gene_usage.html">calc_gene_usage()</a></code> function will calculate the number
of cells (‘freq’) and percentage of cells (‘pct’) with each gene in the
<code>data_cols</code> column(s). The ‘n_cells’ column shows the total
number of cells used for calculating percentages. By default these
results are added to the object meta.data, to return a data.frame set
<code>return_df</code> to <code>TRUE</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_gene_usage.html">calc_gene_usage</a></span><span class="op">(</span></span>
<span>    data_cols <span class="op">=</span> <span class="st">"v_gene"</span>,</span>
<span>    return_df <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 103 × 4</span></span></span>
<span><span class="co">#&gt;    v_gene     n_cells  freq   pct</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> IGKV5-43       329   156 47.4 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> IGKV10-96      329    14  4.26</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> IGKV6-17       329     9  2.74</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> IGLV1          329     9  2.74</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> IGKV14-111     329     8  2.43</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> IGKV1-117      329     7  2.13</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> IGKV13-84      329     7  2.13</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> IGKV1-135      329     6  1.82</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> IGKV19-93      329     6  1.82</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> IGKV3-4        329     6  1.82</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 93 more rows</span></span></span></code></pre></div>
<p>To perform gene usage calculations separately for cell clusters (or
samples), provide the meta.data column containing cluster labels to the
<code>cluster_col</code> argument. Here we see that the MD4 samples
almost exclusively use a single V segment (IGKV5-43), which is expected
since MD4 B cells are monoclonal.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_gene_usage.html">calc_gene_usage</a></span><span class="op">(</span></span>
<span>    data_cols   <span class="op">=</span> <span class="st">"v_gene"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    return_df   <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 618 × 6</span></span></span>
<span><span class="co">#&gt;    v_gene     sample n_cells  freq    pct shared</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> IGKV5-43   MD4-2       55    55 100    TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> IGKV5-43   MD4-3       54    51  94.4  TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> IGKV5-43   MD4-1       50    47  94    TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> IGKV10-96  BL6-1       55     6  10.9  TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> IGKV10-96  BL6-3       52     5   9.62 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> IGKV14-111 BL6-2       63     6   9.52 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> IGLV1      BL6-3       52     4   7.69 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> IGHV6-6    BL6-1       55     4   7.27 FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> IGKV6-17   BL6-2       63     4   6.35 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> IGKV8-24   BL6-2       63     4   6.35 FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 608 more rows</span></span></span></code></pre></div>
<p>To only perform calculations for a specific chain, use the
<code>chain</code> argument. In this example we are only returning
results for the IGK chain. Here we see some values in the ‘v_gene’
column labeled as ‘None’, this shows the number of cells that did not
have a V gene segment identified.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_gene_usage.html">calc_gene_usage</a></span><span class="op">(</span></span>
<span>    data_cols   <span class="op">=</span> <span class="st">"v_gene"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    chain       <span class="op">=</span> <span class="st">"IGK"</span>,</span>
<span>    return_df   <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 336 × 6</span></span></span>
<span><span class="co">#&gt;    v_gene     sample n_cells  freq    pct shared</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> IGKV5-43   MD4-2       55    55 100    TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> IGKV5-43   MD4-3       54    51  94.4  TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> IGKV5-43   MD4-1       50    47  94    TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> IGKV10-96  BL6-1       55     6  10.9  TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> IGKV10-96  BL6-3       52     5   9.62 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> None       BL6-3       52     5   9.62 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> IGKV14-111 BL6-2       63     6   9.52 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> IGKV6-17   BL6-2       63     4   6.35 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> IGKV8-24   BL6-2       63     4   6.35 FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> IGKV3-4    BL6-3       52     3   5.77 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 326 more rows</span></span></span></code></pre></div>
<p>If two columns are provided to the <code>data_cols</code> argument,
the number of cells containing each combination of genes is
returned.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_gene_usage.html">calc_gene_usage</a></span><span class="op">(</span></span>
<span>    data_cols   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"v_gene"</span>, <span class="st">"j_gene"</span><span class="op">)</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    return_df   <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,062 × 7</span></span></span>
<span><span class="co">#&gt;    v_gene     j_gene sample n_cells  freq    pct shared</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> IGKV5-43   IGKJ2  MD4-2       55    55 100    TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> IGKV5-43   IGKJ2  MD4-3       54    51  94.4  TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> IGKV5-43   IGKJ2  MD4-1       50    47  94    TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> IGKV14-111 IGKJ2  BL6-2       63     4   6.35 FALSE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> IGKV10-96  IGKJ1  BL6-3       52     3   5.77 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> IGLV1      IGLJ1  BL6-3       52     3   5.77 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> IGKV10-96  IGKJ2  BL6-1       55     3   5.45 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> IGKV10-96  IGKJ1  BL6-2       63     3   4.76 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> IGKV6-15   IGKJ5  BL6-3       52     2   3.85 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> IGKV10-96  IGKJ2  BL6-3       52     2   3.85 TRUE  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1,052 more rows</span></span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="plotting-gene-usage">Plotting gene usage<a class="anchor" aria-label="anchor" href="#plotting-gene-usage"></a>
</h2>
<p>The <code><a href="../reference/plot_gene_usage.html">plot_gene_usage()</a></code> function will summarize the
frequency of each gene segment. By default if a single column is passed
to the <code>data_cols</code> argument, a bargraph will be returned. The
number of top genes to include in the plot can be specified with the
<code>genes</code> argument.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_gene_usage.html">plot_gene_usage</a></span><span class="op">(</span></span>
<span>    data_cols   <span class="op">=</span> <span class="st">"v_gene"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    genes       <span class="op">=</span> <span class="fl">20</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="gene-usage_files/figure-html/unnamed-chunk-6-1.png" width="768"></p>
<p>By default, percentages are shown on the y-axis, to instead plot the
frequency, set the <code>units</code> argument to ‘frequency’.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_gene_usage.html">plot_gene_usage</a></span><span class="op">(</span></span>
<span>    data_cols   <span class="op">=</span> <span class="st">"v_gene"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    units       <span class="op">=</span> <span class="st">"frequency"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="gene-usage_files/figure-html/unnamed-chunk-7-1.png" width="768"></p>
<p>Plot colors can be adjusted using the <code>plot_colors</code>
argument. In addition, <code><a href="../reference/plot_gene_usage.html">plot_gene_usage()</a></code> returns a ggplot
object that can be modified with ggplot2 functions such as
<code><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">ggplot2::theme()</a></code>. Plots can be further adjusted by passing
aesthetic parameters directly to ggplot2, e.g. <code>alpha</code>,
<code>linetype</code>, <code>color</code>, etc.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_gene_usage.html">plot_gene_usage</a></span><span class="op">(</span></span>
<span>    data_cols   <span class="op">=</span> <span class="st">"v_gene"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>BL6 <span class="op">=</span> <span class="st">"#3288BD"</span>, MD4 <span class="op">=</span> <span class="st">"#D53E4F"</span><span class="op">)</span>,</span>
<span>    </span>
<span>    color <span class="op">=</span> <span class="st">"black"</span>,  <span class="co"># parameters to pass to ggplot2</span></span>
<span>    alpha <span class="op">=</span> <span class="fl">0.7</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="gene-usage_files/figure-html/unnamed-chunk-8-1.png" width="768"></p>
<p>If two columns are passed to the <code>data_cols</code> argument, a
heatmap will be generated summarizing the usage of different pairs of
segments. If a column is provided to the <code>cluster_col</code>
argument, a separate heatmap will be generated for each cluster.</p>
<p>In this example we are plotting the frequency that different heavy
chain V and J segments appear together.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_gene_usage.html">plot_gene_usage</a></span><span class="op">(</span></span>
<span>    data_cols   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"v_gene"</span>, <span class="st">"j_gene"</span><span class="op">)</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    chain       <span class="op">=</span> <span class="st">"IGH"</span>,</span>
<span>    genes       <span class="op">=</span> <span class="fl">15</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="gene-usage_files/figure-html/unnamed-chunk-9-1.png" width="768"></p>
<p>The paired gene usage for two chains can also be plotted using
<code><a href="../reference/plot_gene_pairs.html">plot_gene_pairs()</a></code>. In this example we are plotting the
frequency that different heavy and light chain V segments appear
together.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_gene_pairs.html">plot_gene_pairs</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"v_gene"</span>,</span>
<span>    chains      <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IGH"</span>, <span class="st">"IGK"</span><span class="op">)</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    genes       <span class="op">=</span> <span class="fl">12</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="gene-usage_files/figure-html/unnamed-chunk-10-1.png" width="768"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="circos-plot">Circos plot<a class="anchor" aria-label="anchor" href="#circos-plot"></a>
</h2>
<p>A circos plot can be created by setting the <code>method</code>
argument to ‘circos’. This plot will summarize the number of cells
containing different gene pairs, which is shown as the axis labels for
each sample. This requires the circlize package to be installed.</p>
<p>In this example, we are summarizing the segment usage for the entire
dataset (BL6 and MD4 cells combined). The <code>cluster_col</code>
argument can be used to create a separate plot for each sample. Labels
can be rotated to eliminate overlapping text using the
<code>rotate_labels</code> argument.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_gene_usage.html">plot_gene_usage</a></span><span class="op">(</span></span>
<span>    data_cols     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"v_gene"</span>, <span class="st">"j_gene"</span><span class="op">)</span>,</span>
<span>    method        <span class="op">=</span> <span class="st">"circos"</span>,</span>
<span>    genes         <span class="op">=</span> <span class="fl">6</span>,</span>
<span>    rotate_labels <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="gene-usage_files/figure-html/unnamed-chunk-11-1.png" width="576"></p>
<p>Plot colors can be modified using the <code>plot_colors</code>
argument, additional parameters can be passed directly to
<code><a href="https://rdrr.io/pkg/circlize/man/chordDiagram.html" class="external-link">circlize::chordDiagram()</a></code>. In this example we add a border
around the links and scale the plot so each sample is the same
width.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_gene_usage.html">plot_gene_usage</a></span><span class="op">(</span></span>
<span>    data_cols   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"v_gene"</span>, <span class="st">"j_gene"</span><span class="op">)</span>,</span>
<span>    method      <span class="op">=</span> <span class="st">"circos"</span>,</span>
<span>    genes       <span class="op">=</span> <span class="fl">8</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">10</span>, <span class="st">"Spectral"</span><span class="op">)</span>,</span>
<span>    </span>
<span>    link.border <span class="op">=</span> <span class="st">"black"</span>,  <span class="co"># parameters to pass to chordDiagram()</span></span>
<span>    scale       <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="gene-usage_files/figure-html/unnamed-chunk-12-1.png" width="576"></p>
<p>Gene segment usage can be plotted separately for cell clusters (or
samples) using the <code>cluster_col</code> argument. The number of rows
used to arrange plots can be modified using the <code>panel_nrow</code>
argument.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_gene_usage.html">plot_gene_usage</a></span><span class="op">(</span></span>
<span>    data_cols   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"v_gene"</span>, <span class="st">"j_gene"</span><span class="op">)</span>,</span>
<span>    method      <span class="op">=</span> <span class="st">"circos"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    genes       <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    panel_nrow  <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    scale       <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="gene-usage_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] RColorBrewer_1.1-3 ggplot2_3.4.3      SeuratObject_4.1.3</span></span>
<span><span class="co">#&gt; [4] Seurat_4.3.0.1     djvdj_0.1.0       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RcppAnnoy_0.0.21            splines_4.3.1              </span></span>
<span><span class="co">#&gt;   [3] later_1.3.1                 bitops_1.0-7               </span></span>
<span><span class="co">#&gt;   [5] tibble_3.2.1                polyclip_1.10-4            </span></span>
<span><span class="co">#&gt;   [7] lifecycle_1.0.3             doParallel_1.0.17          </span></span>
<span><span class="co">#&gt;   [9] rprojroot_2.0.3             globals_0.16.2             </span></span>
<span><span class="co">#&gt;  [11] lattice_0.21-8              vroom_1.6.3                </span></span>
<span><span class="co">#&gt;  [13] MASS_7.3-60                 backports_1.4.1            </span></span>
<span><span class="co">#&gt;  [15] magrittr_2.0.3              plotly_4.10.2              </span></span>
<span><span class="co">#&gt;  [17] sass_0.4.7                  rmarkdown_2.24             </span></span>
<span><span class="co">#&gt;  [19] jquerylib_0.1.4             yaml_2.3.7                 </span></span>
<span><span class="co">#&gt;  [21] httpuv_1.6.11               sctransform_0.3.5          </span></span>
<span><span class="co">#&gt;  [23] sp_2.0-0                    spatstat.sparse_3.0-2      </span></span>
<span><span class="co">#&gt;  [25] reticulate_1.31             cowplot_1.1.1              </span></span>
<span><span class="co">#&gt;  [27] pbapply_1.7-2               abind_1.4-5                </span></span>
<span><span class="co">#&gt;  [29] zlibbioc_1.46.0             Rtsne_0.16                 </span></span>
<span><span class="co">#&gt;  [31] GenomicRanges_1.52.0        purrr_1.0.2                </span></span>
<span><span class="co">#&gt;  [33] BiocGenerics_0.46.0         RCurl_1.98-1.12            </span></span>
<span><span class="co">#&gt;  [35] circlize_0.4.15             GenomeInfoDbData_1.2.10    </span></span>
<span><span class="co">#&gt;  [37] IRanges_2.34.1              S4Vectors_0.38.1           </span></span>
<span><span class="co">#&gt;  [39] ggrepel_0.9.3               irlba_2.3.5.1              </span></span>
<span><span class="co">#&gt;  [41] listenv_0.9.0               spatstat.utils_3.0-3       </span></span>
<span><span class="co">#&gt;  [43] goftest_1.2-3               spatstat.random_3.1-5      </span></span>
<span><span class="co">#&gt;  [45] fitdistrplus_1.1-11         parallelly_1.36.0          </span></span>
<span><span class="co">#&gt;  [47] pkgdown_2.0.7               leiden_0.4.3               </span></span>
<span><span class="co">#&gt;  [49] codetools_0.2-19            DelayedArray_0.26.7        </span></span>
<span><span class="co">#&gt;  [51] tidyselect_1.2.0            shape_1.4.6                </span></span>
<span><span class="co">#&gt;  [53] farver_2.1.1                matrixStats_1.0.0          </span></span>
<span><span class="co">#&gt;  [55] stats4_4.3.1                spatstat.explore_3.2-1     </span></span>
<span><span class="co">#&gt;  [57] jsonlite_1.8.7              GetoptLong_1.0.5           </span></span>
<span><span class="co">#&gt;  [59] ellipsis_0.3.2              ggtrace_0.2.0              </span></span>
<span><span class="co">#&gt;  [61] progressr_0.14.0            ggridges_0.5.4             </span></span>
<span><span class="co">#&gt;  [63] survival_3.5-5              iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [65] systemfonts_1.0.4           foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [67] tools_4.3.1                 ragg_1.2.5                 </span></span>
<span><span class="co">#&gt;  [69] ica_1.0-3                   Rcpp_1.0.11                </span></span>
<span><span class="co">#&gt;  [71] glue_1.6.2                  gridExtra_2.3              </span></span>
<span><span class="co">#&gt;  [73] xfun_0.40                   MatrixGenerics_1.12.3      </span></span>
<span><span class="co">#&gt;  [75] GenomeInfoDb_1.36.2         dplyr_1.1.2                </span></span>
<span><span class="co">#&gt;  [77] withr_2.5.0                 fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [79] boot_1.3-28.1               fansi_1.0.4                </span></span>
<span><span class="co">#&gt;  [81] digest_0.6.33               R6_2.5.1                   </span></span>
<span><span class="co">#&gt;  [83] mime_0.12                   textshaping_0.3.6          </span></span>
<span><span class="co">#&gt;  [85] colorspace_2.1-0            scattermore_1.2            </span></span>
<span><span class="co">#&gt;  [87] tensor_1.5                  spatstat.data_3.0-1        </span></span>
<span><span class="co">#&gt;  [89] utf8_1.2.3                  tidyr_1.3.0                </span></span>
<span><span class="co">#&gt;  [91] generics_0.1.3              data.table_1.14.8          </span></span>
<span><span class="co">#&gt;  [93] httr_1.4.7                  htmlwidgets_1.6.2          </span></span>
<span><span class="co">#&gt;  [95] S4Arrays_1.0.5              uwot_0.1.16                </span></span>
<span><span class="co">#&gt;  [97] pkgconfig_2.0.3             gtable_0.3.4               </span></span>
<span><span class="co">#&gt;  [99] ComplexHeatmap_2.16.0       lmtest_0.9-40              </span></span>
<span><span class="co">#&gt; [101] SingleCellExperiment_1.22.0 XVector_0.40.0             </span></span>
<span><span class="co">#&gt; [103] htmltools_0.5.6             clue_0.3-64                </span></span>
<span><span class="co">#&gt; [105] scales_1.2.1                Biobase_2.60.0             </span></span>
<span><span class="co">#&gt; [107] png_0.1-8                   knitr_1.43                 </span></span>
<span><span class="co">#&gt; [109] tzdb_0.4.0                  reshape2_1.4.4             </span></span>
<span><span class="co">#&gt; [111] rjson_0.2.21                nlme_3.1-162               </span></span>
<span><span class="co">#&gt; [113] cachem_1.0.8                zoo_1.8-12                 </span></span>
<span><span class="co">#&gt; [115] GlobalOptions_0.1.2         stringr_1.5.0              </span></span>
<span><span class="co">#&gt; [117] KernSmooth_2.23-21          parallel_4.3.1             </span></span>
<span><span class="co">#&gt; [119] miniUI_0.1.1.1              desc_1.4.2                 </span></span>
<span><span class="co">#&gt; [121] pillar_1.9.0                grid_4.3.1                 </span></span>
<span><span class="co">#&gt; [123] vctrs_0.6.3                 RANN_2.6.1                 </span></span>
<span><span class="co">#&gt; [125] promises_1.2.1              xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [127] cluster_2.1.4               evaluate_0.21              </span></span>
<span><span class="co">#&gt; [129] readr_2.1.4                 cli_3.6.1                  </span></span>
<span><span class="co">#&gt; [131] compiler_4.3.1              rlang_1.1.1                </span></span>
<span><span class="co">#&gt; [133] crayon_1.5.2                abdiv_0.2.0                </span></span>
<span><span class="co">#&gt; [135] future.apply_1.11.0         labeling_0.4.2             </span></span>
<span><span class="co">#&gt; [137] plyr_1.8.8                  fs_1.6.3                   </span></span>
<span><span class="co">#&gt; [139] stringi_1.7.12              viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt; [141] deldir_1.0-9                munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [143] lazyeval_0.2.2              spatstat.geom_3.2-4        </span></span>
<span><span class="co">#&gt; [145] Matrix_1.5-4.1              hms_1.1.3                  </span></span>
<span><span class="co">#&gt; [147] patchwork_1.1.3             bit64_4.0.5                </span></span>
<span><span class="co">#&gt; [149] future_1.33.0               shiny_1.7.5                </span></span>
<span><span class="co">#&gt; [151] highr_0.10                  SummarizedExperiment_1.30.2</span></span>
<span><span class="co">#&gt; [153] ROCR_1.0-11                 igraph_1.5.1               </span></span>
<span><span class="co">#&gt; [155] broom_1.0.5                 memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [157] bslib_0.5.1                 bit_4.0.5</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
