<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="djvdj">
<title>Data Import â€¢ djvdj</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.6/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.6/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Data Import">
<meta property="og:description" content="djvdj">
<meta property="og:image" content="https://rnabioco.github.io/djvdj/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">djvdj</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Data import and manipulation</h6>
    <a class="dropdown-item" href="../articles/import.html">Data Import</a>
    <a class="dropdown-item" href="../articles/mutate.html">Data Manipulation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>VDJ analysis</h6>
    <a class="dropdown-item" href="../articles/frequency.html">Clonotype Frequencies</a>
    <a class="dropdown-item" href="../articles/diversity.html">Repertoire Diversity</a>
    <a class="dropdown-item" href="../articles/similarity.html">Repertoire Similarity</a>
    <a class="dropdown-item" href="../articles/gene-usage.html">Gene Usage</a>
    <a class="dropdown-item" href="../articles/clustering.html">CDR3 Clustering</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualization</h6>
    <a class="dropdown-item" href="../articles/plotting.html">Plotting</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/djvdj/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Data Import</h1>
            
            <h4 data-toc-skip class="date">May 08 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/djvdj/blob/HEAD/vignettes/import.Rmd" class="external-link"><code>vignettes/import.Rmd</code></a></small>
      <div class="d-none name"><code>import.Rmd</code></div>
    </div>

    
    
<p><br></p>
<p>This vignette provides detailed examples for loading gene expression
and V(D)J data into a single object. For the examples shown below, we
use data for splenocytes from BL6 and MD4 mice collected using the 10X
Genomics scRNA-seq platform. MD4 B cells are monoclonal and specifically
bind hen egg lysozyme.</p>
<p><br></p>
<div class="section level2">
<h2 id="basic-usage">Basic usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h2>
<p>To load and parse V(D)J data, FASTQ files must first be processed
using <a href="https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/what-is-cell-ranger" class="external-link">Cell
Ranger</a>. The data used for this vignette was processed using Cell
Ranger v7.0. Cell Ranger generates several files that djvdj uses for the
downstream analysis. This includes the
outs/filtered_contig_annotations.csv file, which contains basic
information about each chain and is required for the
<code><a href="../reference/import_vdj.html">import_vdj()</a></code> function.</p>
<p>The simplest use case for <code><a href="../reference/import_vdj.html">import_vdj()</a></code> involves creating
an object using data from a single run.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/djvdj/">djvdj</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/splen"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create Seurat object</span></span>
<span><span class="va">so</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"BL6_GEX/filtered_feature_bc_matrix"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add V(D)J data to object</span></span>
<span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span>vdj_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"BL6_BCR"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><code><a href="../reference/import_vdj.html">import_vdj()</a></code> adds a variety of per-chain metrics to the
object meta.data. Information for each chain identified for the cell is
separated by a semicolon. The separator used for storing and parsing
per-chain V(D)J data can be specified using the <code>sep</code>
argument included for most djvdj functions. <code>NA</code>s will be
included for cells that lack V(D)J data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="st">"meta.data"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">n_chains</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt;                       orig.ident nCount_RNA nFeature_RNA  clonotype_id</span></span>
<span><span class="co">#&gt; ACACCAAAGAATTGTG-1 SeuratProject         38           21 clonotype2492</span></span>
<span><span class="co">#&gt; ACACCGGCACAAGTAA-1 SeuratProject         80           30 clonotype3124</span></span>
<span><span class="co">#&gt;                    exact_subclonotype_id  chains n_chains</span></span>
<span><span class="co">#&gt; ACACCAAAGAATTGTG-1                     1 IGH;IGK        2</span></span>
<span><span class="co">#&gt; ACACCGGCACAAGTAA-1                     1 IGH;IGK        2</span></span>
<span><span class="co">#&gt;                                          cdr3</span></span>
<span><span class="co">#&gt; ACACCAAAGAATTGTG-1    CAHGSRDFDVW;CWQGTHFPQTF</span></span>
<span><span class="co">#&gt; ACACCGGCACAAGTAA-1 CARHEGYYEAMDYW;CQQGNTLPLTF</span></span>
<span><span class="co">#&gt;                                                                                         cdr3_nt</span></span>
<span><span class="co">#&gt; ACACCAAAGAATTGTG-1          TGTGCTCACGGTAGTCGAGACTTCGATGTCTGG;TGCTGGCAAGGTACACATTTTCCTCAGACGTTC</span></span>
<span><span class="co">#&gt; ACACCGGCACAAGTAA-1 TGTGCAAGACATGAGGGGTACTACGAGGCTATGGACTACTGG;TGCCAACAGGGTAATACGCTTCCTCTCACGTTC</span></span>
<span><span class="co">#&gt;                    cdr3_length cdr3_nt_length             v_gene</span></span>
<span><span class="co">#&gt; ACACCAAAGAATTGTG-1       11;11          33;33 IGHV1-75;IGKV1-135</span></span>
<span><span class="co">#&gt; ACACCGGCACAAGTAA-1       14;11          42;33 IGHV5-15;IGKV10-96</span></span>
<span><span class="co">#&gt;                       d_gene      j_gene    c_gene isotype  reads umis</span></span>
<span><span class="co">#&gt; ACACCAAAGAATTGTG-1 None;None IGHJ1;IGKJ1 IGHM;IGKC    IGHM 20;208 2;11</span></span>
<span><span class="co">#&gt; ACACCGGCACAAGTAA-1 None;None IGHJ4;IGKJ5 IGHM;IGKC    IGHM 80;174 6;12</span></span>
<span><span class="co">#&gt;                    productive full_length paired</span></span>
<span><span class="co">#&gt; ACACCAAAGAATTGTG-1  TRUE;TRUE   TRUE;TRUE   TRUE</span></span>
<span><span class="co">#&gt; ACACCGGCACAAGTAA-1  TRUE;TRUE   TRUE;TRUE   TRUE</span></span></code></pre></div>
<p>To modify/filter/plot per-chain data, djvdj provides a range of
functions that make it easy to parse and visualize this information. For
detailed examples refer to the vignettes and documentation for the
following functions: <code><a href="../reference/filter_vdj.html">filter_vdj()</a></code>,
<code><a href="../reference/mutate_vdj.html">mutate_vdj()</a></code>, <code><a href="../reference/summarize_vdj.html">summarize_vdj()</a></code>,
<code><a href="../reference/plot_numerical.html">plot_histogram()</a></code>, <code><a href="../reference/plot_scatter.html">plot_scatter()</a></code>.</p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="loading-multiple-runs">Loading multiple runs<a class="anchor" aria-label="anchor" href="#loading-multiple-runs"></a>
</h2>
<p>When combining gene expression and V(D)J data from multiple runs into
the same Seurat object, we must ensure that <code><a href="../reference/import_vdj.html">import_vdj()</a></code> is
able to match the cell barcodes from the two data types. The easiest way
to do this is to load the gene expression and V(D)J samples in the same
order.</p>
<p>If the V(D)J samples are not loaded in the same order as the gene
expression data, the cell barcodes will not match and you will receive
an error.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load GEX data</span></span>
<span><span class="va">gex_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"BL6_GEX/filtered_feature_bc_matrix"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"MD4_GEX/filtered_feature_bc_matrix"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">so</span> <span class="op">&lt;-</span> <span class="va">gex_dirs</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load BCR data</span></span>
<span><span class="co"># note that the BL6 and MD4 paths are in the same order as the gene</span></span>
<span><span class="co"># expression data</span></span>
<span><span class="va">vdj_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"BL6_BCR"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"MD4_BCR"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span>vdj_dir <span class="op">=</span> <span class="va">vdj_dirs</span><span class="op">)</span></span></code></pre></div>
<p>Another way to ensure the cell barcodes from the gene expression and
V(D)J data are able to be matched is to specify cell barcode prefixes
for each sample. For both the <code><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Seurat::Read10X()</a></code> and
<code><a href="../reference/import_vdj.html">import_vdj()</a></code> functions this can be done by passing a named
vector. In this scenario, the names will be added as prefixes for the
cell barcodes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load GEX data</span></span>
<span><span class="va">gex_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  BL6 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"BL6_GEX/filtered_feature_bc_matrix"</span><span class="op">)</span>,</span>
<span>  MD4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"MD4_GEX/filtered_feature_bc_matrix"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">so</span> <span class="op">&lt;-</span> <span class="va">gex_dirs</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load BCR data</span></span>
<span><span class="co"># note that the samples are not in the same order as the gene expression data,</span></span>
<span><span class="co"># but this is okay since cell prefixes are provided as names for the</span></span>
<span><span class="co"># input vector</span></span>
<span><span class="va">vdj_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  MD4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"MD4_BCR"</span><span class="op">)</span>,</span>
<span>  BL6 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"BL6_BCR"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span>vdj_dir <span class="op">=</span> <span class="va">vdj_dirs</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="defining-clonotypes">Defining clonotypes<a class="anchor" aria-label="anchor" href="#defining-clonotypes"></a>
</h2>
<p>When results from multiple Cell Ranger runs are added to the object,
the clonotype IDs will not match, i.e.Â clonotype1 will not be the same
for all samples. To allow clonotypes to be directly compared between
multiple samples, the clonotypes can be recalculated using the
<code>define_clonotypes</code> argument. This argument will assign new
clonotype IDs using information available for each chain. There are
three options to specify how this step is performed:</p>
<ul>
<li>â€˜cdr3aaâ€™: use only the CDR3 amino acid sequence</li>
<li>â€˜cdr3ntâ€™: use only the CDR3 nucleotide sequence</li>
<li>â€˜cdr3_geneâ€™: use both the CDR3 nucleotide sequence and the
combination of all V(D)J genes identified for the cell</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span></span>
<span>    vdj_dir <span class="op">=</span> <span class="va">vdj_dirs</span>,</span>
<span>    define_clonotypes <span class="op">=</span> <span class="st">"cdr3_gene"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>The clonotype IDs can also be adjusted for multiple samples using the
<a href="https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/using/aggr" class="external-link">aggregate</a>
function available with Cell Ranger. To load aggregated output files
just pass the <code>cellranger aggr</code> output directory to the
<code>aggr_dir</code> argument. To correctly match cell barcodes from
aggregated V(D)J data with the gene expression data, the gene expression
samples must be loaded into the object in the same order the samples
were listed in the <code>cellranger aggr</code> <a href="https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/using/aggr#vdj_aggr" class="external-link">config
file</a>.</p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="filtering-chains">Filtering chains<a class="anchor" aria-label="anchor" href="#filtering-chains"></a>
</h2>
<p>The <code><a href="../reference/import_vdj.html">import_vdj()</a></code> function has several arguments that can
be used to perform basic filtering for each chain. The
<code>filter_chains</code> argument will only include chains with at
least one productive and full length contig. However, it should be noted
that in recent versions of Cell Ranger the output files are already
filtered based on this criteria, so this argument is only relevant when
loading data processed with earlier versions such as Cell Ranger
v3.0.</p>
<p>The <code>filter_paired</code> argument will only include clonotypes
with paired chains. For TCR data, this means each clonotype must have at
least one TRA and TRB chain. For BCR data each clonotype must have at
least one IGH chain and at least one IGK or IGL chain. It should be
noted that if a clonotype has multiple chains of the same type, it will
still be included, e.g.Â TRA;TRB;TRB or IGH;IGK;IGK will still be
included. Clonotypes that include more than two chains can be filtered
using <code><a href="../reference/filter_vdj.html">filter_vdj()</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vdj_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  MD4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"MD4_BCR"</span><span class="op">)</span>,</span>
<span>  BL6 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"BL6_BCR"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span></span>
<span>    vdj_dir <span class="op">=</span> <span class="va">vdj_dirs</span>,</span>
<span>    filter_chains <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    filter_paired <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># To only include clonotypes with exactly 2 chains</span></span>
<span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so_vdj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/filter_vdj.html">filter_vdj</a></span><span class="op">(</span><span class="va">n_chains</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="loading-mutation-information">Loading mutation information<a class="anchor" aria-label="anchor" href="#loading-mutation-information"></a>
</h2>
<p>Mutation information for each chain can be parsed using two
additional output files from Cell Ranger:</p>
<ul>
<li>outs/concat_ref.bam: this file contains alignment information used
to quantify insertions, deletions, and mismatches for each chain.</li>
<li>outs/airr_rearrangement.tsv: this file contains coordinates for each
V(D)J gene segment and is used to quantify mutations for each segment
and/or junction.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vdj_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  MD4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"MD4_BCR"</span><span class="op">)</span>,</span>
<span>  BL6 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"BL6_BCR"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span></span>
<span>    vdj_dir <span class="op">=</span> <span class="va">vdj_dirs</span>,</span>
<span>    include_mutations <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>The additional columns added to the meta.data will include the number
of insertions, deletions, and mismatches (ending in â€˜insâ€™, â€˜delâ€™, or
â€˜misâ€™) for each V(D)J segment (prefixed with â€˜vâ€™, â€˜dâ€™, â€˜jâ€™, or â€˜câ€™).
Columns containing junction information will be prefixed with either
â€˜vdâ€™ or â€˜djâ€™. Columns ending in â€˜freqâ€™ show the event frequency which is
calculated as the number of events divided by the length of the
region.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="st">"meta.data"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">n_chains</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt;                        orig.ident nCount_RNA nFeature_RNA  clonotype_id</span></span>
<span><span class="co">#&gt; BL6_ACACCAAAGAATTGTG-1        BL6         38           21 clonotype2492</span></span>
<span><span class="co">#&gt; BL6_ACACCGGCACAAGTAA-1        BL6         80           30 clonotype3124</span></span>
<span><span class="co">#&gt;                        exact_subclonotype_id  chains n_chains</span></span>
<span><span class="co">#&gt; BL6_ACACCAAAGAATTGTG-1                     1 IGH;IGK        2</span></span>
<span><span class="co">#&gt; BL6_ACACCGGCACAAGTAA-1                     1 IGH;IGK        2</span></span>
<span><span class="co">#&gt;                                              cdr3</span></span>
<span><span class="co">#&gt; BL6_ACACCAAAGAATTGTG-1    CAHGSRDFDVW;CWQGTHFPQTF</span></span>
<span><span class="co">#&gt; BL6_ACACCGGCACAAGTAA-1 CARHEGYYEAMDYW;CQQGNTLPLTF</span></span>
<span><span class="co">#&gt;                                                                                             cdr3_nt</span></span>
<span><span class="co">#&gt; BL6_ACACCAAAGAATTGTG-1          TGTGCTCACGGTAGTCGAGACTTCGATGTCTGG;TGCTGGCAAGGTACACATTTTCCTCAGACGTTC</span></span>
<span><span class="co">#&gt; BL6_ACACCGGCACAAGTAA-1 TGTGCAAGACATGAGGGGTACTACGAGGCTATGGACTACTGG;TGCCAACAGGGTAATACGCTTCCTCTCACGTTC</span></span>
<span><span class="co">#&gt;                        cdr3_length cdr3_nt_length             v_gene</span></span>
<span><span class="co">#&gt; BL6_ACACCAAAGAATTGTG-1       11;11          33;33 IGHV1-75;IGKV1-135</span></span>
<span><span class="co">#&gt; BL6_ACACCGGCACAAGTAA-1       14;11          42;33 IGHV5-15;IGKV10-96</span></span>
<span><span class="co">#&gt;                           d_gene      j_gene    c_gene isotype   reads</span></span>
<span><span class="co">#&gt; BL6_ACACCAAAGAATTGTG-1 None;None IGHJ1;IGKJ1 IGHM;IGKC    IGHM  40;416</span></span>
<span><span class="co">#&gt; BL6_ACACCGGCACAAGTAA-1 None;None IGHJ4;IGKJ5 IGHM;IGKC    IGHM 160;348</span></span>
<span><span class="co">#&gt;                         umis productive full_length paired v_ins v_del</span></span>
<span><span class="co">#&gt; BL6_ACACCAAAGAATTGTG-1  4;22  TRUE;TRUE   TRUE;TRUE   TRUE   2;0   0;0</span></span>
<span><span class="co">#&gt; BL6_ACACCGGCACAAGTAA-1 12;24  TRUE;TRUE   TRUE;TRUE   TRUE   0;0   0;3</span></span>
<span><span class="co">#&gt;                        v_mis d_ins d_del d_mis j_ins j_del j_mis c_ins</span></span>
<span><span class="co">#&gt; BL6_ACACCAAAGAATTGTG-1   1;0   0;0   0;0   0;0   0;0   0;3   0;0   0;0</span></span>
<span><span class="co">#&gt; BL6_ACACCGGCACAAGTAA-1   0;0   0;0   0;0   0;0   0;0   0;3   3;0   0;0</span></span>
<span><span class="co">#&gt;                        c_del c_mis all_ins all_del all_mis vd_ins vd_del</span></span>
<span><span class="co">#&gt; BL6_ACACCAAAGAATTGTG-1   0;0   0;0     2;0     0;3     1;0    0;0    0;0</span></span>
<span><span class="co">#&gt; BL6_ACACCGGCACAAGTAA-1   0;0   0;0     0;0     0;6     3;0    0;0    0;0</span></span>
<span><span class="co">#&gt;                        dj_ins dj_del v_mis_freq d_mis_freq j_mis_freq</span></span>
<span><span class="co">#&gt; BL6_ACACCAAAGAATTGTG-1    0;0    0;0 0.002849;0        0;0        0;0</span></span>
<span><span class="co">#&gt; BL6_ACACCGGCACAAGTAA-1    0;0    0;0        0;0        0;0 0.055556;0</span></span>
<span><span class="co">#&gt;                        c_mis_freq all_mis_freq</span></span>
<span><span class="co">#&gt; BL6_ACACCAAAGAATTGTG-1        0;0   0.002525;0</span></span>
<span><span class="co">#&gt; BL6_ACACCGGCACAAGTAA-1        0;0   0.007371;0</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="loading-additional-sequence-information">Loading additional sequence information<a class="anchor" aria-label="anchor" href="#loading-additional-sequence-information"></a>
</h2>
<p>By default the only sequence information loaded by
<code><a href="../reference/import_vdj.html">import_vdj()</a></code> will be for the CDR3 region. Newer versions of
Cell Ranger will include additional sequences in the
filtered_contig_annotations.csv file. This includes the FWR1, CDR1,
FWR2, CDR2, FWR3, and FWR4 regions. These additional sequences can be
loaded using the <code>data_cols</code> argument.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span></span>
<span>    vdj_dir   <span class="op">=</span> <span class="va">vdj_dirs</span>,</span>
<span>    data_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cdr1"</span>, <span class="st">"cdr1_nt"</span>, <span class="st">"cdr2"</span>, <span class="st">"cdr2_nt"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">so_vdj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="st">"meta.data"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt;                        orig.ident nCount_RNA nFeature_RNA clonotype_id</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1        BL6        202           25         &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1        BL6         42           20   clonotype7</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1        BL6         32           19  clonotype32</span></span>
<span><span class="co">#&gt;                        exact_subclonotype_id chains n_chains</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1                    NA   &lt;NA&gt;       NA</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1                     1    IGK        1</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1                     1    IGK        1</span></span>
<span><span class="co">#&gt;                                    cdr1</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1             &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1 RSSQSIVHSNGNTYLE</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1      KASQNVGTNVA</span></span>
<span><span class="co">#&gt;                                                                 cdr1_nt</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1                                             &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1 AGATCTAGTCAGAGCATTGTACATAGTAATGGAAACACCTATTTAGAA</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1                AAGGCCAGTCAGAATGTGGGTACTAATGTAGCC</span></span>
<span><span class="co">#&gt;                           cdr2               cdr2_nt        cdr3</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1    &lt;NA&gt;                  &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1 KVSNRFS AAAGTTTCCAACCGATTTTCT CFQGSHVPWTF</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1 SASYRYS TCGGCATCCTACCGGTACAGT CQQYNSYPLTF</span></span>
<span><span class="co">#&gt;                                                  cdr3_nt cdr1_length</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1                              &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1 TGCTTTCAAGGTTCACATGTTCCGTGGACGTTC          16</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1 TGTCAGCAATATAACAGCTATCCTCTCACGTTC          11</span></span>
<span><span class="co">#&gt;                        cdr1_nt_length cdr2_length cdr2_nt_length</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1           &lt;NA&gt;        &lt;NA&gt;           &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1             48           7             21</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1             33           7             21</span></span>
<span><span class="co">#&gt;                        cdr3_length cdr3_nt_length    v_gene d_gene j_gene</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1        &lt;NA&gt;           &lt;NA&gt;      &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1          11             33 IGKV1-117   None  IGKJ1</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1          11             33  IGKV6-15   None  IGKJ5</span></span>
<span><span class="co">#&gt;                        c_gene isotype reads umis productive full_length</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1   &lt;NA&gt;    &lt;NA&gt;  &lt;NA&gt; &lt;NA&gt;       &lt;NA&gt;        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1   IGKC    None   352   21       TRUE        TRUE</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1   IGKC    None   342   18       TRUE        TRUE</span></span>
<span><span class="co">#&gt;                        paired</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1     NA</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1  FALSE</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1  FALSE</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="loading-tcr-and-bcr-data">Loading TCR and BCR data<a class="anchor" aria-label="anchor" href="#loading-tcr-and-bcr-data"></a>
</h2>
<p>To add both BCR and TCR data to the object, run
<code><a href="../reference/import_vdj.html">import_vdj()</a></code> separately for each data type. To distinguish
between columns containing BCR or TCR data, use the <code>prefix</code>
argument to add unique column names.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bcr_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  MD4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"MD4_BCR"</span><span class="op">)</span>,</span>
<span>  BL6 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"BL6_BCR"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">tcr_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  MD4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"MD4_TCR"</span><span class="op">)</span>,</span>
<span>  BL6 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"BL6_TCR"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span><span class="va">bcr_dirs</span>, prefix <span class="op">=</span> <span class="st">"bcr_"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span><span class="va">tcr_dirs</span>, prefix <span class="op">=</span> <span class="st">"tcr_"</span><span class="op">)</span></span></code></pre></div>
<p>This results in two sets of new columns being added to the meta.data.
When performing downstream analysis using other djvdj functions, be sure
to specify the correct columns, i.e.Â â€˜bcr_clonotype_idâ€™ or
â€˜tcr_clonotype_idâ€™.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/methods/slot.html" class="external-link">slot</a></span><span class="op">(</span><span class="st">"meta.data"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt;                        orig.ident nCount_RNA nFeature_RNA</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1        BL6        202           25</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1        BL6         42           20</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1        BL6         32           19</span></span>
<span><span class="co">#&gt;                        bcr_clonotype_id bcr_exact_subclonotype_id</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1             &lt;NA&gt;                        NA</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1       clonotype7                         1</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1      clonotype32                         1</span></span>
<span><span class="co">#&gt;                        bcr_chains bcr_n_chains    bcr_cdr3</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1       &lt;NA&gt;           NA        &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1        IGK            1 CFQGSHVPWTF</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1        IGK            1 CQQYNSYPLTF</span></span>
<span><span class="co">#&gt;                                              bcr_cdr3_nt bcr_cdr3_length</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1                              &lt;NA&gt;            &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1 TGCTTTCAAGGTTCACATGTTCCGTGGACGTTC              11</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1 TGTCAGCAATATAACAGCTATCCTCTCACGTTC              11</span></span>
<span><span class="co">#&gt;                        bcr_cdr3_nt_length bcr_v_gene bcr_d_gene</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1               &lt;NA&gt;       &lt;NA&gt;       &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1                 33  IGKV1-117       None</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1                 33   IGKV6-15       None</span></span>
<span><span class="co">#&gt;                        bcr_j_gene bcr_c_gene bcr_isotype bcr_reads</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1       &lt;NA&gt;       &lt;NA&gt;        &lt;NA&gt;      &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1      IGKJ1       IGKC        None       352</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1      IGKJ5       IGKC        None       342</span></span>
<span><span class="co">#&gt;                        bcr_umis bcr_productive bcr_full_length bcr_paired</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1     &lt;NA&gt;           &lt;NA&gt;            &lt;NA&gt;         NA</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1       21           TRUE            TRUE      FALSE</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1       18           TRUE            TRUE      FALSE</span></span>
<span><span class="co">#&gt;                        tcr_clonotype_id tcr_chains tcr_n_chains tcr_cdr3</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1             &lt;NA&gt;       &lt;NA&gt;           NA     &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1             &lt;NA&gt;       &lt;NA&gt;           NA     &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1             &lt;NA&gt;       &lt;NA&gt;           NA     &lt;NA&gt;</span></span>
<span><span class="co">#&gt;                        tcr_cdr3_nt tcr_cdr3_length tcr_cdr3_nt_length</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1        &lt;NA&gt;            &lt;NA&gt;               &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1        &lt;NA&gt;            &lt;NA&gt;               &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1        &lt;NA&gt;            &lt;NA&gt;               &lt;NA&gt;</span></span>
<span><span class="co">#&gt;                        tcr_v_gene tcr_d_gene tcr_j_gene tcr_c_gene</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1       &lt;NA&gt;       &lt;NA&gt;       &lt;NA&gt;       &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1       &lt;NA&gt;       &lt;NA&gt;       &lt;NA&gt;       &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1       &lt;NA&gt;       &lt;NA&gt;       &lt;NA&gt;       &lt;NA&gt;</span></span>
<span><span class="co">#&gt;                        tcr_reads tcr_umis tcr_productive tcr_full_length</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1      &lt;NA&gt;     &lt;NA&gt;           &lt;NA&gt;            &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1      &lt;NA&gt;     &lt;NA&gt;           &lt;NA&gt;            &lt;NA&gt;</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1      &lt;NA&gt;     &lt;NA&gt;           &lt;NA&gt;            &lt;NA&gt;</span></span>
<span><span class="co">#&gt;                        tcr_paired</span></span>
<span><span class="co">#&gt; BL6_AAACGGGGTTCTGTTT-1         NA</span></span>
<span><span class="co">#&gt; BL6_AAAGATGCAACAACCT-1         NA</span></span>
<span><span class="co">#&gt; BL6_AACGTTGCACGACTCG-1         NA</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="other-object-types">Other object types<a class="anchor" aria-label="anchor" href="#other-object-types"></a>
</h2>
<p>In addition to Seurat objects, djvdj also works with
SingleCellExperiment objects and with data.frames. If no input object is
provided to <code><a href="../reference/import_vdj.html">import_vdj()</a></code>, a data.frame containing V(D)J
information will be returned. This data.frame can be used with other
djvdj functions to perform further downstream analysis.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vdj_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  MD4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"MD4_BCR"</span><span class="op">)</span>,</span>
<span>  BL6 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"BL6_BCR"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># This will load V(D)J data and return a data.frame</span></span>
<span><span class="va">df_vdj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span>vdj_dir <span class="op">=</span> <span class="va">vdj_dirs</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] dplyr_1.1.2        SeuratObject_4.1.3 Seurat_4.3.0      </span></span>
<span><span class="co">#&gt; [4] djvdj_0.1.0       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RcppAnnoy_0.0.20            splines_4.3.0              </span></span>
<span><span class="co">#&gt;   [3] later_1.3.1                 bitops_1.0-7               </span></span>
<span><span class="co">#&gt;   [5] tibble_3.2.1                polyclip_1.10-4            </span></span>
<span><span class="co">#&gt;   [7] lifecycle_1.0.3             doParallel_1.0.17          </span></span>
<span><span class="co">#&gt;   [9] rprojroot_2.0.3             globals_0.16.2             </span></span>
<span><span class="co">#&gt;  [11] lattice_0.21-8              vroom_1.6.3                </span></span>
<span><span class="co">#&gt;  [13] MASS_7.3-58.4               backports_1.4.1            </span></span>
<span><span class="co">#&gt;  [15] magrittr_2.0.3              plotly_4.10.1              </span></span>
<span><span class="co">#&gt;  [17] sass_0.4.6                  rmarkdown_2.21             </span></span>
<span><span class="co">#&gt;  [19] jquerylib_0.1.4             yaml_2.3.7                 </span></span>
<span><span class="co">#&gt;  [21] httpuv_1.6.9                sctransform_0.3.5          </span></span>
<span><span class="co">#&gt;  [23] sp_1.6-0                    spatstat.sparse_3.0-1      </span></span>
<span><span class="co">#&gt;  [25] reticulate_1.28             cowplot_1.1.1              </span></span>
<span><span class="co">#&gt;  [27] pbapply_1.7-0               RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;  [29] abind_1.4-5                 zlibbioc_1.46.0            </span></span>
<span><span class="co">#&gt;  [31] Rtsne_0.16                  GenomicRanges_1.52.0       </span></span>
<span><span class="co">#&gt;  [33] purrr_1.0.1                 BiocGenerics_0.46.0        </span></span>
<span><span class="co">#&gt;  [35] RCurl_1.98-1.12             circlize_0.4.15            </span></span>
<span><span class="co">#&gt;  [37] GenomeInfoDbData_1.2.10     IRanges_2.34.0             </span></span>
<span><span class="co">#&gt;  [39] S4Vectors_0.38.1            ggrepel_0.9.3              </span></span>
<span><span class="co">#&gt;  [41] irlba_2.3.5.1               listenv_0.9.0              </span></span>
<span><span class="co">#&gt;  [43] spatstat.utils_3.0-2        goftest_1.2-3              </span></span>
<span><span class="co">#&gt;  [45] spatstat.random_3.1-4       fitdistrplus_1.1-11        </span></span>
<span><span class="co">#&gt;  [47] parallelly_1.35.0           pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt;  [49] leiden_0.4.3                codetools_0.2-19           </span></span>
<span><span class="co">#&gt;  [51] DelayedArray_0.26.2         tidyselect_1.2.0           </span></span>
<span><span class="co">#&gt;  [53] shape_1.4.6                 matrixStats_0.63.0         </span></span>
<span><span class="co">#&gt;  [55] stats4_4.3.0                spatstat.explore_3.1-0     </span></span>
<span><span class="co">#&gt;  [57] jsonlite_1.8.4              GetoptLong_1.0.5           </span></span>
<span><span class="co">#&gt;  [59] ellipsis_0.3.2              ggtrace_0.2.0              </span></span>
<span><span class="co">#&gt;  [61] progressr_0.13.0            ggridges_0.5.4             </span></span>
<span><span class="co">#&gt;  [63] survival_3.5-5              iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [65] systemfonts_1.0.4           foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [67] tools_4.3.0                 ragg_1.2.5                 </span></span>
<span><span class="co">#&gt;  [69] ica_1.0-3                   Rcpp_1.0.10                </span></span>
<span><span class="co">#&gt;  [71] glue_1.6.2                  gridExtra_2.3              </span></span>
<span><span class="co">#&gt;  [73] xfun_0.39                   MatrixGenerics_1.12.0      </span></span>
<span><span class="co">#&gt;  [75] GenomeInfoDb_1.36.0         withr_2.5.0                </span></span>
<span><span class="co">#&gt;  [77] fastmap_1.1.1               boot_1.3-28.1              </span></span>
<span><span class="co">#&gt;  [79] fansi_1.0.4                 digest_0.6.31              </span></span>
<span><span class="co">#&gt;  [81] R6_2.5.1                    mime_0.12                  </span></span>
<span><span class="co">#&gt;  [83] textshaping_0.3.6           colorspace_2.1-0           </span></span>
<span><span class="co">#&gt;  [85] scattermore_1.0             tensor_1.5                 </span></span>
<span><span class="co">#&gt;  [87] spatstat.data_3.0-1         utf8_1.2.3                 </span></span>
<span><span class="co">#&gt;  [89] tidyr_1.3.0                 generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [91] data.table_1.14.8           httr_1.4.5                 </span></span>
<span><span class="co">#&gt;  [93] htmlwidgets_1.6.2           S4Arrays_1.0.1             </span></span>
<span><span class="co">#&gt;  [95] uwot_0.1.14                 pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [97] gtable_0.3.3                ComplexHeatmap_2.16.0      </span></span>
<span><span class="co">#&gt;  [99] lmtest_0.9-40               SingleCellExperiment_1.22.0</span></span>
<span><span class="co">#&gt; [101] XVector_0.40.0              htmltools_0.5.5            </span></span>
<span><span class="co">#&gt; [103] clue_0.3-64                 scales_1.2.1               </span></span>
<span><span class="co">#&gt; [105] Biobase_2.60.0              png_0.1-8                  </span></span>
<span><span class="co">#&gt; [107] knitr_1.42                  tzdb_0.3.0                 </span></span>
<span><span class="co">#&gt; [109] reshape2_1.4.4              rjson_0.2.21               </span></span>
<span><span class="co">#&gt; [111] nlme_3.1-162                cachem_1.0.8               </span></span>
<span><span class="co">#&gt; [113] zoo_1.8-12                  GlobalOptions_0.1.2        </span></span>
<span><span class="co">#&gt; [115] stringr_1.5.0               KernSmooth_2.23-20         </span></span>
<span><span class="co">#&gt; [117] parallel_4.3.0              miniUI_0.1.1.1             </span></span>
<span><span class="co">#&gt; [119] desc_1.4.2                  pillar_1.9.0               </span></span>
<span><span class="co">#&gt; [121] grid_4.3.0                  vctrs_0.6.2                </span></span>
<span><span class="co">#&gt; [123] RANN_2.6.1                  promises_1.2.0.1           </span></span>
<span><span class="co">#&gt; [125] xtable_1.8-4                cluster_2.1.4              </span></span>
<span><span class="co">#&gt; [127] evaluate_0.20               readr_2.1.4                </span></span>
<span><span class="co">#&gt; [129] Rsamtools_2.16.0            cli_3.6.1                  </span></span>
<span><span class="co">#&gt; [131] compiler_4.3.0              rlang_1.1.1                </span></span>
<span><span class="co">#&gt; [133] crayon_1.5.2                abdiv_0.2.0                </span></span>
<span><span class="co">#&gt; [135] future.apply_1.10.0         plyr_1.8.8                 </span></span>
<span><span class="co">#&gt; [137] fs_1.6.2                    stringi_1.7.12             </span></span>
<span><span class="co">#&gt; [139] BiocParallel_1.34.1         viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt; [141] deldir_1.0-6                Biostrings_2.68.0          </span></span>
<span><span class="co">#&gt; [143] munsell_0.5.0               lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt; [145] spatstat.geom_3.1-0         Matrix_1.5-4               </span></span>
<span><span class="co">#&gt; [147] hms_1.1.3                   patchwork_1.1.2            </span></span>
<span><span class="co">#&gt; [149] bit64_4.0.5                 future_1.32.0              </span></span>
<span><span class="co">#&gt; [151] ggplot2_3.4.2               shiny_1.7.4                </span></span>
<span><span class="co">#&gt; [153] SummarizedExperiment_1.30.1 ROCR_1.0-11                </span></span>
<span><span class="co">#&gt; [155] igraph_1.4.2                broom_1.0.4                </span></span>
<span><span class="co">#&gt; [157] memoise_2.0.1               bslib_0.4.2                </span></span>
<span><span class="co">#&gt; [159] bit_4.0.5</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
