<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="djvdj">
<title>Repertoire Similarity â€¢ djvdj</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.6/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.6/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Repertoire Similarity">
<meta property="og:description" content="djvdj">
<meta property="og:image" content="https://rnabioco.github.io/djvdj/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">djvdj</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Data import and manipulation</h6>
    <a class="dropdown-item" href="../articles/import.html">Data Import</a>
    <a class="dropdown-item" href="../articles/mutate.html">Data Manipulation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>VDJ analysis</h6>
    <a class="dropdown-item" href="../articles/frequency.html">Clonotype Frequencies</a>
    <a class="dropdown-item" href="../articles/diversity.html">Repertoire Diversity</a>
    <a class="dropdown-item" href="../articles/similarity.html">Repertoire Similarity</a>
    <a class="dropdown-item" href="../articles/gene-usage.html">Gene Usage</a>
    <a class="dropdown-item" href="../articles/clustering.html">CDR3 Clustering</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Visualization</h6>
    <a class="dropdown-item" href="../articles/plotting.html">Plotting</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/djvdj/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Repertoire Similarity</h1>
            
            <h4 data-toc-skip class="date">May 08 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/djvdj/blob/HEAD/vignettes/similarity.Rmd" class="external-link"><code>vignettes/similarity.Rmd</code></a></small>
      <div class="d-none name"><code>similarity.Rmd</code></div>
    </div>

    
    
<p><br></p>
<p>This vignette provides detailed examples for quantifying repertoire
overlap between samples. For the examples shown below, we use data for
splenocytes from BL6 and MD4 mice collected using the 10X Genomics
scRNA-seq platform. MD4 B cells are monoclonal and specifically bind hen
egg lysozyme.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rnabioco.github.io/djvdj/">djvdj</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add V(D)J data to object</span></span>
<span><span class="va">vdj_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  BL6 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/splen/BL6_BCR"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span>,</span>
<span>  MD4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/splen/MD4_BCR"</span>, package <span class="op">=</span> <span class="st">"djvdj"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">so</span> <span class="op">&lt;-</span> <span class="va">splen_so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/import_vdj.html">import_vdj</a></span><span class="op">(</span><span class="va">vdj_dirs</span>, define_clonotypes <span class="op">=</span> <span class="st">"cdr3_gene"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<div class="section level2">
<h2 id="calculating-repertoire-overlap">Calculating repertoire overlap<a class="anchor" aria-label="anchor" href="#calculating-repertoire-overlap"></a>
</h2>
<p>The <code><a href="../reference/calc_similarity.html">calc_similarity()</a></code> function will calculate repertoire
overlap between clusters and add the results in the object meta.data.
This function is designed to specifically work with the R package <a href="https://github.com/kylebittinger/abdiv" class="external-link">abdiv</a>. The similarity
metric to use for calculations can be selected by passing the name of
the function to the <code>method</code> argument. Any beta diversity
function from the abdiv package that takes species counts as input can
be used. Be sure to read the documentation for the function you are
using to ensure it is appropriate for your analysis.</p>
<p>In this example we are calculating the Jaccard dissimilarity index
for BL6 and MD4 samples. With this metric, a value close to 0 indicates
that the samples have a high number of shared clonotypes.</p>
<p>Jaccard dissimilarity is calculated using the following equation,
where <span class="math inline">\(a\)</span> is the number of species
present in both x and y, <span class="math inline">\(b\)</span> is the
number of species present in y but not x, and <span class="math inline">\(c\)</span> is the number of species present in x
but not y.</p>
<p><span class="math display">\[ 1 - {a \over a + b + c} \]</span></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_similarity.html">calc_similarity</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    method      <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/jaccard.html" class="external-link">jaccard</a></span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Similarity metrics can also be calculated for a specific chain. To do
this, the column passed to the <code>data_col</code> argument must
contain per-chain data, such as CDR3 amino acid or nucleotide sequences.
In this example similarity is calculated based on heavy chain CDR3
sequences.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_similarity.html">calc_similarity</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"cdr3_nt"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    chain       <span class="op">=</span> <span class="st">"IGH"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Instead of adding the results to the object meta.data, a matrix can
also be returned.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_similarity.html">calc_similarity</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    return_mat  <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt;           BL6-1     BL6-2     BL6-3     MD4-1     MD4-2     MD4-3</span></span>
<span><span class="co">#&gt; BL6-1 0.0000000 0.9639640 0.9696970 1.0000000 1.0000000 0.9830508</span></span>
<span><span class="co">#&gt; BL6-2 0.9639640 0.0000000 0.9813084 1.0000000 1.0000000 0.9848485</span></span>
<span><span class="co">#&gt; BL6-3 0.9696970 0.9813084 0.0000000 0.9807692 1.0000000 1.0000000</span></span>
<span><span class="co">#&gt; MD4-1 1.0000000 1.0000000 0.9807692 0.0000000 0.8571429 0.9000000</span></span>
<span><span class="co">#&gt; MD4-2 1.0000000 1.0000000 1.0000000 0.8571429 0.0000000 0.8750000</span></span>
<span><span class="co">#&gt; MD4-3 0.9830508 0.9848485 1.0000000 0.9000000 0.8750000 0.0000000</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="plotting-overlap">Plotting overlap<a class="anchor" aria-label="anchor" href="#plotting-overlap"></a>
</h2>
<p>The <code><a href="../reference/plot_similarity.html">plot_similarity()</a></code> function will create plots
summarizing repertoire overlap between samples. By default, Jaccard
dissimilarity will be calculated and plotted as a heatmap. For this
example, none of the samples show strong overlap. This may seem
surprising since the MD4 samples are mainly composed of a single
clonotype and should have a very similar repertoire. However, the
Jaccard index is only measuring the number of overlapping clonotypes and
is not influenced by clonotype abundance.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_similarity.html">plot_similarity</a></span><span class="op">(</span></span>
<span>    data_col <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="similarity_files/figure-html/unnamed-chunk-5-1.png" width="576"></p>
<p>A metric that takes into account clonotype abundance is the
Horn-Morisita index. This metric measures the probability that
clonotypes drawn from each sample will be different. Values close to 0
indicate high similarity. In this example we see that when clonotype
abundance is taken into account, the MD4 samples appear very similar to
each other.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_similarity.html">plot_similarity</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    method      <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/morisita.html" class="external-link">horn_morisita</a></span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="similarity_files/figure-html/unnamed-chunk-6-1.png" width="576"></p>
<p>The appearance of the heatmap can be modified with additional
arguments. Setting <code>cluster_heatmap</code> to <code>FALSE</code>
will remove the dendrograph. The <code>remove_upper_triangle</code>
argument can be used to only plot the lower triangle of the heatmap. The
<code>plot_colors</code> argument will adjust the color gradient.
Additional parameters can be passed directly to
<code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_similarity.html">plot_similarity</a></span><span class="op">(</span></span>
<span>    data_col        <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col     <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    method          <span class="op">=</span> <span class="fu">abdiv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/abdiv/man/morisita.html" class="external-link">horn_morisita</a></span>,</span>
<span>    plot_colors     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#3182bd"</span>, <span class="st">"white"</span>, <span class="st">"#fec44f"</span><span class="op">)</span>,</span>
<span>    cluster_heatmap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    remove_upper_triangle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    </span>
<span>    name <span class="op">=</span> <span class="st">"Horn-Morisita"</span>  <span class="co"># parameter to pass to Heatmap()</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="similarity_files/figure-html/unnamed-chunk-7-1.png" width="576"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="circos-plot">Circos plot<a class="anchor" aria-label="anchor" href="#circos-plot"></a>
</h2>
<p>A circos plot can be created by setting the <code>method</code>
argument to â€˜circosâ€™. This plot will summarize the number of clonotypes
overlapping between samples. The <code>group_col</code> argument can be
used to split the graph into distinct sections based on a grouping
variable.</p>
<p>The number of overlapping clonotypes is shown as the axis labels for
each sample. The width of each link reflects the number of clonotypes
shared between the samples. Labels can be rotated to eliminate
overlapping text using the <code>rotate_labels</code> argument.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_similarity.html">plot_similarity</a></span><span class="op">(</span></span>
<span>    data_col      <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col   <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    group_col     <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    method        <span class="op">=</span> <span class="st">"circos"</span>,</span>
<span>    rotate_labels <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="similarity_files/figure-html/unnamed-chunk-8-1.png" width="576"></p>
<p>Plot colors can be modified using the <code>plot_colors</code>
argument, additional parameters can be passed directly to
<code><a href="https://rdrr.io/pkg/circlize/man/chordDiagram.html" class="external-link">circlize::chordDiagram()</a></code>. In this example we scale the plot
so each sample is the same width.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_similarity.html">plot_similarity</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    group_col   <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    method      <span class="op">=</span> <span class="st">"circos"</span>,</span>
<span>    plot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">10</span>, <span class="st">"Spectral"</span><span class="op">)</span>,</span>
<span>    </span>
<span>    scale <span class="op">=</span> <span class="cn">TRUE</span>  <span class="co"># parameter to pass to chordDiagram()</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="similarity_files/figure-html/unnamed-chunk-9-1.png" width="576"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="mds-plot">MDS plot<a class="anchor" aria-label="anchor" href="#mds-plot"></a>
</h2>
<p>Multidimensional scaling (MDS) can be used to visualize the overall
similarity between repertoires. The <code><a href="../reference/calc_mds.html">calc_mds()</a></code> function
will calculate MDS coordinates using either the Jaccard dissimilarity
index or the Horn-Morisita index. The method can be specified using the
<code>method</code> argument. MDS coordinates will get added to the
meta.data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_vdj</span> <span class="op">&lt;-</span> <span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/calc_mds.html">calc_mds</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    method      <span class="op">=</span> <span class="st">"horn_morisita"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/plot_mds.html">plot_mds()</a></code> function will create an MDS plot with
labels added for each sample.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_mds.html">plot_mds</a></span><span class="op">(</span></span>
<span>    data_col <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="similarity_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
<p>ggplot2 functions can be used to further adjust plot aesthetics, or
additional arguments can be passed directly to ggplot2,
e.g.Â <code>size</code>, <code>shape</code>, etc.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_mds.html">plot_mds</a></span><span class="op">(</span></span>
<span>    data_col    <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    </span>
<span>    size  <span class="op">=</span> <span class="fl">4</span>,    <span class="co"># parameters to pass to ggplot </span></span>
<span>    shape <span class="op">=</span> <span class="fl">2</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="similarity_files/figure-html/unnamed-chunk-12-1.png" width="576"></p>
<p>To remove sample labels, set <code>label_points</code> to
<code>FALSE</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_mds.html">plot_mds</a></span><span class="op">(</span></span>
<span>    data_col     <span class="op">=</span> <span class="st">"clonotype_id"</span>,</span>
<span>    cluster_col  <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>    label_points <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="similarity_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] RColorBrewer_1.1-3 ggplot2_3.4.2      SeuratObject_4.1.3</span></span>
<span><span class="co">#&gt; [4] Seurat_4.3.0       djvdj_0.1.0       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RcppAnnoy_0.0.20            splines_4.3.0              </span></span>
<span><span class="co">#&gt;   [3] later_1.3.1                 bitops_1.0-7               </span></span>
<span><span class="co">#&gt;   [5] tibble_3.2.1                polyclip_1.10-4            </span></span>
<span><span class="co">#&gt;   [7] lifecycle_1.0.3             doParallel_1.0.17          </span></span>
<span><span class="co">#&gt;   [9] rprojroot_2.0.3             globals_0.16.2             </span></span>
<span><span class="co">#&gt;  [11] lattice_0.21-8              vroom_1.6.3                </span></span>
<span><span class="co">#&gt;  [13] MASS_7.3-58.4               backports_1.4.1            </span></span>
<span><span class="co">#&gt;  [15] magrittr_2.0.3              plotly_4.10.1              </span></span>
<span><span class="co">#&gt;  [17] sass_0.4.6                  rmarkdown_2.21             </span></span>
<span><span class="co">#&gt;  [19] jquerylib_0.1.4             yaml_2.3.7                 </span></span>
<span><span class="co">#&gt;  [21] httpuv_1.6.9                sctransform_0.3.5          </span></span>
<span><span class="co">#&gt;  [23] sp_1.6-0                    spatstat.sparse_3.0-1      </span></span>
<span><span class="co">#&gt;  [25] reticulate_1.28             cowplot_1.1.1              </span></span>
<span><span class="co">#&gt;  [27] pbapply_1.7-0               abind_1.4-5                </span></span>
<span><span class="co">#&gt;  [29] zlibbioc_1.46.0             Rtsne_0.16                 </span></span>
<span><span class="co">#&gt;  [31] GenomicRanges_1.52.0        purrr_1.0.1                </span></span>
<span><span class="co">#&gt;  [33] BiocGenerics_0.46.0         RCurl_1.98-1.12            </span></span>
<span><span class="co">#&gt;  [35] circlize_0.4.15             GenomeInfoDbData_1.2.10    </span></span>
<span><span class="co">#&gt;  [37] IRanges_2.34.0              S4Vectors_0.38.1           </span></span>
<span><span class="co">#&gt;  [39] ggrepel_0.9.3               irlba_2.3.5.1              </span></span>
<span><span class="co">#&gt;  [41] listenv_0.9.0               spatstat.utils_3.0-2       </span></span>
<span><span class="co">#&gt;  [43] goftest_1.2-3               spatstat.random_3.1-4      </span></span>
<span><span class="co">#&gt;  [45] fitdistrplus_1.1-11         parallelly_1.35.0          </span></span>
<span><span class="co">#&gt;  [47] pkgdown_2.0.7               leiden_0.4.3               </span></span>
<span><span class="co">#&gt;  [49] codetools_0.2-19            DelayedArray_0.26.2        </span></span>
<span><span class="co">#&gt;  [51] tidyselect_1.2.0            shape_1.4.6                </span></span>
<span><span class="co">#&gt;  [53] farver_2.1.1                matrixStats_0.63.0         </span></span>
<span><span class="co">#&gt;  [55] stats4_4.3.0                spatstat.explore_3.1-0     </span></span>
<span><span class="co">#&gt;  [57] jsonlite_1.8.4              GetoptLong_1.0.5           </span></span>
<span><span class="co">#&gt;  [59] ellipsis_0.3.2              ggtrace_0.2.0              </span></span>
<span><span class="co">#&gt;  [61] progressr_0.13.0            ggridges_0.5.4             </span></span>
<span><span class="co">#&gt;  [63] survival_3.5-5              iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [65] systemfonts_1.0.4           foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [67] tools_4.3.0                 ragg_1.2.5                 </span></span>
<span><span class="co">#&gt;  [69] ica_1.0-3                   Rcpp_1.0.10                </span></span>
<span><span class="co">#&gt;  [71] glue_1.6.2                  gridExtra_2.3              </span></span>
<span><span class="co">#&gt;  [73] xfun_0.39                   MatrixGenerics_1.12.0      </span></span>
<span><span class="co">#&gt;  [75] GenomeInfoDb_1.36.0         dplyr_1.1.2                </span></span>
<span><span class="co">#&gt;  [77] withr_2.5.0                 fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [79] boot_1.3-28.1               fansi_1.0.4                </span></span>
<span><span class="co">#&gt;  [81] digest_0.6.31               R6_2.5.1                   </span></span>
<span><span class="co">#&gt;  [83] mime_0.12                   textshaping_0.3.6          </span></span>
<span><span class="co">#&gt;  [85] colorspace_2.1-0            scattermore_1.0            </span></span>
<span><span class="co">#&gt;  [87] tensor_1.5                  spatstat.data_3.0-1        </span></span>
<span><span class="co">#&gt;  [89] utf8_1.2.3                  tidyr_1.3.0                </span></span>
<span><span class="co">#&gt;  [91] generics_0.1.3              data.table_1.14.8          </span></span>
<span><span class="co">#&gt;  [93] httr_1.4.5                  htmlwidgets_1.6.2          </span></span>
<span><span class="co">#&gt;  [95] S4Arrays_1.0.1              uwot_0.1.14                </span></span>
<span><span class="co">#&gt;  [97] pkgconfig_2.0.3             gtable_0.3.3               </span></span>
<span><span class="co">#&gt;  [99] ComplexHeatmap_2.16.0       lmtest_0.9-40              </span></span>
<span><span class="co">#&gt; [101] SingleCellExperiment_1.22.0 XVector_0.40.0             </span></span>
<span><span class="co">#&gt; [103] htmltools_0.5.5             clue_0.3-64                </span></span>
<span><span class="co">#&gt; [105] scales_1.2.1                Biobase_2.60.0             </span></span>
<span><span class="co">#&gt; [107] png_0.1-8                   knitr_1.42                 </span></span>
<span><span class="co">#&gt; [109] tzdb_0.3.0                  reshape2_1.4.4             </span></span>
<span><span class="co">#&gt; [111] rjson_0.2.21                nlme_3.1-162               </span></span>
<span><span class="co">#&gt; [113] cachem_1.0.8                zoo_1.8-12                 </span></span>
<span><span class="co">#&gt; [115] GlobalOptions_0.1.2         stringr_1.5.0              </span></span>
<span><span class="co">#&gt; [117] KernSmooth_2.23-20          parallel_4.3.0             </span></span>
<span><span class="co">#&gt; [119] miniUI_0.1.1.1              desc_1.4.2                 </span></span>
<span><span class="co">#&gt; [121] pillar_1.9.0                grid_4.3.0                 </span></span>
<span><span class="co">#&gt; [123] vctrs_0.6.2                 RANN_2.6.1                 </span></span>
<span><span class="co">#&gt; [125] promises_1.2.0.1            xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [127] cluster_2.1.4               evaluate_0.20              </span></span>
<span><span class="co">#&gt; [129] readr_2.1.4                 cli_3.6.1                  </span></span>
<span><span class="co">#&gt; [131] compiler_4.3.0              rlang_1.1.1                </span></span>
<span><span class="co">#&gt; [133] crayon_1.5.2                abdiv_0.2.0                </span></span>
<span><span class="co">#&gt; [135] future.apply_1.10.0         labeling_0.4.2             </span></span>
<span><span class="co">#&gt; [137] plyr_1.8.8                  fs_1.6.2                   </span></span>
<span><span class="co">#&gt; [139] stringi_1.7.12              viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt; [141] deldir_1.0-6                munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [143] lazyeval_0.2.2              spatstat.geom_3.1-0        </span></span>
<span><span class="co">#&gt; [145] Matrix_1.5-4                hms_1.1.3                  </span></span>
<span><span class="co">#&gt; [147] patchwork_1.1.2             bit64_4.0.5                </span></span>
<span><span class="co">#&gt; [149] future_1.32.0               shiny_1.7.4                </span></span>
<span><span class="co">#&gt; [151] highr_0.10                  SummarizedExperiment_1.30.1</span></span>
<span><span class="co">#&gt; [153] ROCR_1.0-11                 igraph_1.4.2               </span></span>
<span><span class="co">#&gt; [155] broom_1.0.4                 memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [157] bslib_0.4.2                 bit_4.0.5</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
